// @amazon-devices/kepler-module-manifest-builder/dist/test/index.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const src_1 = require("../src");
const manifestBuilder = __importStar(require("../src/manifest-builder"));
const logger_1 = require("../src/utils/logger");
const npmUtils = __importStar(require("../src/utils/npm-utils"));
// Mock all external dependencies
jest.mock('child_process', () => ({
    execSync: jest.fn(),
}));
jest.mock('fs', () => ({
    existsSync: jest.fn(),
    readFileSync: jest.fn(),
}));
jest.mock('path', () => ({
    join: jest.fn((...args) => args.join('/')),
}));
jest.mock('../src/manifest-builder', () => ({
    buildManifestForPackage: jest.fn(),
}));
jest.mock('../src/utils/npm-utils', () => ({
    getVersionsForLibrary: jest.fn(),
    getVersionsIncludedFor: jest.fn(),
}));
describe('Kepler Module Manifest Builder', () => {
    const mockLogger = {
        error: jest.spyOn(logger_1.logger, 'error').mockImplementation(),
        warn: jest.spyOn(logger_1.logger, 'warn').mockImplementation(),
        info: jest.spyOn(logger_1.logger, 'info').mockImplementation(),
        log: jest.spyOn(logger_1.logger, 'log').mockImplementation(),
        setSilent: jest.spyOn(logger_1.logger, 'setSilent').mockImplementation(),
    };
    beforeEach(() => {
        jest.clearAllMocks();
    });
    describe('buildManifest', () => {
        const mockProjectDir = '/test/project';
        const mockOverridesPath = '/test/overrides.json';
        it('Backward Compatibility: Should call buildManifestForRootPaths, wrapping the string-arg in a list', () => {
            // Given
            const mockManifest = {
                needs: ['module1'],
                wants: ['module2'],
            };
            fs.existsSync.mockReturnValue(true);
            manifestBuilder.buildManifestForPackage.mockReturnValue(mockManifest);
            // When
            const result = (0, src_1.buildManifest)(mockProjectDir);
            // Then
            expect(result).toEqual(mockManifest);
            expect(manifestBuilder.buildManifestForPackage).toHaveBeenCalledWith([mockProjectDir], {}, undefined, undefined);
        });
    });
    describe('buildManifestForRootPaths', () => {
        const mockProjectDir = '/test/project';
        const mockOverridesPath = '/test/overrides.json';
        it('should throw error when directory list is empty', () => {
            expect(() => (0, src_1.buildManifestForRootPaths)([])).toThrow(`Required option directory was not provided.`);
        });
        it('should build manifest without overrides', () => {
            // Given
            const mockManifest = {
                needs: ['module1'],
                wants: ['module2'],
            };
            fs.existsSync.mockReturnValue(true);
            manifestBuilder.buildManifestForPackage.mockReturnValue(mockManifest);
            // When
            const result = (0, src_1.buildManifestForRootPaths)([mockProjectDir]);
            // Then
            expect(result).toEqual(mockManifest);
            expect(manifestBuilder.buildManifestForPackage).toHaveBeenCalledWith([mockProjectDir], {}, undefined, undefined);
        });
        it('should build manifest with overrides', () => {
            // Given
            const mockManifest = {
                needs: ['module1'],
                wants: ['module2'],
            };
            const mockPackageJson = { some: 'config' };
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify(mockPackageJson));
            manifestBuilder.buildManifestForPackage.mockReturnValue(mockManifest);
            // When
            const result = (0, src_1.buildManifestForRootPaths)([mockProjectDir], mockOverridesPath);
            // Then
            expect(result).toEqual(mockManifest);
            expect(manifestBuilder.buildManifestForPackage).toHaveBeenCalledWith([mockProjectDir], mockPackageJson, undefined, undefined);
        });
    });
    describe('queryNpm', () => {
        const mockLibraryName = '@amazon-devices/react-native-gesture-handler';
        const mockVersion = '1.0.0';
        describe('version listing mode', () => {
            it('should return available versions when no version specified', () => {
                // Given
                const mockVersions = ['1.0.0', '1.1.0', '2.0.0'];
                npmUtils.getVersionsForLibrary.mockReturnValue(mockVersions);
                // When
                const result = (0, src_1.queryNpm)(mockLibraryName);
                // Then
                expect(result).toEqual({ versions: mockVersions });
                expect(npmUtils.getVersionsForLibrary).toHaveBeenCalledWith(mockLibraryName, 'npm');
                expect(mockLogger.info).toHaveBeenCalledWith(`Available versions to install for ${mockLibraryName}:`);
                mockVersions.forEach((version) => {
                    expect(mockLogger.info).toHaveBeenCalledWith(`*  ${version}`);
                });
            });
            it('should handle single version response', () => {
                // Given
                const mockSingleVersion = '1.0.0';
                npmUtils.getVersionsForLibrary.mockReturnValue(mockSingleVersion);
                // When
                const result = (0, src_1.queryNpm)(mockLibraryName);
                // Then
                expect(result).toEqual({ versions: mockVersion });
                expect(mockLogger.info).toHaveBeenCalledWith(`*  ${mockSingleVersion}`);
            });
        });
        describe('version specific mode', () => {
            it('should return included versions for specific version', () => {
                // Given
                const mockIncludedVersions = {
                    includesModules: ['module1', 'module2'],
                    versionMap: {
                        '1.0.0': 'module1',
                        '1.1.0': 'module2',
                    },
                };
                npmUtils.getVersionsIncludedFor.mockReturnValue(mockIncludedVersions);
                // When
                const result = (0, src_1.queryNpm)(mockLibraryName, mockVersion);
                // Then
                expect(result).toEqual({
                    includedVersions: mockIncludedVersions,
                });
                expect(npmUtils.getVersionsIncludedFor).toHaveBeenCalledWith(mockLibraryName, mockVersion, 'npm');
                expect(mockLogger.info).toHaveBeenCalledTimes(3);
                expect(mockLogger.info).toHaveBeenNthCalledWith(1, `Backward-compatible versions available when using ${mockLibraryName}@${mockVersion}:`);
                expect(mockLogger.info).toHaveBeenNthCalledWith(2, `*  1.0.0`);
                expect(mockLogger.info).toHaveBeenNthCalledWith(3, `*  1.1.0`);
            });
            it('should support custom package managers', () => {
                // Given
                const customPM = 'yarn';
                const mockIncludedVersions = {
                    includesModules: ['module1'],
                    versionMap: { '1.0.0': 'module1' },
                };
                npmUtils.getVersionsIncludedFor.mockReturnValue(mockIncludedVersions);
                // When
                (0, src_1.queryNpm)(mockLibraryName, mockVersion, customPM);
                // Then
                expect(npmUtils.getVersionsIncludedFor).toHaveBeenCalledWith(mockLibraryName, mockVersion, customPM);
            });
        });
        describe('error handling', () => {
            it('should propagate errors from getVersionsForLibrary', () => {
                // Given
                npmUtils.getVersionsForLibrary.mockImplementation(() => {
                    throw new Error('Version fetch failed');
                });
                // Then
                expect(() => (0, src_1.queryNpm)(mockLibraryName)).toThrow('Version fetch failed');
            });
            it('should propagate errors from getVersionsIncludedFor', () => {
                // Given
                npmUtils.getVersionsIncludedFor.mockImplementation(() => {
                    throw new Error('Version processing failed');
                });
                // Then
                expect(() => (0, src_1.queryNpm)(mockLibraryName, mockVersion)).toThrow('Version processing failed');
            });
        });
    });
});
