// @amazon-devices/kepler-module-manifest-builder/dist/test/manifest-builder-scenarios.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const manifest_builder_1 = require("../src/manifest-builder");
const file_utils_1 = require("../src/utils/file-utils");
const logger_1 = require("../src/utils/logger");
// Mock all external dependencies
jest.mock('fs', () => ({
    existsSync: jest.fn(),
    readdirSync: jest.fn(),
    readFileSync: jest.fn(),
    statSync: jest.fn(), // Used by scanNodeModulesRecursively
}));
jest.mock('../src/utils/file-utils', () => ({
    ...jest.requireActual('../src/utils/file-utils'),
    parseJsonFile: jest.fn(), // Used by buildManifestForPackage
}));
jest.mock('path', () => ({
    join: jest.fn((...args) => args.join('/')),
}));
describe('manifest-builder Scenario Tests', () => {
    const depth = 0;
    const mockLogger = {
        error: jest.spyOn(logger_1.logger, 'error').mockImplementation(),
        warn: jest.spyOn(logger_1.logger, 'warn').mockImplementation(),
        info: jest.spyOn(logger_1.logger, 'info').mockImplementation(),
    };
    const mockProjectDir = '/test/project';
    describe('Given App that requires RNGH 2.4.0 and Reanimated 2.5.0.. but Reanimated 2.5.0 depends on RNGH 2.5.0', () => {
        it('Define needs as RNGH 2.4.0 and Reanimated 2.4.0 if no overrides are present', () => {
            fs.existsSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                // Specific to this case, RNGH will not have nested node_modules, but reanimated will..
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'node_modules') {
                    return false;
                }
                return true;
            });
            fs.statSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                return {
                    isDirectory: () => {
                        // Depth of 2 allows recursing to `node_modules/@amazon-devices/<library_name>.
                        if (lastSegment === '@amzn' ||
                            lastSegment in mockLibraries) {
                            return true;
                        }
                        if (penultimateSegment ===
                            'react-native-gesture-handler' &&
                            lastSegment === 'node_modules') {
                            return false;
                        }
                        if (penultimateSegment === 'react-native-reanimated' &&
                            lastSegment === 'node_modules') {
                            return true;
                        }
                        // Anything that isn't package.json or kepler-compatibility.json is
                        // a directory, from the perspective of unit test
                        return !fPath.endsWith('.json');
                    },
                };
            });
            // Given two libraries, RNGH and Reanimated
            const mockLibraries = [
                'react-native-gesture-handler',
                'react-native-reanimated',
            ];
            // Both Libraries have defined a `kepler-compatibility.json`
            const mockRNGH24CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                },
            };
            const mockRNGH25CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                    '2.5.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                },
            };
            const mockReanimatedCompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.reanimated_2@IReanimated_0',
                    '2.1.0': '/com.amazon.kepler.reanimated_2@IReanimated_1',
                    '2.2.0': '/com.amazon.kepler.reanimated_2@IReanimated_2',
                    '2.3.0': '/com.amazon.kepler.reanimated_2@IReanimated_3',
                    '2.4.0': '/com.amazon.kepler.reanimated_2@IReanimated_4',
                },
            };
            const mockFiles = ['kepler-compatibility.json', 'package.json'];
            fs.readdirSync // node_modules
                .mockReturnValueOnce(['@amzn']) // node_modules/@amzn
                .mockReturnValueOnce([
                'react-native-gesture-handler',
                'react-native-reanimated',
            ]) // node_modules/@amazon-devices/[react-native-gesture-handler]
                .mockReturnValueOnce(['@amzn']) // node_modules/@amazon-devices/react-native-reanimated/@amzn
                .mockReturnValueOnce(['react-native-gesture-handler']); // node_modules/@amazon-devices/react-native-gesture-handler/@amazon-devices/react-native-gesture-handler
            let visitedOnce = false;
            file_utils_1.parseJsonFile.mockImplementation((packageJsonPath) => {
                const fileSplits = packageJsonPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'kepler-compatibility.json') {
                    const res = visitedOnce
                        ? mockRNGH25CompatibilityJson
                        : mockRNGH24CompatibilityJson;
                    visitedOnce = true;
                    return res;
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'kepler-compatibility.json') {
                    return mockReanimatedCompatibilityJson;
                }
            });
            // When calling buildManifestForPackage
            // When
            const result = (0, manifest_builder_1.buildManifestForPackage)([mockProjectDir]);
            // Then
            expect(result.needs).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                '/com.amazon.kepler.reanimated_2@IReanimated_4',
            ]);
            expect(result.wants).toEqual([]);
            expect(result.keplerModuleMap).toEqual({
                '@amazon-devices/react-native-gesture-handler': {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                    '2.5.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                },
                '@amazon-devices/react-native-reanimated': {
                    '2.0.0': '/com.amazon.kepler.reanimated_2@IReanimated_0',
                    '2.1.0': '/com.amazon.kepler.reanimated_2@IReanimated_1',
                    '2.2.0': '/com.amazon.kepler.reanimated_2@IReanimated_2',
                    '2.3.0': '/com.amazon.kepler.reanimated_2@IReanimated_3',
                    '2.4.0': '/com.amazon.kepler.reanimated_2@IReanimated_4',
                },
            });
        });
    });
    describe('Given App that requires RNGH 2.5.0 and Reanimated 2.4.0.. but Reanimated 2.4.0 depends on RNGH 2.4.0', () => {
        it('Define needs as RNGH 2.5.0 and Reanimated 2.4.0 if no overrides are present', () => {
            fs.existsSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                // Specific to this case, RNGH will not have nested node_modules, but reanimated will..
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'node_modules') {
                    return false;
                }
                return true;
            });
            fs.statSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                return {
                    isDirectory: () => {
                        // Depth of 2 allows recursing to `node_modules/@amazon-devices/<library_name>.
                        if (lastSegment === '@amzn' ||
                            lastSegment in mockLibraries) {
                            return true;
                        }
                        if (penultimateSegment ===
                            'react-native-gesture-handler' &&
                            lastSegment === 'node_modules') {
                            return false;
                        }
                        if (penultimateSegment === 'react-native-reanimated' &&
                            lastSegment === 'node_modules') {
                            return true;
                        }
                        // Anything that isn't package.json or kepler-compatibility.json is
                        // a directory, from the perspective of unit test
                        return !fPath.endsWith('.json');
                    },
                };
            });
            // Given two libraries, RNGH and Reanimated
            const mockLibraries = [
                'react-native-gesture-handler',
                'react-native-reanimated',
            ];
            // Both Libraries have defined a `kepler-compatibility.json`
            const mockRNGH24CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                },
            };
            const mockRNGH25CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                    '2.5.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                },
            };
            const mockReanimatedCompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.reanimated_2@IReanimated_0',
                    '2.1.0': '/com.amazon.kepler.reanimated_2@IReanimated_1',
                    '2.2.0': '/com.amazon.kepler.reanimated_2@IReanimated_2',
                    '2.3.0': '/com.amazon.kepler.reanimated_2@IReanimated_3',
                    '2.4.0': '/com.amazon.kepler.reanimated_2@IReanimated_4',
                },
            };
            const mockFiles = ['kepler-compatibility.json', 'package.json'];
            fs.readdirSync // node_modules
                .mockReturnValueOnce(['@amzn']) // node_modules/@amzn
                .mockReturnValueOnce([
                'react-native-gesture-handler',
                'react-native-reanimated',
            ]) // node_modules/@amazon-devices/[react-native-gesture-handler]
                .mockReturnValueOnce(['@amzn']) // node_modules/@amazon-devices/react-native-reanimated/@amzn
                .mockReturnValueOnce(['react-native-gesture-handler']); // node_modules/@amazon-devices/react-native-gesture-handler/@amazon-devices/react-native-gesture-handler
            let visitedOnce = false;
            file_utils_1.parseJsonFile.mockImplementation((packageJsonPath) => {
                const fileSplits = packageJsonPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'kepler-compatibility.json') {
                    const res = visitedOnce
                        ? mockRNGH24CompatibilityJson
                        : mockRNGH25CompatibilityJson;
                    visitedOnce = true;
                    return res;
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'kepler-compatibility.json') {
                    return mockReanimatedCompatibilityJson;
                }
            });
            // When calling buildManifestForPackage
            // When
            const result = (0, manifest_builder_1.buildManifestForPackage)([mockProjectDir]);
            // Then
            expect(result.needs).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                '/com.amazon.kepler.reanimated_2@IReanimated_4',
            ]);
            expect(result.wants).toEqual([]);
        });
        it('Define needs as RNGH 2.0.0 and Reanimated 2.4.0, wants as RNGH 2.4.0 if overrides ask for RNGH specifically', () => {
            // Given two libraries, RNGH and Reanimated
            const mockLibraries = [
                'react-native-gesture-handler',
                'react-native-reanimated',
            ];
            // Both Libraries have defined a `kepler-compatibility.json`
            const mockRNGH24CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                },
            };
            const mockRNGH25CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                    '2.5.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                },
            };
            const mockReanimatedCompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.reanimated_2@IReanimated_0',
                    '2.1.0': '/com.amazon.kepler.reanimated_2@IReanimated_1',
                    '2.2.0': '/com.amazon.kepler.reanimated_2@IReanimated_2',
                    '2.3.0': '/com.amazon.kepler.reanimated_2@IReanimated_3',
                    '2.4.0': '/com.amazon.kepler.reanimated_2@IReanimated_4',
                },
            };
            const mockReanimatedBackwardCompatibilityJson = {
                kepler: {
                    backwardCompatibility: {
                        '@amazon-devices/react-native-gesture-handler': {
                            needs: '2.0.0',
                            wants: '^2.0.0',
                        },
                    },
                },
            };
            fs.existsSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                // Specific to this case, RNGH will not have nested node_modules, but reanimated will..
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'node_modules') {
                    return false;
                }
                return true;
            });
            fs.statSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                return {
                    isDirectory: () => {
                        // Depth of 2 allows recursing to `node_modules/@amazon-devices/<library_name>.
                        if (lastSegment === '@amzn' ||
                            lastSegment in mockLibraries) {
                            return true;
                        }
                        if (penultimateSegment ===
                            'react-native-gesture-handler' &&
                            lastSegment === 'node_modules') {
                            return false;
                        }
                        if (penultimateSegment === 'react-native-reanimated' &&
                            lastSegment === 'node_modules') {
                            return true;
                        }
                        // Anything that isn't package.json or kepler-compatibility.json is
                        // a directory, from the perspective of unit test
                        return !fPath.endsWith('.json');
                    },
                };
            });
            fs.readdirSync // node_modules
                .mockReturnValueOnce(['@amzn']) // node_modules/@amzn
                .mockReturnValueOnce([
                'react-native-gesture-handler',
                'react-native-reanimated', // Represents 2.5.0
            ]) // node_modules/@amazon-devices/[gesture-handler, reanimated]
                .mockReturnValueOnce(['@amzn']) // node_modules/@amazon-devices/react-native-reanimated/@amzn
                .mockReturnValueOnce(['react-native-gesture-handler']); // node_modules/@amazon-devices/react-native-gesture-handler/@amazon-devices/react-native-gesture-handler -- Represents 2.5.0
            let firstVisit = true;
            file_utils_1.parseJsonFile.mockImplementation((packageJsonPath) => {
                const fileSplits = packageJsonPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'kepler-compatibility.json') {
                    const res = firstVisit
                        ? mockRNGH24CompatibilityJson // First visited is `node_modules/@amazon-devices/react-native-gesture-handler`, the variable represents 2.4.0
                        : mockRNGH25CompatibilityJson; // Next visited is `node_modules/@amazon-devices/react-native-reanimated/node_modules/@amazon-devices/react-native-gesture-handler`, the variable represents 2.5.0
                    firstVisit = false;
                    return res;
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'kepler-compatibility.json') {
                    return mockReanimatedCompatibilityJson; // Represents Reanimated @ 2.4.0
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'package.json') {
                    return mockReanimatedBackwardCompatibilityJson;
                }
                else {
                    console.log(`\n---\nMOCK WARNING: Encountered a JSON file that was unmocked: ${packageJsonPath}, returning undefined.\n---\n`);
                }
            });
            // When calling buildManifestForPackage
            const result = (0, manifest_builder_1.buildManifestForPackage)([mockProjectDir], {
                kepler: {
                    backwardCompatibility: {
                        '@amazon-devices/react-native-gesture-handler': {
                            needs: '2.0.0',
                            wants: '^2.0.0',
                        },
                    },
                },
            });
            // Then
            expect(result.needs).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                '/com.amazon.kepler.reanimated_2@IReanimated_4',
            ]);
            expect(result.wants).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
            ]);
        });
        it('Define needs as RNGH 2.0.0 and Reanimated 2.5.0, wants as RNGH 2.4.0 if overrides ask for RNGH < 2.5.0', () => {
            // Given two libraries, RNGH and Reanimated
            const mockLibraries = [
                'react-native-gesture-handler',
                'react-native-reanimated',
            ];
            // Both Libraries have defined a `kepler-compatibility.json`
            const mockRNGH24CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                },
            };
            const mockRNGH25CompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                    '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
                    '2.2.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_2',
                    '2.3.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_3',
                    '2.4.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
                    '2.5.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_5',
                },
            };
            const mockReanimatedCompatibilityJson = {
                versions: {
                    '2.0.0': '/com.amazon.kepler.reanimated_2@IReanimated_0',
                    '2.1.0': '/com.amazon.kepler.reanimated_2@IReanimated_1',
                    '2.2.0': '/com.amazon.kepler.reanimated_2@IReanimated_2',
                    '2.3.0': '/com.amazon.kepler.reanimated_2@IReanimated_3',
                    '2.4.0': '/com.amazon.kepler.reanimated_2@IReanimated_4',
                    '2.5.0': '/com.amazon.kepler.reanimated_2@IReanimated_5',
                },
            };
            const mockReanimatedBackwardCompatibilityJson = {
                kepler: {
                    backwardCompatibility: {
                        '@amazon-devices/react-native-gesture-handler': {
                            needs: '2.0.0',
                            wants: '^2.0.0',
                        },
                    },
                },
            };
            fs.existsSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                // Specific to this case, RNGH will not have nested node_modules, but reanimated will..
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'node_modules') {
                    return false;
                }
                return true;
            });
            fs.statSync.mockImplementation((...args) => {
                const fPath = args[0];
                const fileSplits = fPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                return {
                    isDirectory: () => {
                        // Depth of 2 allows recursing to `node_modules/@amazon-devices/<library_name>.
                        if (lastSegment === '@amzn' ||
                            lastSegment in mockLibraries) {
                            return true;
                        }
                        if (penultimateSegment ===
                            'react-native-gesture-handler' &&
                            lastSegment === 'node_modules') {
                            return false;
                        }
                        if (penultimateSegment === 'react-native-reanimated' &&
                            lastSegment === 'node_modules') {
                            return true;
                        }
                        // Anything that isn't package.json or kepler-compatibility.json is
                        // a directory, from the perspective of unit test
                        return !fPath.endsWith('.json');
                    },
                };
            });
            fs.readdirSync // node_modules
                .mockReturnValueOnce(['@amzn']) // node_modules/@amzn
                .mockReturnValueOnce([
                'react-native-gesture-handler',
                'react-native-reanimated', // Represents 2.5.0
            ]) // node_modules/@amazon-devices/[gesture-handler, reanimated]
                .mockReturnValueOnce(['@amzn']) // node_modules/@amazon-devices/react-native-reanimated/@amzn
                .mockReturnValueOnce(['react-native-gesture-handler']); // node_modules/@amazon-devices/react-native-gesture-handler/@amazon-devices/react-native-gesture-handler -- Represents 2.5.0
            let firstVisit = true;
            file_utils_1.parseJsonFile.mockImplementation((packageJsonPath) => {
                const fileSplits = packageJsonPath.split('/');
                const lastSegment = fileSplits.pop();
                const penultimateSegment = fileSplits.pop();
                if (penultimateSegment === 'react-native-gesture-handler' &&
                    lastSegment === 'kepler-compatibility.json') {
                    const res = firstVisit
                        ? mockRNGH25CompatibilityJson // First visited is `node_modules/@amazon-devices/react-native-gesture-handler`, the variable represents 2.4.0
                        : mockRNGH24CompatibilityJson; // Next visited is `node_modules/@amazon-devices/react-native-reanimated/node_modules/@amazon-devices/react-native-gesture-handler`, the variable represents 2.5.0
                    firstVisit = false;
                    return res;
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'kepler-compatibility.json') {
                    return mockReanimatedCompatibilityJson; // Represents Reanimated @ 2.4.0
                }
                else if (penultimateSegment === 'react-native-reanimated' &&
                    lastSegment === 'package.json') {
                    return mockReanimatedBackwardCompatibilityJson;
                }
                else {
                    console.log(`\n---\nMOCK WARNING: Encountered a JSON file that was unmocked: ${packageJsonPath}, returning undefined.\n---\n`);
                }
            });
            // When calling buildManifestForPackage
            const result = (0, manifest_builder_1.buildManifestForPackage)([mockProjectDir], {
                kepler: {
                    backwardCompatibility: {
                        '@amazon-devices/react-native-gesture-handler': {
                            needs: '2.0.0',
                            wants: '<2.5.0',
                        },
                    },
                },
            });
            // Then
            expect(result.needs).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                '/com.amazon.kepler.reanimated_2@IReanimated_5',
            ]);
            expect(result.wants).toEqual([
                '/com.amazon.kepler.gesture_handler_2@IGestureHandler_4',
            ]);
        });
    });
});
