// @amzn/react-native-kepler/Libraries/NativeComponent/BaseViewConfig.kepler.js
/**
 * Copyright (c) 2021-2025 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @format
 * @flow
 */

/**
 * CHANGELOG:
 * Added enableSynchronousFocusEventsKepler
 * Rename keplerAtlLabels to keplerAltLabels
 * Added accessibilityLabelledBy
 * Added topPressIn,topPressOut to support D-PAD key events
 */

import type {PartialViewConfigWithoutName} from './PlatformBaseViewConfig';
// amznmod_react - Explicitly import, Platform is used in this file to set Transform appropriately
import Platform from '../Utilities/Platform';
import ReactNativeStyleAttributes from '../Components/View/ReactNativeStyleAttributes';
import {DynamicallyInjectedByGestureHandler} from './ViewConfigIgnore';

const bubblingEventTypes = {
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */
  // Bubbling events from UIManagerModuleConstants.java
  topChange: {
    phasedRegistrationNames: {
      captured: 'onChangeCapture',
      bubbled: 'onChange',
    },
  },
  topSelect: {
    phasedRegistrationNames: {
      captured: 'onSelectCapture',
      bubbled: 'onSelect',
    },
  },
  topTouchEnd: {
    phasedRegistrationNames: {
      captured: 'onTouchEndCapture',
      bubbled: 'onTouchEnd',
    },
  },
  topTouchCancel: {
    phasedRegistrationNames: {
      captured: 'onTouchCancelCapture',
      bubbled: 'onTouchCancel',
    },
  },
  topTouchStart: {
    phasedRegistrationNames: {
      captured: 'onTouchStartCapture',
      bubbled: 'onTouchStart',
    },
  },
  topTouchMove: {
    phasedRegistrationNames: {
      captured: 'onTouchMoveCapture',
      bubbled: 'onTouchMove',
    },
  },

  // Experimental/Work in Progress Pointer Events (not yet ready for use)
  topPointerCancel: {
    phasedRegistrationNames: {
      captured: 'onPointerCancelCapture',
      bubbled: 'onPointerCancel',
    },
  },
  topPointerDown: {
    phasedRegistrationNames: {
      captured: 'onPointerDownCapture',
      bubbled: 'onPointerDown',
    },
  },
  topPointerEnter: {
    phasedRegistrationNames: {
      captured: 'onPointerEnterCapture',
      bubbled: 'onPointerEnter',
      skipBubbling: true,
    },
  },
  topPointerLeave: {
    phasedRegistrationNames: {
      captured: 'onPointerLeaveCapture',
      bubbled: 'onPointerLeave',
      skipBubbling: true,
    },
  },
  topPointerMove: {
    phasedRegistrationNames: {
      captured: 'onPointerMoveCapture',
      bubbled: 'onPointerMove',
    },
  },
  topPointerUp: {
    phasedRegistrationNames: {
      captured: 'onPointerUpCapture',
      bubbled: 'onPointerUp',
    },
  },
  topPointerOut: {
    phasedRegistrationNames: {
      captured: 'onPointerOutCapture',
      bubbled: 'onPointerOut',
    },
  },
  topPointerOver: {
    phasedRegistrationNames: {
      captured: 'onPointerOverCapture',
      bubbled: 'onPointerOver',
    },
  },
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   * topAssetDidLoad does not exist in 0.72 and kept from 0.67.3
   */
  topAssetDidLoad: {
    phasedRegistrationNames: {
      bubbled: 'onAssetDidLoad',
      captured: 'onAssetDidLoadCapture',
    },
  },
  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   */
  topBlur: {
    phasedRegistrationNames: {
      bubbled: 'onBlur',
      captured: 'onBlurCapture',
    },
  },
  topEndEditing: {
    phasedRegistrationNames: {
      bubbled: 'onEndEditing',
      captured: 'onEndEditingCapture',
    },
  },
  topKeyPress: {
    phasedRegistrationNames: {
      bubbled: 'onKeyPress',
      captured: 'onKeyPressCapture',
    },
  },
  topSubmitEditing: {
    phasedRegistrationNames: {
      bubbled: 'onSubmitEditing',
      captured: 'onSubmitEditingCapture',
    },
  },
  // amznmod_react - topChangeText does not exist in 0.72, 0.67.3 and 0.64
  topChangeText: {
    phasedRegistrationNames: {
      bubbled: 'onChangeText',
      captured: 'onChangeTextCapture',
    },
  },
  topFocus: {
    phasedRegistrationNames: {
      bubbled: 'onFocus',
      captured: 'onFocusCapture',
    },
  },
  topPress: {
    phasedRegistrationNames: {
      bubbled: 'onPress',
      captured: 'onPressCapture',
    },
  },
  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   */
};

const directEventTypes = {
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */
  topAccessibilityAction: {
    registrationName: 'onAccessibilityAction',
  },
  onGestureHandlerEvent: DynamicallyInjectedByGestureHandler({
    registrationName: 'onGestureHandlerEvent',
  }),
  onGestureHandlerStateChange: DynamicallyInjectedByGestureHandler({
    registrationName: 'onGestureHandlerStateChange',
  }),

  // Direct events from UIManagerModuleConstants.java
  topContentSizeChange: {
    registrationName: 'onContentSizeChange',
  },
  topScrollBeginDrag: {
    registrationName: 'onScrollBeginDrag',
  },
  topMessage: {
    registrationName: 'onMessage',
  },
  topSelectionChange: {
    registrationName: 'onSelectionChange',
  },
  topLoadingFinish: {
    registrationName: 'onLoadingFinish',
  },
  topMomentumScrollEnd: {
    registrationName: 'onMomentumScrollEnd',
  },
  topClick: {
    registrationName: 'onClick',
  },
  topLoadingStart: {
    registrationName: 'onLoadingStart',
  },
  topLoadingError: {
    registrationName: 'onLoadingError',
  },
  topMomentumScrollBegin: {
    registrationName: 'onMomentumScrollBegin',
  },
  topScrollEndDrag: {
    registrationName: 'onScrollEndDrag',
  },
  topScroll: {
    registrationName: 'onScroll',
  },
  topLayout: {
    registrationName: 'onLayout',
  },
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   */
  onAssetDidLoad: {
    registrationName: 'onAssetDidLoad',
  },
  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   * topLongPress does not exist in 0.72, 0.67.3 and 0.64
   */
  topLongPress: {
    registrationName: 'onLongPressEvent',
  },
   /**
   * amznmod_react - Add from BaseViewConfig.android.js
   * topPressIn,topPressOut do not exist in 0.72, 0.67.3 and 0.64
   */
   topPressIn: {
    registrationName: 'onPressIn',
  },
  topPressOut: {
    registrationName: 'onPressOut',
  },
  topAccessibilityEscape: {
    registrationName: 'onAccessibilityEscape',
  },
  topAccessibilityTap: {
    registrationName: 'onAccessibilityTap',
  },
  topMagicTap: {
    registrationName: 'onMagicTap',
  },
  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   */
};

const validAttributesForNonEventProps = {
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */
  // @ReactProps from BaseViewManager
  backgroundColor: {process: require('../StyleSheet/processColor').default},
  transform: true,
  opacity: true,
  elevation: true,
  shadowColor: {process: require('../StyleSheet/processColor').default},
  zIndex: true,
  renderToHardwareTextureKepler: true,
  testID: true,
  nativeID: true,
  accessibilityLabelledBy: true,
  accessibilityLabel: true,
  accessibilityHint: true,
  accessibilityRole: true,
  accessibilityCollection: true,
  accessibilityCollectionItem: true,
  accessibilityState: true,
  // amznmod_react - add transportControls
  transportControls: true,
  accessibilityActions: true,
  accessibilityValue: true,
  importantForAccessibility: true,
  rotation: true,
  scaleX: true,
  scaleY: true,
  translateX: true,
  translateY: true,
  accessibilityLiveRegion: true,

  // @ReactProps from LayoutShadowNode
  width: true,
  minWidth: true,
  collapsable: true,
  maxWidth: true,
  height: true,
  minHeight: true,
  maxHeight: true,
  flex: true,
  flexGrow: true,
  rowGap: true,
  columnGap: true,
  gap: true,
  flexShrink: true,
  flexBasis: true,
  aspectRatio: true,
  flexDirection: true,
  flexWrap: true,
  alignSelf: true,
  alignItems: true,
  alignContent: true,
  justifyContent: true,
  overflow: true,
  display: true,

  margin: true,
  marginBlock: true,
  marginBlockEnd: true,
  marginBlockStart: true,
  marginBottom: true,
  marginEnd: true,
  marginHorizontal: true,
  marginInline: true,
  marginInlineEnd: true,
  marginInlineStart: true,
  marginLeft: true,
  marginRight: true,
  marginStart: true,
  marginTop: true,
  marginVertical: true,

  padding: true,
  paddingBlock: true,
  paddingBlockEnd: true,
  paddingBlockStart: true,
  paddingBottom: true,
  paddingEnd: true,
  paddingHorizontal: true,
  paddingInline: true,
  paddingInlineEnd: true,
  paddingInlineStart: true,
  paddingLeft: true,
  paddingRight: true,
  paddingStart: true,
  paddingTop: true,
  paddingVertical: true,

  borderWidth: true,
  borderStartWidth: true,
  borderEndWidth: true,
  borderTopWidth: true,
  borderBottomWidth: true,
  borderLeftWidth: true,
  borderRightWidth: true,

  start: true,
  end: true,
  left: true,
  right: true,
  top: true,
  bottom: true,

  position: true,

  style: ReactNativeStyleAttributes,
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */


  /**
   * amznmod_react - Add from BaseViewConfig.ios.js
   * Newly added in 0.72
   */
  accessibilityLanguage: true,
  borderCurve: true,
  borderEndEndRadius: true,
  borderEndStartRadius: true,
  borderStartEndRadius: true,
  borderStartStartRadius: true,
  /**
   * amznmod_react - Add from BaseViewConfig.ios.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   */
  hasTVPreferredFocus: true,
  focusable: true,
  tabIndex: true,
  // amznmod_react - Add isFocusableInTouchMode
  isFocusableInTouchMode: true,
  nativeBackgroundAndroid: true,
  nativeForegroundAndroid: true,
  nextFocusDown: true,
  nextFocusForward: true,
  nextFocusLeft: true,
  nextFocusRight: true,
  nextFocusUp: true,
  // amznmod_react - add enableSynchronousFocusEventsKepler
  enableSynchronousFocusEventsKepler: true,
  // amznmod_react - enable TVFocusGuide props/methods
  autoFocus: true,
  trapFocusUp: true,
  trapFocusDown: true,
  trapFocusLeft: true,
  trapFocusRight: true,
  destinations: true, //amznmod_react - Using destinations as prop instead of method
  /**
   * amznmod_react - Add from ReactNativeViewViewConfigAndroid.js
   */


  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   */
  accessibilityElementsHidden: true,
  accessibilityIgnoresInvertColors: true,
  accessibilityStates: true, // TODO: Can be removed after next release
  accessibilityViewIsModal: true,
  // amznmod_react - add accessibilityDescribedBy, accessibilityOrientationText
  accessibilityDescribedBy: true,
  accessibilityDescribedByKepler: true, // TODO: to be deprecated
  accessibilityOrientationText: true,
  accessibilityOrientationTextKepler: true, // TODO: to be deprecated

  // amznmod_react - add accessibilityLabelledBy
  accessibilityLabelledBy: true,
  // amznmod_react Add isPressable
  isPressable: true,
  // amznmod_react Add isLongPressRegistered
  isLongPressRegistered: true,
  // amznmod_react - Add voiceOrdinal
  voiceOrdinal: true,
  // amznmod_react - Add voiceEntityType
  voiceEntityType: true,
  // amznmod_react - Add voiceExternalIds
  voiceExternalIds: true,
  // amznmod_react - Add voiceAlternativeNames
  voiceAlternativeNames: true,
  // amznmod_react - Add ariaPosInset
  ariaPosInSet: true,
  // amznmod_react - Add alexaEntityType
  alexaEntityType: true,
  // amznmod_react - Add keplerAtlLabels (Deprecated, use keplerAltLabels)
  keplerAtlLabels: true,
  // amznmod_react - Add keplerAltLabels
  keplerAltLabels: true,
  // amznmod_react - Add alexaExternalIds
  alexaExternalIds: true,
  // amznmod_react - Add ariaLabel
  ariaLabel: true,
  // amznmod_react - Add texture cache flag on Kepler
  renderToHardwareTextureKepler: true,
  accessible: true,
  backfaceVisibility: true,
  borderBottomColor: {process: require('../StyleSheet/processColor').default},
  borderBottomEndRadius: true,
  borderBottomLeftRadius: true,
  borderBottomRightRadius: true,
  borderBottomStartRadius: true,
  borderColor: {process: require('../StyleSheet/processColor').default},
  borderEndColor: {process: require('../StyleSheet/processColor').default},
  borderLeftColor: {process: require('../StyleSheet/processColor').default},
  borderRadius: true,
  borderRightColor: {process: require('../StyleSheet/processColor').default},
  borderStartColor: {process: require('../StyleSheet/processColor').default},
  borderStyle: true,
  borderTopColor: {process: require('../StyleSheet/processColor').default},
  borderTopEndRadius: true,
  borderTopLeftRadius: true,
  borderTopRightRadius: true,
  borderTopStartRadius: true,
  borderWidth: true,
  clickable: true,
  direction: true,
  hitSlop: {diff: require('../Utilities/differ/insetsDiffer')},
  needsOffscreenAlphaCompositing: true,
  onAccessibilityAction: true, // this shoul be part of validAttributesForEventProps
  onAccessibilityEscape: true, // this shoul be part of validAttributesForEventProps
  onAccessibilityTap: true, // this shoul be part of validAttributesForEventProps
  onLayout: true,
  onMagicTap: true, // this shoul be part of validAttributesForEventProps
  pointerEvents: true,
  removeClippedSubviews: true,
  shadowOffset: {diff: require('../Utilities/differ/sizesDiffer')},
  shadowOpacity: true,
  shadowRadius: true,
  shouldRasterizeIOS: true,
  transform:
    Platform.OS === 'ios'
      ? {diff: require('../Utilities/differ/matricesDiffer')}
      : {process: require('../StyleSheet/processTransform')},
  /**
   * amznmod_react - Add from ReactNativeViewConfig.js
   */
};

/**
 * amznmod_react - Add from BaseViewConfig.android.js
 */
// Props for bubbling and direct events
const validAttributesForEventProps = {
  onLayout: true,

  // PanResponder handlers
  onMoveShouldSetResponder: true,
  onMoveShouldSetResponderCapture: true,
  onStartShouldSetResponder: true,
  onStartShouldSetResponderCapture: true,
  onResponderGrant: true,
  onResponderReject: true,
  onResponderStart: true,
  onResponderEnd: true,
  onResponderRelease: true,
  onResponderMove: true,
  onResponderTerminate: true,
  onResponderTerminationRequest: true,
  onShouldBlockNativeResponder: true,

  // Touch events
  onTouchStart: true,
  onTouchMove: true,
  onTouchEnd: true,
  onTouchCancel: true,

  // Pointer events
  onPointerEnter: true,
  onPointerEnterCapture: true,
  onPointerLeave: true,
  onPointerLeaveCapture: true,
  onPointerMove: true,
  onPointerMoveCapture: true,
  onPointerOut: true,
  onPointerOutCapture: true,
  onPointerOver: true,
  onPointerOverCapture: true,
  /**
   * amznmod_react - Add from BaseViewConfig.android.js
   */


  /**
   * amznmod_react - Add from BaseViewConfig.ios.js
   * Newly added in 0.72
   */
  onPointerUp: true,
  onPointerDown: true,
  onPointerCancel: true,
  /**
   * amznmod_react - Add from BaseViewConfig.ios.js
   */
};

/**
 * On Kepler, Props are derived from a ViewManager and its ShadowNode.
 *
 * Where did we find these base platform props from?
 * - Nearly all component ViewManagers descend from BaseViewManager,
 * - and BaseViewManagers' ShadowNodes descend from LayoutShadowNode.
 * - Also, all components inherit ViewConfigs from UIManagerModuleConstants.java.
 *
 * So, these ViewConfigs are generated from LayoutShadowNode and BaseViewManager.
 */
const PlatformBaseViewConfigKepler: PartialViewConfigWithoutName = {
  directEventTypes,
  bubblingEventTypes,
  validAttributes: {
    ...validAttributesForNonEventProps,
    ...validAttributesForEventProps,
  },
};

export default PlatformBaseViewConfigKepler;
