// @amzn/react-native-w3cmedia/dist/components/mediacontrols/SeekBarUI/ProgressBar/ProgressBarUtilities.js
"use strict";
/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.progressBarTokens = void 0;
const BaseUtilities_1 = require("../BaseUtilities");
const progressBarTokens = (type) => ({
    track: {
        color: {
            backgroundColor: BaseUtilities_1.base[type].color.gray[200]
        },
        size: {
            borderRadius: {
                xs: 8,
                sm: 8,
                md: 10,
                lg: 12,
                xl: 14
            },
            height: {
                xs: 8,
                sm: 8,
                md: 10,
                lg: 12,
                xl: 14
            },
            width: {
                xs: 350,
                sm: 556,
                md: 700,
                lg: 800,
                xl: 1000
            }
        }
    },
    indicator: {
        color: {
            backgroundColor: BaseUtilities_1.base[type].color.gray[600]
        },
        size: {
            borderRadius: {
                xs: 8,
                sm: 8,
                md: 10,
                lg: 12,
                xl: 14
            },
            width: {
                indeterminate: 70
            }
        },
        motion: {
            duration: {
                determinate: 250,
                indeterminate: 1000
            }
        }
    },
    buffer: {
        color: {
            backgroundColor: BaseUtilities_1.base[type].color.gray[400]
        },
        size: {
            borderRadius: {
                xs: 8,
                sm: 8,
                md: 10,
                lg: 12,
                xl: 14
            }
        }
    },
    thumb: {
        color: {
            backgroundColor: {
                enabled: BaseUtilities_1.base[type].color.white,
                disabled: BaseUtilities_1.base[type].color.gray[200]
            }
        },
        size: {
            xs: 12,
            sm: 18,
            md: 24,
            lg: 36,
            xl: 48,
            opacity: {
                disabled: 1
            }
        }
    }
});
exports.progressBarTokens = progressBarTokens;
