// @amazon-devices/react-native-kepler/Libraries/Image/Image.kepler.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @flow
 * @format
 */

import ImageViewNativeComponent from './ImageViewNativeComponent';
import * as React from 'react';
import StyleSheet from '../StyleSheet/StyleSheet';
import TextAncestor from '../Text/TextAncestor';
import ImageInjection from './ImageInjection';
import ImageAnalyticsTagContext from './ImageAnalyticsTagContext';
import flattenStyle from '../StyleSheet/flattenStyle';
import resolveAssetSource from './resolveAssetSource';
import NativeImageLoaderKepler from './NativeImageLoaderKepler';
import TextInlineImageNativeComponent from './TextInlineImageNativeComponent';
import type {ImageProps as ImagePropsType} from './ImageProps';
import type {RootTag} from '../Types/RootTagTypes';
import {convertObjectFitToResizeMode} from './ImageUtils';
import {getImageSourcesFromImageProps} from './ImageSourceUtils';

let _requestId = 1;
function generateRequestId() {
  return _requestId++;
}

/**
 * Retrieve the width and height (in pixels) of an image prior to displaying it
 *
 * See https://reactnative.dev/docs/image.html#getsize
 */
function getSize(
  url: string,
  success: (width: number, height: number) => void,
  failure?: (error: any) => void,
): any {
  return NativeImageLoaderKepler.getSize(url)
    .then(function (sizes) {
      success(sizes.width, sizes.height);
    })
    .catch(
      failure ||
        function () {
          console.warn('Failed to get size for image: ' + url);
        },
    );
}

/**
 * Retrieve the width and height (in pixels) of an image prior to displaying it
 * with the ability to provide the headers for the request
 *
 * See https://reactnative.dev/docs/image.html#getsizewithheaders
 */
function getSizeWithHeaders(
  url: string,
  headers: {[string]: string, ...},
  success: (width: number, height: number) => void,
  failure?: (error: any) => void,
): any {
  return NativeImageLoaderKepler.getSizeWithHeaders(url, headers)
    .then(function (sizes) {
      success(sizes.width, sizes.height);
    })
    .catch(
      failure ||
        function () {
          console.warn('Failed to get size for image: ' + url);
        },
    );
}

function prefetchWithMetadata(
  url: string,
  queryRootName: string,
  rootTag?: ?RootTag,
  callback: ?Function,
): any {
  // TODO: T79192300 Log queryRootName and rootTag
  prefetch(url, callback);
}

function prefetch(url: string, callback: ?Function): any {
  const requestId = generateRequestId();
  callback && callback(requestId);
  return NativeImageLoaderKepler.prefetchImage(url, requestId);
}

function abortPrefetch(requestId: number) {
  NativeImageLoaderKepler.abortRequest(requestId);
}

/**
 * Perform cache interrogation.
 *
 * See https://reactnative.dev/docs/image.html#querycache
 */
async function queryCache(
  urls: Array<string>,
): Promise<{[string]: 'memory' | 'disk' | 'disk/memory', ...}> {
  return await NativeImageLoaderKepler.queryCache(urls);
}

type ImageComponentStatics = $ReadOnly<{|
  getSize: typeof getSize,
  getSizeWithHeaders: typeof getSizeWithHeaders,
  prefetch: typeof prefetch,
  prefetchWithMetadata: typeof prefetchWithMetadata,
  abortPrefetch: typeof abortPrefetch,
  queryCache: typeof queryCache,
  resolveAssetSource: typeof resolveAssetSource,
|}>;

/**
 * A React component for displaying different types of images,
 * including network images, static resources, temporary local images, and
 * images from local disk, such as the camera roll.
 *
 * See https://reactnative.dev/docs/image.html
 */
let Image = (props: ImagePropsType, forwardedRef) => {
  const source = getImageSourcesFromImageProps(props) || {
    uri: undefined,
    width: undefined,
    height: undefined,
  };

  const defaultSource = resolveAssetSource(props.defaultSource);
  const loadingIndicatorSource = resolveAssetSource(props.loadingIndicatorSource);
  let sources;
  let style: ImageStyleProp;
  if (Array.isArray(source)) {
    // $FlowFixMe flattenStyle is not strong enough
    style = flattenStyle([styles.base, props.style]) || {};
    sources = source;
  } else {
    if (source?.uri != null) {
      const {uri} = source;
      const width = props.width ?? source.width;
      const height = props.height ?? source.height;
      // $FlowFixMe flattenStyle is not strong enough
      style = flattenStyle([{width, height}, styles.base, props.style]) || {};
      sources = [source];
    } else {
      // Additional logic to handle URIs being passed as strings which is supported in android but not on iOS.
      style = flattenStyle([styles.base, props.style]);
      sources = [props.source];
    }
  }

  const {onLoadStart, onLoad, onLoadEnd, onError} = props;
  const objectFit =
    style && style.objectFit
      ? convertObjectFitToResizeMode(style.objectFit)
      : null;
  // $FlowFixMe[prop-missing]
  const resizeMode =
    objectFit || props.resizeMode || (style && style.resizeMode) || 'cover';
  const accessible = props.accessible || props.alt !== undefined;
  // amznmod_react: Make all accessible views focusable.
  const focusable = accessible || props.focusable;

  if (props.children) {
    throw new Error(
      'The <Image> component cannot contain children. If you want to render content on top of the image, consider using the <ImageBackground> component or absolute positioning.',
    );
  }

  if (props.defaultSource && props.loadingIndicatorSource) {
    throw new Error(
      'The <Image> component cannot have defaultSource and loadingIndicatorSource at the same time. Please use either defaultSource or loadingIndicatorSource.',
    );
  }

  const nativeProps = {
    ...props,
    style,
    resizeMode: resizeMode,
    focusable: focusable, // amznmod_react: Make all accessible views focusable.
    shouldNotifyLoadEvents: !!(onLoadStart || onLoad || onLoadEnd || onError),
    src: sources,
    source: sources,
    /* $FlowFixMe(>=0.78.0 site=react_native_kepler_fb) This issue was found
     * when making Flow check .kepler.js files. */
    headers: (source?.[0]?.headers || source?.headers: ?{[string]: string}), // amznmod_react: Add support for image request headers.
    defaultSrc: defaultSource ? defaultSource.uri : null,
    loadingIndicatorSrc: loadingIndicatorSource
      ? loadingIndicatorSource.uri
      : null,
    ref: forwardedRef,
    accessible: accessible,
    accessibilityLabel: props.accessibilityLabel ?? props.alt,
  };

  return (
    <ImageAnalyticsTagContext.Consumer>
      {analyticTag => {
        const nativePropsWithAnalytics =
          analyticTag !== null
            ? {
                ...nativeProps,
                internal_analyticTag: analyticTag,
              }
            : nativeProps;
        return (
          <TextAncestor.Consumer>
            {hasTextAncestor =>
              hasTextAncestor ? (
                <TextInlineImageNativeComponent {...nativePropsWithAnalytics} />
              ) : (
                <ImageViewNativeComponent {...nativePropsWithAnalytics} />
              )
            }
          </TextAncestor.Consumer>
        );
      }}
    </ImageAnalyticsTagContext.Consumer>
  );
};

Image = React.forwardRef<
  ImagePropsType,
  | React.ElementRef<typeof TextInlineImageNativeComponent>
  | React.ElementRef<typeof ImageViewNativeComponent>,
>(Image);

if (ImageInjection.unstable_createImageComponent != null) {
  Image = ImageInjection.unstable_createImageComponent(Image);
}

Image.displayName = 'Image';

/**
 * Retrieve the width and height (in pixels) of an image prior to displaying it
 *
 * See https://reactnative.dev/docs/image.html#getsize
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.getSize = getSize;

/**
 * Retrieve the width and height (in pixels) of an image prior to displaying it
 * with the ability to provide the headers for the request
 *
 * See https://reactnative.dev/docs/image.html#getsizewithheaders
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.getSizeWithHeaders = getSizeWithHeaders;

/**
 * Prefetches a remote image for later use by downloading it to the disk
 * cache
 *
 * See https://reactnative.dev/docs/image.html#prefetch
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.prefetch = prefetch;

/**
 * Prefetches a remote image for later use by downloading it to the disk
 * cache, and adds metadata for queryRootName and rootTag.
 *
 * See https://reactnative.dev/docs/image.html#prefetch
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.prefetchWithMetadata = prefetchWithMetadata;

/**
 * Abort prefetch request.
 *
 * See https://reactnative.dev/docs/image.html#abortprefetch
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.abortPrefetch = abortPrefetch;

/**
 * Perform cache interrogation.
 *
 * See https://reactnative.dev/docs/image.html#querycache
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.queryCache = queryCache;

/**
 * Resolves an asset reference into an object.
 *
 * See https://reactnative.dev/docs/image.html#resolveassetsource
 */
/* $FlowFixMe(>=0.89.0 site=react_native_kepler_fb) This comment suppresses an
 * error found when Flow v0.89 was deployed. To see the error, delete this
 * comment and run Flow. */
Image.resolveAssetSource = resolveAssetSource;

/**
 * Switch to `deprecated-react-native-prop-types` for compatibility with future
 * releases. This is deprecated and will be removed in the future.
 */
Image.propTypes = require('deprecated-react-native-prop-types').ImagePropTypes;

const styles = StyleSheet.create({
  base: {
    overflow: 'hidden',
  },
});

module.exports = ((Image: any): ImageAndroid);
