// @amzn/react-native-w3cmedia/dist/interface/index.js
"use strict";
/*
 * Copyright 2022-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerMediaControlHandler = exports.requestMediaKeysSystemAccess = exports.requestMediaKeySystemAccess = exports.MediaSource = exports.QuotaExceededError = exports.InvalidStateError = exports.NotSupportedError = exports.TrackEvent = exports.Event = exports.HTMLAudioElement = exports.HTMLVideoElement = exports.HTMLMediaElement = exports.EndOfStreamError = exports.ReadyState = exports.AppendMode = exports.MediaPlayer = exports.VideoPlayer = exports.AudioPlayer = exports.Video = exports.Audio = void 0;
const EncryptedMediaImpl_1 = require("../EncryptedMediaImpl");
Object.defineProperty(exports, "requestMediaKeysSystemAccess", { enumerable: true, get: function () { return EncryptedMediaImpl_1.requestMediaKeysSystemAccess; } });
Object.defineProperty(exports, "requestMediaKeySystemAccess", { enumerable: true, get: function () { return EncryptedMediaImpl_1.requestMediaKeySystemAccess; } });
const MediaPlayer_1 = require("../MediaPlayer");
Object.defineProperty(exports, "MediaPlayer", { enumerable: true, get: function () { return MediaPlayer_1.MediaPlayer; } });
const Audio_1 = require("./Audio");
Object.defineProperty(exports, "Audio", { enumerable: true, get: function () { return Audio_1.Audio; } });
const AudioPlayer_1 = require("./AudioPlayer");
Object.defineProperty(exports, "AudioPlayer", { enumerable: true, get: function () { return AudioPlayer_1.AudioPlayer; } });
const EncryptedMediaInterface_1 = require("./EncryptedMediaInterface");
Object.defineProperty(exports, "InvalidStateError", { enumerable: true, get: function () { return EncryptedMediaInterface_1.InvalidStateError; } });
Object.defineProperty(exports, "NotSupportedError", { enumerable: true, get: function () { return EncryptedMediaInterface_1.NotSupportedError; } });
Object.defineProperty(exports, "QuotaExceededError", { enumerable: true, get: function () { return EncryptedMediaInterface_1.QuotaExceededError; } });
const EventTargetInterface_1 = require("./EventTargetInterface");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return EventTargetInterface_1.Event; } });
Object.defineProperty(exports, "TrackEvent", { enumerable: true, get: function () { return EventTargetInterface_1.TrackEvent; } });
const HTMLMediaElement_1 = require("./HTMLMediaElement");
Object.defineProperty(exports, "HTMLAudioElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLAudioElement; } });
Object.defineProperty(exports, "HTMLMediaElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLMediaElement; } });
Object.defineProperty(exports, "HTMLVideoElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLVideoElement; } });
const MediaSource_1 = require("./MediaSource");
Object.defineProperty(exports, "MediaSource", { enumerable: true, get: function () { return MediaSource_1.MediaSource; } });
const MediaSourceInterface_1 = require("./MediaSourceInterface");
Object.defineProperty(exports, "AppendMode", { enumerable: true, get: function () { return MediaSourceInterface_1.AppendMode; } });
Object.defineProperty(exports, "EndOfStreamError", { enumerable: true, get: function () { return MediaSourceInterface_1.EndOfStreamError; } });
Object.defineProperty(exports, "ReadyState", { enumerable: true, get: function () { return MediaSourceInterface_1.ReadyState; } });
const Video_1 = require("./Video");
Object.defineProperty(exports, "Video", { enumerable: true, get: function () { return Video_1.Video; } });
const VideoPlayer_1 = require("./VideoPlayer");
Object.defineProperty(exports, "VideoPlayer", { enumerable: true, get: function () { return VideoPlayer_1.VideoPlayer; } });
const KeplerMediaControlHandler_1 = require("./KeplerMediaControlHandler");
Object.defineProperty(exports, "KeplerMediaControlHandler", { enumerable: true, get: function () { return KeplerMediaControlHandler_1.KeplerMediaControlHandler; } });
