// @amzn/react-native-w3cmedia/dist/interface/KeplerMediaControlHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerMediaControlHandler = void 0;
const LogUtils_1 = require("../LogUtils");
const MediaControlHandler_1 = require("../mediacontrolhandler/MediaControlHandler");
class KeplerMediaControlHandler extends MediaControlHandler_1.MediaControlHandler {
    /**
     * @brief Callback from KMC server to execute the media control command "play" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating play command to pipeline as a default mode,
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handlePlay(sessionId) {
        return await super.handlePlay(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "pause" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating pause command to pipeline as a default mode,
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handlePause(sessionId, context) {
        return await super.handlePause(sessionId, context);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "toggleplaypause" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating play/pause based on current state of pipeline as a default mode,
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleTogglePlayPause(sessionId) {
        return await super.handleTogglePlayPause(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "stop" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating pause command to pipeline as a default mode,
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleStop(sessionId) {
        return await super.handleStop(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "startover" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek to start position and initiate the play command to pipeline as a default mode,
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleStartOver(sessionId) {
        return await super.handleStartOver(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "fastforward" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek forward command to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleFastForward(sessionId) {
        return await super.handleFastForward(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "rewind" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek backward command to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleRewind(sessionId) {
        return await super.handleRewind(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "seek" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek forward/backward command to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSeek(position, sessionId) {
        return await super.handleSeek(position, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "setplaybackspeed" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by configure the playbackspeed command to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSetPlaybackSpeed(speed, sessionId) {
        return await super.handleSetPlaybackSpeed(speed, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "skipforward" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek command with delta position to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSkipForward(delta, sessionId) {
        return await super.handleSkipForward(delta, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "skipbackward" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by seek command with delta position to pipeline as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSkipBackward(delta, sessionId) {
        return await super.handleSkipBackward(delta, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "setVolume" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating setvolume command as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSetAudioVolume(volume, sessionId) {
        return await super.handleSetAudioVolume(volume, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "setAudioTrack" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating setaudiotrack command as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     *        Note: Currently this is NOP in default mode and handling will be available in future releases.
     */
    async handleSetAudioTrack(audioTrack, sessionId) {
        return await super.handleSetAudioTrack(audioTrack, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "enableTextTrack" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating enableTextTrack command as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     *        Note: Currently this is NOP in default mode and handling will be available in future releases.
     */
    async handleEnableTextTrack(textTrack, sessionId) {
        return await super.handleEnableTextTrack(textTrack, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "next" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleNext(sessionId) {
        return await super.handleNext(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "previous" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handlePrevious() {
        LogUtils_1.LogUtil.info('KeplerMediaControlHandler: handlePrevious');
        return await super.handlePrevious();
    }
    /**
     * @brief Callback from KMC server to execute the media control command "shuffle" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleEnableShuffle(enable, sessionId) {
        return await super.handleEnableShuffle(enable, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "repeate" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSetRepeatMode(mode, sessionId) {
        return await super.handleSetRepeatMode(mode, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "setRating" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleSetRating(id, rating, sessionId) {
        return await super.handleSetRating(id, rating, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "disableTextTrack" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by initiating disableTextTrack command as a default mode.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     *        Note: Currently this is NOP in default mode and handling will be available in future releases.
     */
    async handleDisableTextTrack(sessionId) {
        return await super.handleDisableTextTrack(sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "getMetaInfo" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleGetMetadataInfo(id) {
        return await super.handleGetMetadataInfo(id);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "customAction" for
     *        current mediaplayer instance.
     *        This is NOP for w3cmedia playback handling as this is playlist management feature.
     *        App clients are requested to choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleCustomAction(action, sessionId) {
        return await super.handleCustomAction(action, sessionId);
    }
    /**
     * @brief Callback from KMC server to execute the media control command "getSessionState" for
     *        current mediaplayer instance.
     *        w3cmedia will handle this internally by updating the sessionstate as a default.
     *        alternatively App client can choose to override the callback function using OverrideMediaControlHandler
     *        and implement accordingly.
     */
    async handleGetSessionState(sessionId) {
        return await super.handleGetSessionState(sessionId);
    }
}
exports.KeplerMediaControlHandler = KeplerMediaControlHandler;
;
