// @amzn/react-native-w3cmedia/dist/interface/KeplerVideoSurfaceView.js
"use strict";
/*
 * Copyright 2021-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerVideoSurfaceView = void 0;
/** This file is copied from ATVKeplerVideoSurface vmpf-dev branch */
const React = __importStar(require("react"));
const LogUtils_1 = require("../LogUtils");
const react_native_1 = require("react-native");
const KeplerVideoSurfaceViewSpec_1 = __importDefault(require("../components/videosurfaceview/KeplerVideoSurfaceViewSpec"));
const styles = react_native_1.StyleSheet.create({
    base: {
        flexGrow: 1,
        flexShrink: 1
    }
});
;
/**
 * @brief This is a react native component that Renders video frames on the screen.
 * Use this component only when operating in pre-buffering mode using the `VideoPlayer`
 * component, where more than one player can be pre-buffered simultaneously with
 * different contents. The app then attaches the surface handle passed by the
 * `onSurfaceViewCreated` event callback to the video player instance that needs
 * to be rendered.
 */
class KeplerVideoSurfaceView extends React.Component {
    constructor(props) {
        super(props);
    }
    /**
     * @hidden
     */
    componentDidMount() {
        LogUtils_1.LogUtil.info('KeplerVideoSurfaceView: componentDidMount');
        if (this.props.onComponentDidMount !== undefined) {
            this.props.onComponentDidMount();
        }
    }
    /**
     * @hidden
     */
    componentWillUnmount() {
        LogUtils_1.LogUtil.info('KeplerVideoSurfaceView: componentWillUnmount');
        if (this.props.onComponentWillUnmount !== undefined) {
            this.props.onComponentWillUnmount();
        }
    }
    /**
     * @hidden
     */
    render() {
        const { scalingmode, onSurfaceViewCreated, onSurfaceViewDestroyed, style } = this.props;
        LogUtils_1.LogUtil.info('render KeplerVideoSurfaceView');
        return (<KeplerVideoSurfaceViewSpec_1.default scalingmode={scalingmode} onKeplerVideoSurfaceViewCreated={(ev) => {
                LogUtils_1.LogUtil.info('KeplerVideoSurfaceView: onKeplerVideoSurfaceViewCreated');
                onSurfaceViewCreated?.(ev.nativeEvent.value);
            }} onKeplerVideoSurfaceViewDestroyed={(ev) => {
                LogUtils_1.LogUtil.info('KeplerVideoSurfaceView: onKeplerVideoSurfaceViewDestroyed');
                onSurfaceViewDestroyed?.(ev.nativeEvent.value);
            }} style={[styles.base, style]}/>);
    }
}
exports.KeplerVideoSurfaceView = KeplerVideoSurfaceView;
;
