// @amzn/react-native-w3cmedia/dist/interface/KeplerVideoView.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerVideoView = void 0;
/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const React = __importStar(require("react"));
const LogUtils_1 = require("../LogUtils");
const KeplerCaptionsView_1 = require("./KeplerCaptionsView");
const KeplerVideoSurfaceView_1 = require("./KeplerVideoSurfaceView");
const MediaControls_1 = __importDefault(require("../components/mediacontrols/MediaControls"));
/**
 * @brief This is a react native component that renders the following as specified
 * by optional props set by the app developer:
 *
 * * Video frames on the screen.
 * * Closed captions and subtitles on the screen (optional).
 * * Media controls (optional).
 *
 */
const KeplerVideoView = (props) => {
    const showKeplerCaptionsView = props.showCaptions;
    const showControls = props.showControls;
    const surfaceViewHandle = React.useRef(null);
    const videoPlayer = props.videoPlayer;
    const scalingmode = props.scalingmode;
    /**
     * @hidden
     */
    const onSurfaceViewCreated = (_surfaceViewHandle) => {
        LogUtils_1.LogUtil.info(`KeplerVideoView: onSurfaceViewCreated ${_surfaceViewHandle}`);
        surfaceViewHandle.current = _surfaceViewHandle;
        if (videoPlayer) {
            videoPlayer.setSurfaceHandle(surfaceViewHandle.current);
            videoPlayer.play();
        }
    };
    /**
     * @hidden
     */
    const onSurfaceViewDestroyed = (_surfaceViewHandle) => {
        LogUtils_1.LogUtil.info(`KeplerVideoView: onSurfaceViewDestroyed: ${_surfaceViewHandle}`);
        if (surfaceViewHandle.current === _surfaceViewHandle) {
            videoPlayer?.clearSurfaceHandle(_surfaceViewHandle);
            surfaceViewHandle.current = null;
        }
    };
    /**
     * @hidden
     */
    const onCaptionViewCreated = (_captionsViewHandle) => {
        LogUtils_1.LogUtil.info(`KeplerVideoView: onCaptionViewCreated ${_captionsViewHandle}`);
        if (videoPlayer) {
            videoPlayer.setCaptionViewHandle(_captionsViewHandle);
        }
    };
    /**
     * @hidden
     */
    const onCaptionViewDestroyed = (_captionsViewHandle) => {
        LogUtils_1.LogUtil.info(`KeplerVideoView: onCaptionViewDestroyed ${_captionsViewHandle}`);
        videoPlayer?.clearCaptionViewHandle(_captionsViewHandle);
    };
    return (<>
            <KeplerVideoSurfaceView_1.KeplerVideoSurfaceView style={{ zIndex: 0 }} scalingmode={scalingmode} onSurfaceViewCreated={onSurfaceViewCreated} onSurfaceViewDestroyed={onSurfaceViewDestroyed}/>
            {showKeplerCaptionsView ?
            <KeplerCaptionsView_1.KeplerCaptionsView onCaptionViewCreated={onCaptionViewCreated} onCaptionViewDestroyed={onCaptionViewDestroyed} show={props.showCaptions} style={{ width: '100%',
                    height: '100%',
                    top: 0,
                    left: 0,
                    position: 'absolute',
                    backgroundColor: "transparent",
                    flexDirection: 'column',
                    alignItems: 'center', zIndex: 2 }}/> : null}
            {showControls ?
            <MediaControls_1.default initFocus={true} mediaPlayer={videoPlayer} style={{ zIndex: 3 }}/> : null}
        </>);
};
exports.KeplerVideoView = KeplerVideoView;
KeplerVideoView.defaultProps = {
    showControls: true,
    showCaptions: true,
    scalingmode: 'fit',
};
