// @amzn/react-native-w3cmedia/dist/interface/MediaSourceInterface.js
"use strict";
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndOfStreamError = exports.ReadyState = exports.AppendMode = void 0;
/**
 * For additional information, see
 * https://www.w3.org/TR/media-source-2/#dom-appendmode.
 */
var AppendMode;
(function (AppendMode) {
    AppendMode["segments"] = "segments";
    AppendMode["sequence"] = "sequence";
})(AppendMode = exports.AppendMode || (exports.AppendMode = {}));
/**
 * For additional information, see
 * https://www.w3.org/TR/media-source-2/#dom-readystate.
 */
var ReadyState;
(function (ReadyState) {
    ReadyState["closed"] = "closed";
    ReadyState["open"] = "open";
    ReadyState["ended"] = "ended";
})(ReadyState = exports.ReadyState || (exports.ReadyState = {}));
/**
 * For additional information, see
 * https://www.w3.org/TR/media-source-2/#dom-endofstreamerror.
 */
var EndOfStreamError;
(function (EndOfStreamError) {
    /**
     * For additional information, see
     * https://www.w3.org/TR/media-source-2/#dom-endofstreamerror-network.
     */
    EndOfStreamError["network"] = "network";
    /**
     * For additional information, see
     * https://www.w3.org/TR/media-source-2/#dom-endofstreamerror-decode.
     */
    EndOfStreamError["decode"] = "decode";
})(EndOfStreamError = exports.EndOfStreamError || (exports.EndOfStreamError = {}));
