// @amzn/react-native-w3cmedia/dist/interface/MediaTypes.js
"use strict";
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioUsageType = exports.AudioContentType = void 0;
// temporary definition of AudioContentType and AudioUsageType till we import from audio turbo module
/**
 *  Describes audio content types.
 */
var AudioContentType;
(function (AudioContentType) {
    AudioContentType[AudioContentType["CONTENT_TYPE_NONE"] = 0] = "CONTENT_TYPE_NONE";
    AudioContentType[AudioContentType["CONTENT_TYPE_SPEECH"] = 1] = "CONTENT_TYPE_SPEECH";
    AudioContentType[AudioContentType["CONTENT_TYPE_MUSIC"] = 2] = "CONTENT_TYPE_MUSIC";
    AudioContentType[AudioContentType["CONTENT_TYPE_MOVIE"] = 3] = "CONTENT_TYPE_MOVIE";
    AudioContentType[AudioContentType["CONTENT_TYPE_SONIFICATION"] = 4] = "CONTENT_TYPE_SONIFICATION";
})(AudioContentType = exports.AudioContentType || (exports.AudioContentType = {}));
;
/**
 * Describes audio usage types.
 */
var AudioUsageType;
(function (AudioUsageType) {
    AudioUsageType[AudioUsageType["USAGE_NONE"] = 0] = "USAGE_NONE";
    AudioUsageType[AudioUsageType["USAGE_MEDIA"] = 1] = "USAGE_MEDIA";
    AudioUsageType[AudioUsageType["USAGE_COMMUNICATION"] = 2] = "USAGE_COMMUNICATION";
    AudioUsageType[AudioUsageType["USAGE_ALARM"] = 3] = "USAGE_ALARM";
    AudioUsageType[AudioUsageType["USAGE_NOTIFICATION"] = 4] = "USAGE_NOTIFICATION";
    AudioUsageType[AudioUsageType["USAGE_ACCESSIBILITY"] = 5] = "USAGE_ACCESSIBILITY";
    AudioUsageType[AudioUsageType["USAGE_GAME"] = 6] = "USAGE_GAME";
    AudioUsageType[AudioUsageType["USAGE_NAVIGATION"] = 7] = "USAGE_NAVIGATION";
    AudioUsageType[AudioUsageType["USAGE_RING"] = 8] = "USAGE_RING";
    AudioUsageType[AudioUsageType["USAGE_VOICE_ASSISTANT"] = 9] = "USAGE_VOICE_ASSISTANT";
})(AudioUsageType = exports.AudioUsageType || (exports.AudioUsageType = {}));
;
