// @amzn/react-native-kepler/Libraries/ReactNative/I18nManager.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Portions of this file are copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * @flow strict-local
 * @format
 */

import NativeI18nManager from './NativeI18nManager';
import type { EventSubscription } from '../vendor/emitter/EventEmitter';

const i18nConstants: {|
  doLeftAndRightSwapInRTL: boolean,
  isRTL: boolean,
  localeIdentifier?: ?string,
|} = getI18nManagerConstants();

function getI18nManagerConstants() {
  if (NativeI18nManager) {
    const {isRTL, doLeftAndRightSwapInRTL, localeIdentifier} =
      NativeI18nManager.getConstants();
    return {isRTL, doLeftAndRightSwapInRTL, localeIdentifier};
  }

  return {
    isRTL: false,
    doLeftAndRightSwapInRTL: true,
  };
}

// amznmod_react - APIs for retrieving and listening for changes on system locale and timezone config
// SettingEventName should match with volta::i18n::I18nManager::Setting
const SettingEventName = {
  Locale: "Locale",
  Timezone: "Timezone"
};

module.exports = {
  SettingEventName,

  getConstants: (): {|
    doLeftAndRightSwapInRTL: boolean,
    isRTL: boolean,
    localeIdentifier: ?string,
  |} => {
    return i18nConstants;
  },

  allowRTL: (shouldAllow: boolean) => {
    if (!NativeI18nManager) {
      return;
    }

    NativeI18nManager.allowRTL(shouldAllow);
  },

  forceRTL: (shouldForce: boolean) => {
    if (!NativeI18nManager) {
      return;
    }

    NativeI18nManager.forceRTL(shouldForce);
  },

  swapLeftAndRightInRTL: (flipStyles: boolean) => {
    if (!NativeI18nManager) {
      return;
    }

    NativeI18nManager.swapLeftAndRightInRTL(flipStyles);
  },

  addListener: (eventName: string, callback: Function) => {
    console.warn("I18nManager::addListener is deprecated. Use addEventListener instead");
    if (!NativeI18nManager || !SettingEventName[eventName]) {
      return;
    }
    NativeI18nManager.addListener(eventName, callback);
  },

  addEventListener: (eventName: string, callback: Function): EventSubscription => {
    if (!NativeI18nManager || !SettingEventName[eventName]) {
      return;
    }
    return NativeI18nManager.addEventListener(eventName, callback);
  },

  removeListeners: () => {
    if (!NativeI18nManager) {
      return;
    }
    NativeI18nManager.removeListeners();
  },

  getAppLocale: () => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.getAppLocale();
  },

  getSystemLocale: () => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.getSystemLocale();
  },

  setAppLocale: (locale: string) => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.setAppLocale(locale);
  },

  getTimezone: () => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.getTimezone();
  },

  getSystemTimezone: () => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.getSystemTimezone();
  },

  setTimezone: (timezone: string) => {
    if (!NativeI18nManager) {
      return;
    }
    return NativeI18nManager.setTimezone(timezone);
  },

  resetLocale: () => {
    if (!NativeI18nManager) {
      return;
    }
    NativeI18nManager.resetLocale();
  },

  resetTimezone: () => {
    if (!NativeI18nManager) {
      return;
    }
    NativeI18nManager.resetTimezone();
  },

  isRTL: i18nConstants.isRTL,
  doLeftAndRightSwapInRTL: i18nConstants.doLeftAndRightSwapInRTL,
};
