// @amazon-devices/react-native-kepler/Libraries/Utilities/BackHandler.kepler.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @flow strict-local
 * @format
 */

import { UserInputManager, UserInputEventName } from '../UserInputManager/UserInputManager';
import NativeDeviceEventManager from "../NativeModules/specs/NativeDeviceEventManager";

type BackPressEventName = 'backPress' | 'hardwareBackPress';

const _backPressSubscriptions = [];
const backPressEventSubscription = UserInputManager.addListener(UserInputEventName.Back, function({ phase }) {
  // Ensure this is the release phase of the key press.
  if (phase !== 'RELEASED') {
    return false;
  }

  for (let i = _backPressSubscriptions.length - 1; i >= 0; i--) {
    if (_backPressSubscriptions[i]()) {
      return true;
    }
  }

  BackHandler.exitApp();
  return false; // unreached
});

// Add backPressEventSubscription to global scope to prevent it from being garbage collected
Object.defineProperty(global, '__backPressEventSubscription', {
  configurable: true,
  value: backPressEventSubscription,
});

/**
 * Detect hardware button presses for back navigation.
 *
 * Android: Detect hardware back button presses, and programmatically invoke the default back button
 * functionality to exit the app if there are no listeners or if none of the listeners return true.
 *
 * iOS: Not applicable.
 *
 * The event subscriptions are called in reverse order (i.e. last registered subscription first),
 * and if one subscription returns true then subscriptions registered earlier will not be called.
 *
 * Example:
 *
 * ```javascript
 * BackHandler.addEventListener('hardwareBackPress', function() {
 *  // this.onMainScreen and this.goBack are just examples, you need to use your own implementation here
 *  // Typically you would use the navigator here to go to the last state.
 *
 *  if (!this.onMainScreen()) {
 *    this.goBack();
 *    return true;
 *  }
 *  return false;
 * });
 * ```
 */
type TBackHandler = {|
  +exitApp: () => void,
  +addEventListener: (
    eventName: BackPressEventName,
    handler: () => ?boolean,
  ) => {remove: () => void, ...},
  +removeEventListener: (
    eventName: BackPressEventName,
    handler: () => ?boolean,
  ) => void,
|};
const BackHandler: TBackHandler = {
  exitApp: function(): void {
    if (!NativeDeviceEventManager) {
      return;
    }

    NativeDeviceEventManager.invokeDefaultBackPressHandler();
  },

  /**
   * Adds an event handler. Supported events:
   *
   * - `hardwareBackPress`: Fires when the Android hardware back button is pressed.
   */
  addEventListener: function(
    eventName: BackPressEventName,
    handler: () => ?boolean,
  ): {remove: () => void, ...} {

    if (_backPressSubscriptions.indexOf(handler) === -1) {
      _backPressSubscriptions.push(handler);
    }
    return {
      remove: (): void => BackHandler.removeEventListener(eventName, handler),
    };
  },

  /**
   * Removes the event handler.
   */
  removeEventListener: function(
    eventName: BackPressEventName,
    handler: () => ?boolean,
  ): void {
    if (_backPressSubscriptions.indexOf(handler) !== -1) {
      _backPressSubscriptions.splice(
        _backPressSubscriptions.indexOf(handler),
        1,
      );
    }
  },
};

module.exports = BackHandler;
