// @amazon-devices/react-native-kepler/Libraries/Utilities/NativeDeviceInfo.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

/**
 * CHANGELOG:
 * 2c3c038b - Use NativeDeviceInfo instead of RCTDeviceEventEmitter to add listener events
 * 36d57fab - Converted Core TM to use Kepler TM interface
 * a482431d - Added attributions to modified react-native files
 * Added Kepler support for multiple windows
 */

// amznmod_react - convert core TM to use Kepler TM interface
import type {KeplerTurboModule} from '../TurboModule/KeplerTurboModule';
import * as TurboModuleRegistry from '../TurboModule/TurboModuleRegistry';
import EmitterSubscription from '../vendor/emitter/EventEmitter';

// amznmod_react - Add Kepler support for multiple windows
export type DisplayMetricsKepler = {|
  width: number,
  height: number,
  scale: number,
  fontScale: number,
  moduleName: string,
|};

type DisplayMetricsAndroid = {|
  width: number,
  height: number,
  scale: number,
  fontScale: number,
  densityDpi: number,
|};

export type DisplayMetrics = {|
  width: number,
  height: number,
  scale: number,
  fontScale: number,
|};

export type DimensionsPayload = {|
  window?: DisplayMetrics,
  screen?: DisplayMetrics,
  windowPhysicalPixels?: DisplayMetricsAndroid,
  screenPhysicalPixels?: DisplayMetricsAndroid,
  // amznmod_react - Add Kepler support for multiple windows
  windows?: DisplayMetricsKepler[],
|};

// amznmod_react - convert core TM to use Kepler TM interface
export interface Spec extends KeplerTurboModule {
  +getConstants: () => {|
    +Dimensions: DimensionsPayload,
    +isIPhoneX_deprecated?: boolean,
  |};
  +getDimensionsV2: (rootTag: number) => {|
    +Dimensions: DisplayMetrics,
  |};
  // amznmod_react Use NativeDeviceInfo instead of RCTDeviceEventEmitter to add listener events
  +addListener: (eventName: string) => void;

  // amznmod_react Use NativeDeviceInfo instead of RCTDeviceEventEmitter to add listener events
  +addListenerV2: (
    rootTag: number,
    eventName: string,
    handler: (event: any) => void,
  ) => {|
    +subscription: EmitterSubscription,
  |};
}

const NativeModule: Spec = TurboModuleRegistry.getEnforcing<Spec>('DeviceInfo');
let constants = null;

const NativeDeviceInfo = {
  getConstants(): {|
    +Dimensions: DimensionsPayload,
    +isIPhoneX_deprecated?: boolean,
  |} {
    if (constants == null) {
      constants = NativeModule.getConstants();
    }
    return constants;
  },
  // amznmod_react Use NativeDeviceInfo instead of RCTDeviceEventEmitter to add listener events
  addListener(eventName, handler): {} {
    NativeModule.addListener(eventName, handler);
  },

  getDimensionsV2(rootTag: number): {|
    +Dimensions: DisplayMetrics,
  |} {
    return NativeModule.getDimensionsV2(rootTag);
  },
  // amznmod_react Use NativeDeviceInfo instead of RCTDeviceEventEmitter to add listener events
  addListenerV2(
    rootTag: number,
    eventName: string,
    callback: (event: any) => void,
  ): {|
    +subscription: EmitterSubscription,
  |} {
    return NativeModule.addListenerV2(rootTag, eventName, callback);
  },
};

export default NativeDeviceInfo;
