// @amzn/react-native-kepler/Libraries/Network/NativeNetworkingAndroid.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

/**
 * CHANGELOG:
 * a482431d - Added attributions to modified react-native files
 */

import type {TurboModule} from '../TurboModule/RCTExport';

import * as TurboModuleRegistry from '../TurboModule/TurboModuleRegistry';

type Header = [string, string];

//amznmod_react -
export interface NativeEventSubscription {
    +remove: () => void;
}

export interface Spec extends TurboModule {
  +sendRequest: (
    method: string,
    url: string,
    requestId: number,
    headers: Array<Header>,
    data: Object,
    responseType: string,
    useIncrementalUpdates: boolean,
    timeout: number,
    withCredentials: boolean,
  ) => void;
  +abortRequest: (requestId: number) => void;
  +clearCookies: (callback: (result: boolean) => void) => void;

  // RCTEventEmitter
  //amznmod_react -
  +addListener: (eventName: string) => NativeEventSubscription;
  +removeListeners: (count: number) => void;
}

export default (TurboModuleRegistry.getEnforcing<Spec>('Networking'): Spec);
