// @amzn/react-native-kepler/Libraries/Network/NativeNetworkingKepler.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @flow
 * @format
 */

import type {KeplerTurboModule} from '../TurboModule/KeplerTurboModule';
import * as TurboModuleRegistry from '../TurboModule/TurboModuleRegistry';

type Header = [string, string];

export interface NativeEventSubscription {
  +remove: () => void;
}

export interface Spec extends KeplerTurboModule {
  +sendRequest: (
      method: string,
      url: string,
      requestId: number,
      headers: Array<Header>,
      data: Object,
      responseType: string,
      useIncrementalUpdates: boolean,
      timeout: number,
      withCredentials: boolean,
  ) => void;
  +abortRequest: (requestId: number) => void;
  +clearCookies: (callback: (result: boolean) => void) => void;
  +setAuthCertificate: (authData: Object) => void;
  // RCTEventEmitter
  +addListener: (eventName: string, callback:Object) => NativeEventSubscription;
  +removeListeners: (count: number) => void;
}

export default (TurboModuleRegistry.getEnforcing<Spec>('Networking'): Spec);
