// @amzn/react-native-kepler/Libraries/Network/RCTNetworking.kepler.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @format
 * @flow
 */

'use strict';

// Do not require the native RCTNetworking module directly! Use this wrapper module instead.
// It will add the necessary requestId, so that you don't have to generate it yourself.
import NativeEventEmitter from '../EventEmitter/NativeEventEmitter';
import NativeNetworkingKepler from './NativeNetworkingKepler';
import type {NativeResponseType} from './XMLHttpRequest';
import convertRequestBody from './convertRequestBody';
import type {RequestBody} from './convertRequestBody';
import KeplerEventSubscription from "./KeplerEventSubscription";

type Header = [string, string];

// Convert FormData headers to arrays, which are easier to consume in
// native on Kepler.
function convertHeadersMapToArray(headers: Object): Array<Header> {
  const headerArray = [];
  for (const name in headers) {
    headerArray.push([name, headers[name]]);
  }
  return headerArray;
}

let _requestId = 1;
function generateRequestId(): number {
  return _requestId++;
}

/**
 * This class is a wrapper around the native RCTNetworking module. It adds a necessary unique
 * requestId to each network request that can be used to abort that request later on.
 */

// CHANGELOG : e326dc898e - Temporarily removing extending NativeEventEmitter until the N/W is ported
// to use NativeEventEmitter#emit.
class RCTNetworking {

  sendRequest(
      method: string,
      trackingName: string,
      url: string,
      headers: Object,
      data: RequestBody,
      responseType: NativeResponseType,
      incrementalUpdates: boolean,
      timeout: number,
      callback: (requestId: number) => mixed,
      withCredentials: boolean,
      requestId?: number,
  ) {
    const body = convertRequestBody(data);
    if (body && body.formData) {
      body.formData = body.formData.map(part => ({
        ...part,
        headers: convertHeadersMapToArray(part.headers),
      }));
    }
    const requestIdToUse = requestId ? requestId : generateRequestId();
    NativeNetworkingKepler.sendRequest(
        method,
        url,
        requestIdToUse,
        convertHeadersMapToArray(headers),
        {...body, trackingName},
        responseType,
        incrementalUpdates,
        timeout,
        withCredentials,
    );
    callback(requestIdToUse);
  }

  abortRequest(requestId: number) {
    NativeNetworkingKepler.abortRequest(requestId);
  }

  clearCookies(callback: (result: boolean) => any) {
    NativeNetworkingKepler.clearCookies(callback);
  }

  setAuthCertificate(authData: Object) {
    NativeNetworkingKepler.setAuthCertificate(authData);
  }
  
  addListener(eventType, listener, context) {
    const nativeEventSubscription = NativeNetworkingKepler.addListener(eventType, listener);

    return new KeplerEventSubscription(eventType, nativeEventSubscription);
  }
}

export default (new RCTNetworking(): RCTNetworking);
