// @amzn/react-native-w3cmedia/dist/components/mediacontrols/FloatingButton.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonType = void 0;
/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
// rs_out
// @ts-nocheck
const react_1 = require("react");
const react_native_1 = require("react-native");
const ic_caption_greyed_png_1 = __importDefault(require("../../../image/ic_caption_greyed.png"));
const ic_caption_off_png_1 = __importDefault(require("../../../image/ic_caption_off.png"));
const ic_caption_on_png_1 = __importDefault(require("../../../image/ic_caption_on.png"));
const ic_disable_loop_png_1 = __importDefault(require("../../../image/ic_disable_loop.png"));
const ic_enable_loop_png_1 = __importDefault(require("../../../image/ic_enable_loop.png"));
const ic_jump_back_10_144dp_png_1 = __importDefault(require("../../../image/ic_jump_back_10_144dp.png"));
const ic_jump_forward_10_144dp_png_1 = __importDefault(require("../../../image/ic_jump_forward_10_144dp.png"));
const ic_pause_144dp_png_1 = __importDefault(require("../../../image/ic_pause_144dp.png"));
const ic_play_144dp_png_1 = __importDefault(require("../../../image/ic_play_144dp.png"));
const ic_volume_mute_png_1 = __importDefault(require("../../../image/ic_volume_mute.png"));
const ic_volume_unmute_png_1 = __importDefault(require("../../../image/ic_volume_unmute.png"));
var ButtonType;
(function (ButtonType) {
    ButtonType["PLAY"] = "Play Button";
    ButtonType["PAUSE"] = "Pause Button";
    ButtonType["SEEK_FORWARD"] = "Forward Button";
    ButtonType["SEEK_BACKWARD"] = "Backward Button";
    ButtonType["MUTE"] = "Mute Button";
    ButtonType["UNMUTE"] = "Unmute Button";
    ButtonType["ENABLE_LOOP"] = "Enable Loop";
    ButtonType["DISABLE_LOOP"] = "Disable Loop";
    ButtonType["PLAYBACK_SPEED"] = "Playback Speed Button";
    ButtonType["CAPTION_ENABLED"] = "Caption Enabled";
    ButtonType["CAPTION_DISABLED"] = "Caption Disabled";
    ButtonType["CAPTION_NOT_AVAILABLE"] = "Caption Not Available";
})(ButtonType = exports.ButtonType || (exports.ButtonType = {}));
const PlayButtonImage = () => (<react_native_1.Image style={styles.playPauseImageContainer} source={ic_play_144dp_png_1.default}/>);
const PauseButtonImage = () => (<react_native_1.Image style={styles.playPauseImageContainer} source={ic_pause_144dp_png_1.default}/>);
const SeekForwardButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_jump_forward_10_144dp_png_1.default}/>);
const SeekBackwardButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_jump_back_10_144dp_png_1.default}/>);
const MuteButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_volume_mute_png_1.default}/>);
const UnMuteButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_volume_unmute_png_1.default}/>);
const EnableLoopButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_enable_loop_png_1.default}/>);
const DisableLoopButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_disable_loop_png_1.default}/>);
const EnableCaptionButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_caption_on_png_1.default}/>);
const DisableCaptionButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_caption_off_png_1.default}/>);
const CaptionNotAvailableButtonImage = () => (<react_native_1.Image style={styles.imageContainer} source={ic_caption_greyed_png_1.default}/>);
const PlaybackSpeedButtonImage = (props) => (<react_native_1.Text style={styles.textButton}>{`${props.value}X`}</react_native_1.Text>);
const DEFAULT_FONT_FAMILY = 'Amazon-Ember-Regular';
const DEFAULT_PLAYBACK_SPEED = '1';
const FloatingButton = (0, react_1.forwardRef)((props, ref) => {
    const { buttonType, onPressHandler, value, hasTVPreferredFocus, ...rest } = props;
    const [color, setColor] = (0, react_1.useState)('');
    const updateColor = (color_) => setColor(color_);
    let button;
    switch (buttonType) {
        case ButtonType.PLAY:
            button = <PlayButtonImage />;
            break;
        case ButtonType.PAUSE:
            button = <PauseButtonImage />;
            break;
        case ButtonType.SEEK_FORWARD:
            button = <SeekForwardButtonImage />;
            break;
        case ButtonType.SEEK_BACKWARD:
            button = <SeekBackwardButtonImage />;
            break;
        case ButtonType.MUTE:
            button = <MuteButtonImage />;
            break;
        case ButtonType.UNMUTE:
            button = <UnMuteButtonImage />;
            break;
        case ButtonType.ENABLE_LOOP:
            button = <EnableLoopButtonImage />;
            break;
        case ButtonType.DISABLE_LOOP:
            button = <DisableLoopButtonImage />;
            break;
        case ButtonType.CAPTION_ENABLED:
            button = <EnableCaptionButtonImage />;
            break;
        case ButtonType.CAPTION_DISABLED:
            button = <DisableCaptionButtonImage />;
            break;
        case ButtonType.CAPTION_NOT_AVAILABLE:
            button = <CaptionNotAvailableButtonImage />;
            break;
        case ButtonType.PLAYBACK_SPEED:
            button = <PlaybackSpeedButtonImage value={typeof value === 'string' ? value : DEFAULT_PLAYBACK_SPEED}/>;
            break;
        default: button = null;
    }
    const isPlayPauseButton = buttonType === ButtonType.PLAY || buttonType === ButtonType.PAUSE;
    return (<react_native_1.TouchableOpacity accessibilityLabel={buttonType} accessibilityRole={"button"} hasTVPreferredFocus={hasTVPreferredFocus} ref={ref} onFocus={() => {
            updateColor(styles.pressedColor.color);
        }} onBlur={() => {
            updateColor('');
        }} onPress={(e) => {
            onPressHandler();
            e.stopPropagation();
        }} onPressIn={() => {
            updateColor(styles.pressedColor.color);
        }} onPressOut={() => {
            updateColor('');
        }} {...rest} activeOpacity={1}>
            <react_native_1.View nativeID={"FloatingButtonContainer"} style={[isPlayPauseButton ? styles.backgroundContainerPlayPause : styles.backgroundContainer,
            value != null && styles.playbackContainer,
            { backgroundColor: color }]}>
                {button}
            </react_native_1.View>
        </react_native_1.TouchableOpacity>);
});
const styles = react_native_1.StyleSheet.create({
    backgroundContainerPlayPause: {
        width: 230,
        height: 230,
        alignItems: 'center',
        justifyContent: 'center',
        alignContent: 'center',
        borderRadius: 115,
        bottom: 30,
        right: 30
    },
    backgroundContainer: {
        width: 80,
        height: 80,
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: 40
    },
    playbackContainer: {
        minWidth: 160,
        width: 'auto',
        overflow: 'hidden'
    },
    imageContainer: {
        width: 60,
        height: 60,
    },
    playPauseImageContainer: {
        width: 170,
        height: 170,
    },
    textButton: {
        fontFamily: DEFAULT_FONT_FAMILY,
        color: 'white',
        fontSize: 50
    },
    pressedColor: {
        color: 'rgba(43, 47, 53, 0.4)'
    }
});
exports.default = FloatingButton;
