// @amzn/react-native-w3cmedia/dist/components/mediacontrols/FocusableButton.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const react_1 = __importStar(require("react"));
const react_native_1 = require("react-native");
const DEFAULT_FONT_FAMILY = 'Amazon-Ember-Regular';
const DEFAULT_STYLE = react_native_1.StyleSheet.create({
    BUTTON: {
        minWidth: 140,
        backgroundColor: '#7F7F7F',
        paddingVertical: 10,
        paddingLeft: 10,
        paddingRight: 50,
        justifyContent: 'center'
    },
    BUTTON_FOCUSED: {
        backgroundColor: '#2B2F35'
    },
    TITLE: {
        color: 'black',
        fontFamily: DEFAULT_FONT_FAMILY,
        fontSize: 20
    },
    TITLE_SELECTED: {
        fontWeight: 'bold',
        fontSize: 23
    }
});
const FocusableButton = (0, react_1.forwardRef)((props, ref) => {
    const [focused, setFocused] = (0, react_1.useState)(false);
    return (<react_native_1.TouchableOpacity {...props} ref={ref} 
    // @ts-ignore
    focusable={true} onPress={() => {
            if (props.onPress) {
                props?.onPress();
            }
        }} onFocus={() => {
            setFocused(true);
        }} onBlur={() => {
            setFocused(false);
        }} hasTVPreferredFocus={props.hasTVPreferredFocus} style={[
            DEFAULT_STYLE.BUTTON,
            focused && DEFAULT_STYLE.BUTTON_FOCUSED
        ]} activeOpacity={1}>
                <react_native_1.Text style={[DEFAULT_STYLE.TITLE, props.selected && DEFAULT_STYLE.TITLE_SELECTED]}>
                    {props.title}
                </react_native_1.Text>
            </react_native_1.TouchableOpacity>);
});
exports.default = FocusableButton;
