// @amzn/react-native-w3cmedia/dist/components/mediacontrols/SliderMetaData.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const react_2 = require("react");
const react_native_1 = require("react-native");
const styles = react_native_1.StyleSheet.create({
    seekContainer: {
        width: '100%',
        bottom: 25,
        position: 'absolute',
        flexDirection: 'row',
        justifyContent: 'space-between',
        paddingLeft: "2%",
        paddingRight: "2%",
    },
    seekValue: {
        padding: 10,
        color: '#fafafa',
        fontSize: 20,
        fontWeight: '500',
        backgroundColor: 'rgba(33, 33, 33, 0.23)',
        borderRadius: 5,
        fontFamily: 'Amazon Ember Display'
    },
    liveValue: {
        padding: 10,
        color: '#fafafa',
        fontSize: 20,
        fontWeight: '500',
        backgroundColor: 'rgba(33, 33, 33, 0.23)',
        borderRadius: 5,
        fontFamily: 'Amazon Ember Display'
    }
});
const formatTime = (milliseconds) => {
    const totalSeconds = Math.floor(milliseconds / 1000);
    const hours = Math.floor(totalSeconds / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60).toString().padStart(2, '0');
    const seconds = (totalSeconds % 60).toString().padStart(2, '0');
    const hoursComponent = hours.toString().padStart(2, '0');
    return hours > 0 ? `${hoursComponent}:${minutes}:${seconds}` : `${minutes}:${seconds}`;
};
const SliderMetaData = (0, react_2.forwardRef)((props) => {
    return (<react_native_1.View style={styles.seekContainer}>
               {props.totalTime === Infinity ? null : <react_native_1.Text style={styles.seekValue}>
                    {formatTime(props.currentTime)}
                </react_native_1.Text>}
                {props.totalTime === Infinity ?
            <react_native_1.Text style={styles.liveValue}>
                    {'\uD83D\uDD34'} LIVE
                </react_native_1.Text> :
            <react_native_1.Text style={styles.seekValue}>
                    {formatTime(props.totalTime)}
                </react_native_1.Text>}
            </react_native_1.View>);
});
exports.default = SliderMetaData;
