// @amzn/react-native-w3cmedia/dist/components/captionsview/KeplerCaptionsViewSpec.js
"use strict";
/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const registerGeneratedViewConfig_1 = __importDefault(require("@amzn/react-native-kepler/Libraries/Utilities/registerGeneratedViewConfig"));
const KeplerCaptionsViewSpec = "KeplerCaptionsViewStable";
const viewConfig = {
    bubblingEventTypes: {
        topOnCaptionViewCreated: {
            phasedRegistrationNames: {
                bubbled: 'onCaptionViewCreated',
                captured: 'onCaptionViewCreatedCapture'
            }
        }
    },
    validAttributes: {
        show: true,
    }
};
(0, registerGeneratedViewConfig_1.default)(KeplerCaptionsViewSpec, {
    uiViewClassName: KeplerCaptionsViewSpec,
    ...viewConfig
});
exports.default = KeplerCaptionsViewSpec;
