// @amzn/react-native-kepler/Libraries/Components/ScrollView/KeplerHorizontalScrollViewNativeComponent.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * 
 * @format
 * @flow strict-local
 */

import type {ScrollViewNativeProps as Props} from './ScrollViewNativeComponentType';
import type {HostComponent} from '../../Renderer/shims/ReactNativeTypes';
import * as NativeComponentRegistry from '../../NativeComponent/NativeComponentRegistry';

const KeplerHorizontalScrollViewNativeComponent: HostComponent<Props> = NativeComponentRegistry.get<Props>(
  'KeplerHorizontalScrollView',
  () => ({
    uiViewClassName: 'KeplerHorizontalScrollView',
    bubblingEventTypes: {},
    directEventTypes: {},
    validAttributes: {
      bounces: true,
      decelerationRate: true,
      disableIntervalMomentum: true,
      endFillColor: {process: require('../../StyleSheet/processColor').default},
      fadingEdgeLength: true,
      horizontal: true,
      inverted: true,
      nestedScrollEnabled: true,
      maintainVisibleContentPosition: true,
      overScrollMode: true,
      pagingEnabled: true,
      persistentScrollbar: true,
      scrollEnabled: true,
      scrollPerfTag: true,
      sendMomentumEvents: true,
      showsHorizontalScrollIndicator: true,
      snapToAlignment: true,
      snapToEnd: true,
      snapToInterval: true,
      snapToStart: true,
      snapToOffsets: true,
      contentOffset: true,
      focusItemAlignment: true,
      scrollSlop: true,
      scrollEventThrottle: true
    },
  }),
);

export default KeplerHorizontalScrollViewNativeComponent;
