// @amazon-devices/react-native-w3cmedia/dist/components/shareablesurfaceview/KeplerShareableSurfaceView.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
/** This file is copied from ATVKeplerVideoSurface vmpf-dev branch */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const registerGeneratedViewConfig_1 = __importDefault(require("@amazon-devices/react-native-kepler/Libraries/Utilities/registerGeneratedViewConfig"));
// This is left commented out because it is needed for codegen. When you run codegen, uncomment this.
// export const Commands: NativeCommands = codegenNativeCommands<NativeCommands>({
//     supportedCommands: ["setValue"],
// });
const KeplerShareableSurfaceViewSpec = "KeplerShareableSurfaceViewStable";
// This is left commented out because it is needed for codegen. When you run codegen, uncomment this.
// export default codegenNativeComponent<NativeProps>("SampleUIComponent");
const viewConfig = {
    validAttributes: {
        scalingmode: true,
    },
    bubblingEventTypes: {
        topOnKeplerShareableSurfaceViewCreated: {
            phasedRegistrationNames: {
                bubbled: 'onKeplerShareableSurfaceViewCreated',
                captured: 'onKeplerShareableSurfaceViewCreatedCapture'
            }
        },
        topOnKeplerShareableSurfaceViewDestroyed: {
            phasedRegistrationNames: {
                bubbled: 'onKeplerShareableSurfaceViewDestroyed',
                captured: 'onKeplerShareableSurfaceViewDestroyedCapture'
            }
        }
    }
};
(0, registerGeneratedViewConfig_1.default)(KeplerShareableSurfaceViewSpec, {
    uiViewClassName: KeplerShareableSurfaceViewSpec,
    ...viewConfig
});
// Comment out this line, when you run codegen.
exports.default = KeplerShareableSurfaceViewSpec;
