// @amazon-devices/kepler-cli-platform/dist/resources/build.js
"use strict";
/*
 * Copyright (c) 2023 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const options_1 = require("../options");
const common_1 = require("../common");
const process = __importStar(require("node:process"));
const verbose = process.argv.includes('--verbose');
common_1.logger.setVerbose(verbose);
const build = {
    name: 'build-kepler',
    description: 'Build the current project',
    text: {
        awaiting: 'Awaiting resources',
        inProgress: 'Building project',
        success: 'Built project',
        failure: 'Failed to build project',
    },
    func: async (ctx, args) => {
        await runBuild(ctx, args);
    },
    dependencies: {},
    options: options_1.buildKeplerOptionsArray,
};
async function runBuild(ctx, args) {
    // Use environment variable override supported by Kepler CLI for consistency
    args.buildType = process.env.BUILDTYPE || args.buildType || 'Release';
    args.verbose = verbose;
    // Create JS bundle and compile it with Hermes, else skip if instructed to
    if (args.skipBundling) {
        common_1.logger.info('Skipping bundle file generation, running native builds');
    }
    else {
        await (0, common_1.buildKeplerBundle)(ctx, args);
    }
    // Run native build
    common_1.logger.info('Initiating native artifacts build for Kepler platform');
    await buildNative(args);
}
async function buildNative(args) {
    const options = common_1.KeplerExecutor.createBuildOptions(args);
    common_1.KeplerExecutor.build(options);
}
exports.default = build;
