// @amazon-devices/kepler-cli-platform/dist/resources/buildBundle.js
"use strict";
/*
 * Copyright (c) 2023 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const options_1 = require("../options");
const common_1 = require("../common");
const process = __importStar(require("node:process"));
const verbose = process.argv.includes('--verbose');
common_1.logger.setVerbose(verbose);
const bundle = {
    name: 'bundle-kepler',
    description: 'Create Metro bundle(s) for application entry points and compile them to Hermes bundle(s)',
    text: {
        awaiting: 'Awaiting resources',
        inProgress: 'Building bundle(s)',
        success: 'Built bundle(s)',
        failure: 'Failed to build bundle(s)',
    },
    func: async (ctx, args) => {
        // Use environment variable override supported by Kepler CLI for consistency
        args.buildType = process.env.BUILDTYPE || args.buildType || 'Release';
        await (0, common_1.buildKeplerBundle)(ctx, args);
    },
    dependencies: {},
    options: options_1.bundleKeplerOptionsArray,
};
exports.default = bundle;
