// @amazon-devices/kepler-cli-platform/dist/resources/hermes.js
"use strict";
/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const options_1 = require("../options");
const common_1 = require("../common");
common_1.logger.setVerbose(process.argv.includes('--verbose'));
// This resource provides the absolute path to the bundle
const hermes = {
    name: 'hermes-kepler',
    description: 'Produces a Hermes bundle given a JS/RAM bundle for the kepler platform',
    text: {
        starting: 'Starting hermes compilation',
        awaiting: 'Awaiting hermes compilation',
        inProgress: 'Compiling bundle with Hermes',
        success: 'Hermes bundle created',
        failure: 'Failed to create Hermes bundle',
    },
    func: async (ctx, args) => common_1.Hermes.createBundle(ctx, {
        out: args.out,
        bundlePath: args.bundlePath,
        hermesFlags: args.hermesFlags,
    }),
    dependencies: {},
    options: options_1.hermesKeplerOptionsArray,
};
exports.default = hermes;
