// @amazon-devices/kepler-cli-platform/dist/resources/run.js
"use strict";
/*
 * Copyright (c) 2023 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cli_tools_1 = require("@react-native-community/cli-tools");
const log_1 = __importDefault(require("../common/log"));
log_1.default.setVerbose(process.argv.includes('--verbose'));
/**
 * The run resource performs the build
 *
 * @remarks
 * This command is a placeholder for future implementation. It is currently unimplemented.
 */
// eslint-disable-next-line @typescript-eslint/ban-types
const run = {
    name: 'run-kepler',
    description: 'Placeholder for later implementation - this command is currently unimplemented',
    text: {
        awaiting: 'Command not implemented',
        inProgress: 'Command not implemented',
        success: 'Command not implemented',
        failure: 'Command not implemented',
    },
    // eslint-disable-next-line @typescript-eslint/no-shadow
    func: async ({ root }, args) => {
        throw new cli_tools_1.CLIError('This command is unimplemented. Please use `npx kepler run-kepler`');
    },
    dependencies: {},
    options: [],
};
exports.default = run;
