// @amazon-devices/react-native-kepler/Libraries/Components/ToastKepler/NativeNotificationManager.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderTemplate = exports.Duration = exports.Importance = void 0;
const TurboModuleRegistry_1 = require("../../TurboModule/TurboModuleRegistry");
/**
 * The following types are taken from NotificationManager turbo module,
 * if you need more context, visit its source code.
 */
var Importance;
(function (Importance) {
    Importance["IMPORTANCE_MIN"] = "MIN";
    Importance["IMPORTANCE_LOW"] = "LOW";
    Importance["IMPORTANCE_NORMAL"] = "NORMAL";
    Importance["IMPORTANCE_HIGH"] = "HIGH";
    Importance["IMPORTANCE_MAX"] = "MAX";
})(Importance = exports.Importance || (exports.Importance = {}));
var Duration;
(function (Duration) {
    Duration[Duration["SHORT"] = 4] = "SHORT";
    Duration[Duration["LONG"] = 8] = "LONG";
})(Duration = exports.Duration || (exports.Duration = {}));
var RenderTemplate;
(function (RenderTemplate) {
    RenderTemplate["SIMPLE_TOAST"] = "SimpleToastRenderTemplate";
    RenderTemplate["MENU_TOAST"] = "MenuToastRenderTemplate";
})(RenderTemplate = exports.RenderTemplate || (exports.RenderTemplate = {}));
exports.default = TurboModuleRegistry_1.getEnforcing('NotificationManager');
