// @amazon-devices/react-native-kepler/Libraries/Components/ToastKepler/ToastKepler.kepler.js
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
import { useCallback, useContext } from 'react';
import { RootTagContext } from "../../ReactNative/RootTag";

import type { 
  LaunchTarget, 
  NotificationContent,
   } from './NativeNotificationManager';
import NativeNotificationManager, {Duration, RenderTemplate, Importance} from './NativeNotificationManager';
import { AppLaunchTarget, ToastKind } from './NotificationManagerUtils';

type IconSize = 'small' | 'normal' | 'large';

/**
 * Interface defining the structure for the ToastKepler component functionalities.
 */
interface IToastKeplerDuration {
  /**
   * Constant representing the short duration for the toast visibility, used as a default.
   */
  SHORT: number;

  /**
   * Constant representing a long duration for the toast visibility, for messages needing longer 
   * user attention.
   */
  LONG: number;
}

/**
 * Standard durations for displaying toast notifications.
 *
 * @property {number} SHORT - Short duration for brief messages.
 * @property {number} LONG - Longer duration for detailed messages.
 */
export const ToastKeplerDuration: IToastKeplerDuration = {
  SHORT: Duration.SHORT,
  LONG: Duration.LONG,
};

export const useToastKepler = () => {
  const rootTag: number = useContext(RootTagContext); // no-op since NotificationManagerTM currently lacks context support.

  /**
   * Displays a toast notification on the screen.
   * @param props Object containing the parameters for showing the toast.
   * @param props.title The title displayed on the toast.
   * @param props.description A brief description of the toast message.
   * @param props.iconUri The URI for the icon to display in the toast.
   * @param props.iconSize Optional. The size category of the icon. Must be one of 'small' (default), 'normal', or 'large'.
   * @param props.duration Optional. The duration of the toast - either 'SHORT' (default) or 'LONG'.
   * @param props.options Optional. Additional options for the toast, including:
   *  - `launchUri`: URI that the toast can reference for additional actions.
   *  - `launchTitle`: The title for any specific action associated with the toast.
   */
  const show = useCallback(({
    title,
    description,
    iconUri,
    options,
    iconSize = 'normal',
    duration = Duration.SHORT,
  }: {
    title: string;
    description: string;
    iconUri: string;
    duration?: number;
    iconSize?: IconSize;
    options?: {
      launchUri: string;
      launchTitle: string;
    };
  }) => {
    if (!NativeNotificationManager) {
      return;
    }
    let launchTargets: LaunchTarget[] = [];
    let template: RenderTemplate = RenderTemplate.SIMPLE_TOAST;

    if (options) {
      launchTargets.push(
        new AppLaunchTarget(options.launchUri, options.launchTitle, 'not_used')
      );
      template = RenderTemplate.MENU_TOAST;
    }

    let iconSizeKey;
    if (iconSize === 'small') {
      iconSizeKey = 'mSmallIconUri';
    } else if (iconSize === 'large') {
      iconSizeKey = 'mLargeIconUri';
    }

    const content: NotificationContent = {
      mTitle: title,
      mDescription: description,
      mIconUri: iconUri,
      ...(iconSizeKey && { [iconSizeKey]: iconUri }),
      mTimeout: duration,
      mLaunchTargets: launchTargets,
      mChannel: {
        importance: {
          importance: Importance.IMPORTANCE_NORMAL,
        },
        kind: new ToastKind(),
        name: template,
      },
    };

    NativeNotificationManager.post(content);
  }, [rootTag]);

  return { show };
}
