// @amazon-devices/react-native-kepler/Libraries/Components/ToastKepler/ToastKepler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useToastKepler = exports.ToastKeplerDuration = void 0;
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const react_1 = require("react");
const RootTag_1 = require("../../ReactNative/RootTag");
const NativeNotificationManager_1 = __importStar(require("./NativeNotificationManager"));
const NotificationManagerUtils_1 = require("./NotificationManagerUtils");
/**
 * Standard durations for displaying toast notifications.
 *
 * @property {number} SHORT - Short duration for brief messages.
 * @property {number} LONG - Longer duration for detailed messages.
 */
exports.ToastKeplerDuration = {
    SHORT: NativeNotificationManager_1.Duration.SHORT,
    LONG: NativeNotificationManager_1.Duration.LONG,
};
const useToastKepler = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext); // no-op since NotificationManagerTM currently lacks context support.
    /**
     * Displays a toast notification on the screen.
     * @param props Object containing the parameters for showing the toast.
     * @param props.title The title displayed on the toast.
     * @param props.description A brief description of the toast message.
     * @param props.iconUri The URI for the icon to display in the toast.
     * @param props.iconSize Optional. The size category of the icon. Must be one of 'small' (default), 'normal', or 'large'.
     * @param props.duration Optional. The duration of the toast - either 'SHORT' (default) or 'LONG'.
     * @param props.options Optional. Additional options for the toast, including:
     *  - `launchUri`: URI that the toast can reference for additional actions.
     *  - `launchTitle`: The title for any specific action associated with the toast.
     */
    const show = react_1.useCallback(({ title, description, iconUri, options, iconSize = 'normal', duration = NativeNotificationManager_1.Duration.SHORT, }) => {
        if (!NativeNotificationManager_1.default) {
            return;
        }
        let launchTargets = [];
        let template = NativeNotificationManager_1.RenderTemplate.SIMPLE_TOAST;
        if (options) {
            launchTargets.push(new NotificationManagerUtils_1.AppLaunchTarget(options.launchUri, options.launchTitle, 'not_used'));
            template = NativeNotificationManager_1.RenderTemplate.MENU_TOAST;
        }
        let iconSizeKey;
        if (iconSize === 'small') {
            iconSizeKey = 'mSmallIconUri';
        }
        else if (iconSize === 'large') {
            iconSizeKey = 'mLargeIconUri';
        }
        const content = {
            mTitle: title,
            mDescription: description,
            mIconUri: iconUri,
            ...(iconSizeKey && { [iconSizeKey]: iconUri }),
            mTimeout: duration,
            mLaunchTargets: launchTargets,
            mChannel: {
                importance: {
                    importance: NativeNotificationManager_1.Importance.IMPORTANCE_NORMAL,
                },
                kind: new NotificationManagerUtils_1.ToastKind(),
                name: template,
            },
        };
        NativeNotificationManager_1.default.post(content);
    }, [rootTag]);
    return { show };
};
exports.useToastKepler = useToastKepler;
