// @amzn/react-native-kepler/Libraries/TV/TVFocusGuideView.js
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Portions of this file are copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Obtained the file from https://github.dev/react-native-tvos/react-native-tvos/tree/tvos-v0.72.4

/**
 * CHANGELOG:
 * Update type to solve syntax error
 * Using destinationProp as prop instead of method
 */
const React = require('react');
const ReactNative = require('react-native');

import {Commands} from '../Components/View/ViewNativeComponent';
import type {ViewProps} from '../Components/View/ViewPropTypes';
import type {HostComponent} from '../Renderer/shims/ReactNativeTypes';
import setAndForwardRef from '../Utilities/setAndForwardRef';

const {View} = ReactNative;

type TVFocusGuideViewProps = $ReadOnly<{
  ...ViewProps,

  /**
   * If the view should be "visible". display "flex" if visible, otherwise "none".
   * Useful for absolute focus guides without children
   * Defaults to true
   */
  enabled?: boolean,

  /**
   * The views the focus should go to
   */
  destinations?: (?React.ElementRef<HostComponent<mixed>>)[],

  /**
   * @deprecated Don't use it, no longer necessary.
   */
  safePadding?: 'vertical' | 'horizontal' | 'both' | null,

  autoFocus?: boolean,

  trapFocusUp?: boolean,
  trapFocusDown?: boolean,
  trapFocusLeft?: boolean,
  trapFocusRight?: boolean,
}>;

export type TVFocusGuideViewImperativeMethods = $ReadOnly<{
  setDestinations: (
    destinations: (?React.ElementRef<HostComponent<mixed>>)[],
  ) => void,
}>;

function TVFocusGuideView(
  {
    enabled = true,
    safePadding,
    destinations: destinationsProp,
    ...props
  }: TVFocusGuideViewProps,
  forwardedRef,
): React.Node {
  // amznmod_react: Update type to solve syntax error
  const focusGuideRef = React.useRef(null);

  const setDestinations = React.useCallback(
    (destinations: (?React.ElementRef<HostComponent<mixed>>)[]) => {
      const dests: number[] = (destinations || [])
        .map(ReactNative.findNodeHandle)
        .filter(Boolean);

      if (focusGuideRef.current != null) {
        Commands.setDestinations(focusGuideRef.current, dests);
      }
    },
    [],
  );

  const _setNativeRef = setAndForwardRef({
    getForwardedRef: () => forwardedRef,
    setLocalRef: ref => {
      focusGuideRef.current = ref;

      // This is a hack. Ideally we would forwardRef to the underlying
      // host component. However, since TVFocusGuide has its own methods that can be
      // called as well, if we used the standard forwardRef then these
      // methods wouldn't be accessible
      //
      // Here we mutate the ref, so that the user can use the standard native
      // methods like `focus()`, `blur()`, etc. while also having access to
      // imperative methods of this component like `setDestinations()`.
      if (ref) {
        ref.setDestinations = setDestinations;
      }
    },
  });

  // amznmod_rect - Using destinationProp as prop instead of method as PROJ*-125820-UseEffect is 
  // called before mount resulting in method call as no op. 
  // React.useEffect(() => {
  //   if (destinationsProp != null) {
  //     setDestinations(destinationsProp);
  //   }
  // }, [setDestinations, destinationsProp]);
  const dests: number[] = (destinationsProp || [])
    .map(ReactNative.findNodeHandle)
    .filter(Boolean);

  const enabledStyle = {display: enabled ? 'flex' : 'none'};
  const style = [styles.container, props.style, enabledStyle];

  return (
    // $FlowFixMe[prop-missing]
    <ReactNative.View
      {...props}
      destinations={dests} //amznmod_react - Using destinationProp as prop instead of method
      style={style}
      ref={_setNativeRef}
      collapsable={false}
    />
  );
}

const styles = ReactNative.StyleSheet.create({
  container: {
    minWidth: 1,
    minHeight: 1,
  },
});

const ForwardedTVFocusGuideView: React.AbstractComponent<
  TVFocusGuideViewProps,
  React.ElementRef<typeof View> & TVFocusGuideViewImperativeMethods,
> = React.forwardRef(TVFocusGuideView);
ForwardedTVFocusGuideView.displayName = 'TVFocusGuideView';

module.exports = ForwardedTVFocusGuideView;
