// @amzn/react-native-kepler/Libraries/TV/TVTextScrollView.js
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Portions of this file are copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Obtained the file from https://github.dev/react-native-tvos/react-native-tvos/tree/tvos-v0.72.4

/**
 * CHANGELOG:
 * Modified paths to align with kepler
 */

const React = require('react');
//amznmod_react - Modified paths to align with kepler
const ScrollView = require('../Components/ScrollView/ScrollView');
const TVEventHandler = require('./TVEventHandler');
//amznmod_react - Modified paths to align with kepler
const findNodeHandle =
  require('../Renderer/shims/ReactNative').findNodeHandle;

//amznmod_react - Modified paths to align with kepler
import typeof Props from '../Components/ScrollView/ScrollView';

/**
 * Convenience wrapper to create a scroll view that will scroll correctly
 * using swipe gestures on tvOS, even if the scroll view has no focusable
 * subviews.
 *
 * The main use case would be when a large scrolling block of text needs
 * to be presented to the user.
 *
 * Props:
 *
 */

class TVTextScrollView extends React.Component<{
  ...Props,
  /**
   * The duration of the scroll animation when a swipe is detected.
   * Default value is 0.3 s
   */
  scrollDuration?: number,
  /**
   * Scrolling distance when a swipe is detected
   * Default value is half the visible height (vertical scroller)
   * or width (horizontal scroller)
   */
  pageSize?: number,
  /**
   * If true, will scroll to start when focus moves out past the beginning
   * of the scroller
   * Defaults to true
   */
  snapToStart?: boolean,
  /**
   * If true, will scroll to end when focus moves out past the end of the
   * scroller
   * Defaults to true
   */
  snapToEnd?: boolean,
  /**
   * Called when the scroller comes into focus (e.g. for highlighting)
   */
  onFocus?: (evt: Event) => void,
  /**
   * Called when the scroller goes out of focus
   */
  onBlur?: (evt: Event) => void,
}> {
  _tvEventHandler: ?TVEventHandler;

  componentDidMount() {
    this._tvEventHandler = new TVEventHandler();
    this._tvEventHandler.enable(this, function (cmp, evt) {
      const myTag = findNodeHandle(cmp);
      evt.dispatchConfig = {};
      if (myTag === evt.tag) {
        if (evt.eventType === 'focus') {
          cmp.props.onFocus && cmp.props.onFocus(evt);
        } else if (evt.eventType === 'blur') {
          cmp.props.onBlur && cmp.props.onBlur(evt);
        }
      }
    });
  }

  componentWillUnmount() {
    if (this._tvEventHandler) {
      this._tvEventHandler.disable();
      delete this._tvEventHandler;
    }
  }

  render(): React.Node | React.Element<string> {
    const props: $FlowFixMe = {
      ...this.props,
      tvParallaxProperties: {
        pressDuration: this.props.scrollDuration || 0.0,
      },
      isTVSelectable: true,
      snapToInterval: this.props.pageSize || 0.0,
      removeClippedSubviews: false,
      automaticallyAdjustContentInsets: false,
    };
    return <ScrollView {...props}>{this.props.children}</ScrollView>;
  }
}

module.exports = TVTextScrollView;
