// @amazon-devices/keplerscript-commonmodules/bin/modifyAllowListScript.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { env } = require('node:process');

// Utility to safely read a JSON file
function readJsonSafe(filePath) {
  if (!fs.existsSync(filePath)) {
    console.error(`File path ${filePath} not found. Please ensure that modify-allow-list.json is defined in the same directory as app package.json while calling this script.`);
    process.exit(1);
  }
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  } catch (e) {
    console.error(`Failed to parse JSON at ${filePath}`);
    process.exit(1);
  }
}

function writeContentToFile(modifyPath, allowListPath) {
  // Read both files (or empty objects)
  const modifyData = readJsonSafe(modifyPath);
  const currentAllowList = readJsonSafe(allowListPath);
  const merged = deepMerge(currentAllowList, modifyData);

  // Write merged data back into allow-list.json file
  fs.writeFileSync(allowListPath, JSON.stringify(merged, null, 2), 'utf-8');
  console.log(`Merged from ${modifyPath} into ${allowListPath}`);  
}

// Deep merge function with validation for required structure
function deepMerge(target, source) {
  // Validate the structure of target and source
  if (typeof target !== 'object' || typeof source !== 'object') {
    console.error('Both target and source must be objects');
    process.exit(1);
  }

  // Iterate through each key in the source object
  for (const key in source) {
    if (source.hasOwnProperty(key)) {
      // Ensure both target and source values are objects with string keys and string values
      if (typeof target[key] === 'object' && typeof source[key] === 'object') {
        // Validate the structure of the nested object
        if (isValidObject(source[key]) && isValidObject(target[key])) {
          // Recursively merge
          target[key] = deepMerge(target[key], source[key]);
        }
      } else if (typeof target[key] === 'undefined') {
        // If the key doesn't exist in the target, simply assign the value
        target[key] = source[key];
      } else {
        // Directly overwrite the value in target
        target[key] = source[key];
      }
    }
  }
  return target;
}

// Helper function to check if an object has string keys and string values
function isValidObject(obj) {
  const value = Object.keys(obj).every(key => typeof key === 'string' && typeof obj[key] === 'boolean');
  if (value) {
    return true;
  } else {
    console.error('Invalid format for modify-allow-list.json file. Expected an object with string-boolean key-value pairs.');
    process.exit(1);
  }
}

function main() {
  console.log("Running modify-allow-list script. This script will add/remove libraries with their major version to the existing allow-list present in keplerscript-commonmodules");
  // This should be the path to the app's package.json file as this script is meant to be called from postinstall in app's package.json file
  const targetPath = process.cwd();

  const modifyPath = path.resolve(targetPath, 'modify-allow-list.json');
  const allowListPath = path.resolve(targetPath, 'node_modules', '@amzn', 'js-bundle-system-allow-list', 'allow-list.json');

  writeContentToFile(modifyPath, allowListPath);
}

// Only run main if this script is run directly (not required)
if (require.main === module) {
  main();
}

if (env.NODE_ENV === 'test') {
  module.exports = {
    readJsonSafe,
    writeContentToFile,
    deepMerge,
    isValidObject,
    main,
  };
}
