// @amazon-devices/react-native-w3cmedia/dist/components/videosurfaceview/KeplerVideoSurfaceViewSpec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const registerGeneratedViewConfig_1 = __importDefault(require("@amazon-devices/react-native-kepler/Libraries/Utilities/registerGeneratedViewConfig"));
// This is left commented out because it is needed for codegen. When you run codegen, uncomment this.
// export const Commands: NativeCommands = codegenNativeCommands<NativeCommands>({
//     supportedCommands: ["setValue"],
// });
const KeplerVideoSurfaceViewSpec = "KeplerVideoSurfaceViewStable";
// This is left commented out because it is needed for codegen. When you run codegen, uncomment this.
// export default codegenNativeComponent<NativeProps>("SampleUIComponent");
const viewConfig = {
    validAttributes: {
        scalingmode: true,
    },
    bubblingEventTypes: {
        topOnKeplerVideoSurfaceViewCreated: {
            phasedRegistrationNames: {
                bubbled: 'onKeplerVideoSurfaceViewCreated',
                captured: 'onKeplerVideoSurfaceViewCreatedCapture'
            }
        },
        topOnKeplerVideoSurfaceViewDestroyed: {
            phasedRegistrationNames: {
                bubbled: 'onKeplerVideoSurfaceViewDestroyed',
                captured: 'onKeplerVideoSurfaceViewDestroyedCapture'
            }
        }
    }
};
(0, registerGeneratedViewConfig_1.default)(KeplerVideoSurfaceViewSpec, {
    uiViewClassName: KeplerVideoSurfaceViewSpec,
    ...viewConfig
});
// Comment out this line, when you run codegen.
exports.default = KeplerVideoSurfaceViewSpec;
