// @amazon-devices/keplerscript-turbomodule-api/dist/kepler-codegen/RNCodegenUtils.js
"use strict";
/**
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * PROPRIETARY/CONFIDENTIAL
 *
 * Use is subject to license terms.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeScriptParser = exports.unwrapNullable = exports.getModules = exports.createAliasResolver = exports.TypeAnnotation = exports.indent = void 0;
/**
 * This file serves as the interface between RNCodegen and Kepler Codegen.
 * TS types provided in this file are based on Flow types used by RN Codegen.
 */
// @ts-expect-error RN Codegen provides no TS types for these
const Utils_1 = require("../../rn-codegen/lib/generators/Utils");
exports.indent = Utils_1.indent;
var TypeAnnotation;
(function (TypeAnnotation) {
    TypeAnnotation["RESERVED"] = "ReservedTypeAnnotation";
    TypeAnnotation["VOID"] = "VoidTypeAnnotation";
    TypeAnnotation["STRING"] = "StringTypeAnnotation";
    TypeAnnotation["NUMBER"] = "NumberTypeAnnotation";
    TypeAnnotation["DOUBLE"] = "DoubleTypeAnnotation";
    TypeAnnotation["FLOAT"] = "FloatTypeAnnotation";
    TypeAnnotation["INT32"] = "Int32TypeAnnotation";
    TypeAnnotation["INT64"] = "Int64TypeAnnotation";
    TypeAnnotation["UINT32"] = "UInt32TypeAnnotation";
    TypeAnnotation["BIGINT"] = "BigIntTypeAnnotation";
    TypeAnnotation["BOOLEAN"] = "BooleanTypeAnnotation";
    TypeAnnotation["ENUM"] = "EnumDeclaration";
    TypeAnnotation["GENERIC_OBJECT"] = "GenericObjectTypeAnnotation";
    TypeAnnotation["NATIVE_OBJECT"] = "NativeObjectTypeAnnotation";
    TypeAnnotation["UNION"] = "UnionTypeAnnotation";
    TypeAnnotation["OBJECT"] = "ObjectTypeAnnotation";
    TypeAnnotation["ARRAY"] = "ArrayTypeAnnotation";
    TypeAnnotation["FUNCTION"] = "FunctionTypeAnnotation";
    TypeAnnotation["PROMISE"] = "PromiseTypeAnnotation";
    TypeAnnotation["ARRAY_BUFFER"] = "ArrayBufferTypeAnnotation";
    TypeAnnotation["JSON_CONTAINER"] = "JSONContainerTypeAnnotation";
    TypeAnnotation["MIXED"] = "MixedTypeAnnotation";
})(TypeAnnotation = exports.TypeAnnotation || (exports.TypeAnnotation = {}));
const Utils_2 = require("../../rn-codegen/lib/generators/modules/Utils");
exports.createAliasResolver = Utils_2.createAliasResolver;
exports.getModules = Utils_2.getModules;
// @ts-expect-error RN Codegen provides no TS types for these
const parsers_commons_1 = require("../../rn-codegen/lib/parsers/parsers-commons");
exports.unwrapNullable = parsers_commons_1.unwrapNullable;
var parser_1 = require("../../rn-codegen/lib/parsers/typescript/parser");
Object.defineProperty(exports, "TypeScriptParser", { enumerable: true, get: function () { return parser_1.TypeScriptParser; } });
