// @amzn/react-native-kepler/Libraries/Components/AccessibilityInfo/NativeAccessibilityInfo.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

import type {TurboModule} from '../../TurboModule/RCTExport';

import * as TurboModuleRegistry from '../../TurboModule/TurboModuleRegistry';

// amznmod_react - Added native subscription matching RCTEventEmitter behaviour
export interface NativeEventSubscription {
  +remove: () => void;
}

// amznmod_react - Converted Accessibilityinfo to use Kepler TM
export interface Spec extends KeplerTurboModule {
  +isReduceMotionEnabled: (
    onSuccess: (isReduceMotionEnabled: boolean) => void,
  ) => void;
  +isTouchExplorationEnabled: (
    onSuccess: (isScreenReaderEnabled: boolean) => void,
  ) => void;
  +isAccessibilityServiceEnabled?: ?(
    onSuccess: (isAccessibilityServiceEnabled: boolean) => void,
  ) => void;
  +setAccessibilityFocus: (reactTag: number) => void;
  +announceForAccessibility: (announcement: string) => void;
  +announceForAccessibilityWithOptions: (announcement: string, polite: boolean) => void;
  // RCTEventEmitter
  +addListener: (
    eventName: string,
    callback: Function,
  ) => NativeEventSubscription;
  +getRecommendedTimeoutMillis?: (
    mSec: number,
    onSuccess: (recommendedTimeoutMillis: number) => void,
  ) => void;
}

export default (TurboModuleRegistry.get<Spec>('AccessibilityInfo'): ?Spec);
