// @amazon-devices/react-native-kepler/Libraries/Storage/NativeAsyncLocalStorage.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

/**
 * CHANGELOG:
 * 5ca035ce - Converted core TM to use the Kepler TM interface
 * a482431d - Added attributions to modified react-native files
 */

// amznmod_react - Ensure that TM uses Kepler TM interface
import type {KeplerTurboModule} from '../TurboModule/KeplerTurboModule';
import * as TurboModuleRegistry from '../TurboModule/TurboModuleRegistry';

// amznmod_react - Ensure that TM uses Kepler TM interface
export interface Spec extends KeplerTurboModule {
  +getConstants: () => {};
  +multiGet: (
    keys: Array<string>,
    callback: (
      errors: ?Array<{message: string, key?: string}>,
      kvPairs: ?Array<Array<string>>,
    ) => void,
  ) => void;
  +multiSet: (
    kvPairs: Array<Array<string>>,
    callback: (errors: ?Array<{message: string, key?: string}>) => void,
  ) => void;
  +multiMerge: (
    kvPairs: Array<Array<string>>,
    callback: (errors: ?Array<{message: string, key?: string}>) => void,
  ) => void;
  +multiRemove: (
    keys: Array<string>,
    callback: (errors: ?Array<{message: string, key?: string}>) => void,
  ) => void;
  // amznmod_react - Add removeValuesForKeysWithPrefix
  +removeValuesForKeysWithPrefix: (
      prefix: string,
      callback: (errors: ?Array<{message: string, key?: string}>) => void,
  ) => void;
  +clear: (callback: (error: {message: string, key?: string}) => void) => void;
  +getAllKeys: (
    callback: (
      error: ?{message: string, key?: string},
      allKeys: ?Array<string>,
    ) => void,
  ) => void;
}

export default (TurboModuleRegistry.get<Spec>('AsyncLocalStorage'): ?Spec);
