// @amzn/react-native-kepler/Libraries/Utilities/Platform.kepler.js
/**
 * Copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @format
 * @flow strict
 */

import NativePlatformConstantsKepler from './NativePlatformConstantsKepler';

export type PlatformSelectSpec<T> = {
  android?: T,
  native?: T,
  default?: T,
  kepler?: T,
  ...
};

const Platform = {
  __constants: null,
  OS: 'kepler',
  // $FlowFixMe[unsafe-getters-setters]
  get Version(): number {
    return this.constants.Version;
  },
  // $FlowFixMe[unsafe-getters-setters]
  get constants(): {|
    isTesting: boolean,
    keplerOSVariant: string,
    reactNativeVersion: {|
      major: number,
      minor: number,
      patch: number,
      prerelease: ?number,
    |},
    Version: number,
    Release: string,
    Serial: string,
    Fingerprint: string,
    Model: string,
    ServerHost?: string,
    uiMode: string,
    Brand: string,
    Manufacturer: string,
  |} {
    if (this.__constants == null) {
      this.__constants = NativePlatformConstantsKepler.getConstants();
    }
    return this.__constants;
  },
  // $FlowFixMe[unsafe-getters-setters]
  get isTesting(): boolean {
    if (__DEV__) {
      return this.constants.isTesting;
    }
    return false;
  },
  // $FlowFixMe[unsafe-getters-setters]
  get isTV(): boolean {
    return this.constants.uiMode === 'tv';
  },
  select: <T>(spec: PlatformSelectSpec<T>): T =>
    'kepler' in spec
      ? // $FlowFixMe[incompatible-return]
        spec.kepler
      : 'native' in spec
        ? // $FlowFixMe[incompatible-return]
          spec.native
        : // $FlowFixMe[incompatible-return]
          spec.default,
};

module.exports = Platform;
