// @amzn/react-native-kepler/Libraries/RestorationManager/RestorationManager.js
/*
* Copyright 2024-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
*
* AMAZON PROPRIETARY/CONFIDENTIAL
*
* You may not use this file except in compliance with the terms and
* conditions set forth in the accompanying LICENSE.TXT file.
*
* THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
* DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
* IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
*/
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRestorationManager = void 0;
const RestorationManagerTurboModuleSpec_1 = __importDefault(require("./RestorationManagerTurboModuleSpec"));
const react_1 = require("react");
const RootTag_1 = require("../ReactNative/RootTag");
/**
 * RestorationManager allows app developers to save arbitrary state data (scroll position, navigation position, etc.)
 * and retrieve this data after the app was killed and restored by the system (usually because it was in the background
 * and the system needed more resources for the foreground app).
 */
class RestorationManager {
    constructor(rootTag) {
        this.rootTag = rootTag;
    }
    /**
     * Set a key-value pair. If the key already has an associated value, it will be overwritten.
     * @param key A string key. A valid key must meet the following requirements:
     * - It should be between 2 and 64 characters in length.
     * - It should be created using only these allowed characters: `[a-zA-Z0-9-._]`
     * @param value The value to set. A valid value is a UTF-8 encoded string with a maximum size of 1KB.
     */
    async setString(key, value) {
        return RestorationManagerTurboModuleSpec_1.default.setString(this.rootTag, key, value);
    }
    /**
     *
     * @param key A string key.
     */
    async getString(key) {
        return RestorationManagerTurboModuleSpec_1.default.getString(this.rootTag, key);
    }
    /**
     * Check if the given key exists.
     * @param key A string key.
     */
    async exists(key) {
        return RestorationManagerTurboModuleSpec_1.default.exists(this.rootTag, key);
    }
    /**
     * Erases the value associated with the given key. Does nothing if the key was not associated with any value.
     * @param key A string key.
     */
    async erase(key) {
        return RestorationManagerTurboModuleSpec_1.default.erase(this.rootTag, key);
    }
    /**
     * Check if we have state data available to restore. Note this will return true after calling any of the following
     * methods: setStringValue, getStringValue, exists, erase. This will return false only the following conditions are
     * met:
     * - None of these APIs have been called yet: setStringValue, getStringValue, exists, erase
     * - The app launch was a fresh start (not killed by the system and then restored).
     */
    async restorableStateExists() {
        return RestorationManagerTurboModuleSpec_1.default.restorableStateExists(this.rootTag);
    }
    /**
     * Check if the app was restored by the system. If true, that means any previously saved state should be available.
     */
    async isRestored() {
        return RestorationManagerTurboModuleSpec_1.default.isRestored(this.rootTag);
    }
}
/**
 * A custom hook that returns an instance of IRestorationManager.
 */
const useRestorationManager = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    return react_1.useMemo(() => new RestorationManager(rootTag), [rootTag]);
};
exports.useRestorationManager = useRestorationManager;
