// @amazon-devices/react-native-w3cmedia/dist/interface/AudioPlayer.js
"use strict";
/*
 * Copyright 2022-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioPlayer = void 0;
const MediaPlayer_1 = require("../MediaPlayer");
const LogUtils_1 = require("../LogUtils");
/** *
* @brief `AudioPlayer` is a component that implements the `HTMLAudioElement` interface
 * that extends `HTMLMediaElement` interface.
 * @remarks This is a typescript class and not a react native component. Apps
 * need to create an instance of `AudioPlayer` and use it to start buffering
 * the content without starting playback. It does not render the media controls GUI.
 * Apps are expected to build their own media controls UI and control the playback experience.
 *
 */
class AudioPlayer extends MediaPlayer_1.MediaPlayer {
    /**
     * @brief Creates a `AudioPlayer` instance.
     * @param audioType Type of audio to play.
     * @param audioUsage  Reason for audio playback.
     */
    constructor(audioType, audioUsage) {
        LogUtils_1.LogUtil.info("AudioPlayer: +constructor");
        super("audio", audioType, audioUsage);
    }
    /**
     * @brief Initiailzes the player.
     * Wait for the promise to be fulfilled before calling any other function.
     * @returns Promise that is resolved when the player is initialized.
     */
    initialize() {
        return super.initialize();
    }
    /**
     * @brief DeInitiailzes the player.
     *
     * @returns Promise that is resolved when the player is deinitialized.
     */
    deinitialize() {
        return super.deinitialize();
    }
    /**
     * @brief DeInitiailzes the player synchronously
     * @param timeoutInMs time out threshold set by client to return the deinit control.
     * @returns MediaPlayerDeInitStatus
     */
    deinitializeSync(timeoutInMs) {
        return super.deinitializeSync(timeoutInMs);
    }
    /**
     * @brief setup the component instance with current mediaplayer being used and pass mediaControlHandler
     *        in case client wants to opt handling of mediacontrol commands.
     */
    setMediaControlFocus(componentInstance, mediaControlHandler) {
        return super.setMediaControlFocus(componentInstance, mediaControlHandler);
    }
}
exports.AudioPlayer = AudioPlayer;
;
