// @amazon-devices/react-native-w3cmedia/dist/interface/Audio.js
"use strict";
/*
 * Copyright 2022-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Audio = void 0;
const React = __importStar(require("react"));
const LogUtils_1 = require("../LogUtils");
const react_native_1 = require("react-native");
const KeplerCaptionsView_1 = require("./KeplerCaptionsView");
const MediaControls_1 = __importDefault(require("../components/mediacontrols/MediaControls"));
const Media_1 = require("../Media");
const AudioPlayer_1 = require("./AudioPlayer");
/**
 * A react native component that Implements the `HTMLAudioElement` interface that extends
 * `HTMLMediaElement` interface..
 */
class Audio extends Media_1.Media {
    isMediaPlayerInitialized = false;
    constructor(props) {
        super(props, new AudioPlayer_1.AudioPlayer(props.audioType, props.audioUsage));
        if (this.props.mediaControlFocus !== undefined) {
            super.setMediaControlFocus(this.props.mediaControlFocus, this.props.overrideMediaControlHandler).then(() => {
                LogUtils_1.LogUtil.info("Audio: setMediaControlFocus is configured..");
            });
        }
        else {
            LogUtils_1.LogUtil.warn("Audio: setMediaControlFocus is not configured..");
        }
        this.initializeMediaPlayer().then(() => {
            LogUtils_1.LogUtil.info("Audio: Media Player Initialized");
            this.isMediaPlayerInitialized = true;
            super.onMediaPlayerInitialized();
        });
    }
    /**
     * @hidden
     */
    componentDidMount() {
        LogUtils_1.LogUtil.info("Audio: componentDidMount");
    }
    /**
     * @hidden
     */
    componentWillUnmount() {
        LogUtils_1.LogUtil.info("Audio: componentWillUnmount");
        if (this.isMediaPlayerInitialized && this.mediaPlayer_ !== null) {
            if (this.captionsViewHandle_ !== null) {
                this.mediaPlayer_?.clearCaptionViewHandle(this.captionsViewHandle_);
                this.captionsViewHandle_ = null;
            }
        }
        this.destroyMediaPlayer().then(() => {
            LogUtils_1.LogUtil.info("Audio: media player destroyed");
        });
    }
    /**
     * @hidden
     */
    componentDidUpdate(prevProps) {
        LogUtils_1.LogUtil.info("Audio: componentDidUpdate");
        /* check if seek interval updated by app */
        if (prevProps.defaultSeekIntervalInSec !== this.props.defaultSeekIntervalInSec) {
            super.defaultSeekIntervalInSec = this.props.defaultSeekIntervalInSec;
        }
    }
    /**
     * @brief DeInitiailzes the player synchronously
     * @param timeoutInMs timeout threshold set by client
     * @returns MediaPlayerDeInitStatus
     */
    destroyMediaPlayerSync(timeoutInMs) {
        return super.destroyMediaPlayerSync(timeoutInMs);
    }
    /**
     * @hidden
     */
    render() {
        return (<react_native_1.View>
                <KeplerCaptionsView_1.KeplerCaptionsView onCaptionViewCreated={this.onCaptionViewCreated} show={this.props.showCaptions ?? true} style={{ zIndex: 1 }}/>
                {this.controls ? <MediaControls_1.default initFocus={true} mediaPlayer={this.mediaPlayer_} style={{ zIndex: 2 }}/> : null}
            </react_native_1.View>);
    }
}
exports.Audio = Audio;
;
