// @amazon-devices/react-native-w3cmedia/dist/interface/EventTargetInterface.js
"use strict";
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaEncryptedEvent = exports.TrackEvent = exports.Event = void 0;
;
class Event {
    type;
    target;
    constructor(...args) {
        if (args.length === 2) {
            this.type = args[0];
            this.target = args[1];
        }
    }
}
exports.Event = Event;
/**
 * https://html.spec.whatwg.org/multipage/media.html#trackevent
 */
class TrackEvent extends Event {
    /**
     * https://html.spec.whatwg.org/multipage/media.html#dom-trackevent-track
     */
    track;
    constructor(type, target, track) {
        super(type, target);
        this.track = track;
    }
}
exports.TrackEvent = TrackEvent;
class MediaEncryptedEvent extends Event {
    initDataType;
    initData;
    constructor(type, target, initDataType, initData) {
        super(type, target);
        this.initDataType = initDataType;
        this.initData = initData;
    }
}
exports.MediaEncryptedEvent = MediaEncryptedEvent;
;
;
