// @amazon-devices/react-native-w3cmedia/dist/interface/KeplerCaptionsView.js
"use strict";
/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerCaptionsView = void 0;
const React = __importStar(require("react"));
const LogUtils_1 = require("../LogUtils");
const react_native_1 = require("react-native");
const KeplerCaptionsViewSpec_1 = __importDefault(require("../components/captionsview/KeplerCaptionsViewSpec"));
const styles = react_native_1.StyleSheet.create({
    base: {
        flexGrow: 1,
        flexShrink: 1
    }
});
;
/**
 * @brief This is a react native component that closed captions and subtitles on the
 * screen. Use the component only when operating in pre-buffering mode using `VideoPlayer`
 * and `AudioPlayer` components.

 * Apps attach the captions handle passed by `onCaptionViewCreated` event callback to
 * the `VideoPlayer` & `AudioPlayer` by calling the `setCaptionViewHandle` function so that
 * the player is able to render captions or subtitles on screen.
 */
class KeplerCaptionsView extends React.Component {
    constructor(props) {
        super(props);
    }
    /**
     * @hidden
     */
    render() {
        const { style, onCaptionViewCreated } = this.props;
        LogUtils_1.LogUtil.info('render KeplerCaptionsView');
        return (<KeplerCaptionsViewSpec_1.default onCaptionViewCreated={(ev) => {
                LogUtils_1.LogUtil.info('got the caption consumer in CaptionWindow');
                onCaptionViewCreated?.(ev.nativeEvent.value);
            }} show={this.props.show ?? true} style={[styles.base, style]}/>);
    }
}
exports.KeplerCaptionsView = KeplerCaptionsView;
;
