// @amazon-devices/react-native-w3cmedia/dist/interface/Video.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Video = void 0;
/*
 * Copyright 2022-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const React = __importStar(require("react"));
const LogUtils_1 = require("../LogUtils");
const react_native_1 = require("react-native");
const KeplerCaptionsView_1 = require("./KeplerCaptionsView");
const MediaControls_1 = __importDefault(require("../components/mediacontrols/MediaControls"));
const KeplerVideoSurfaceView_1 = require("./KeplerVideoSurfaceView");
const Media_1 = require("../Media");
const VideoPlayer_1 = require("./VideoPlayer");
/**
 * Video is a react native component that implements the `HTMLVideoElement` interface
 * that extends the `HTMLMediaElement` interface.
 */
class Video extends Media_1.Media {
    poster_ = "";
    width_ = NaN;
    height_ = NaN;
    surfaceHandle = null;
    isMediaPlayerInitialized = false;
    /**
     * @hidden
     */
    constructor(props) {
        super(props, new VideoPlayer_1.VideoPlayer());
        this.onSurfaceMounted = this.onSurfaceMounted.bind(this);
        this.onSurfaceUnmounted = this.onSurfaceUnmounted.bind(this);
        this.onSurfaceCreated = this.onSurfaceCreated.bind(this);
        this.onSurfaceDestroyed = this.onSurfaceDestroyed.bind(this);
        if (this.props.width !== undefined) {
            this.width_ = this.props.width;
        }
        if (this.props.height !== undefined) {
            this.height_ = this.props.height;
        }
        if (this.props.mediaControlFocus !== undefined) {
            super.setMediaControlFocus(this.props.mediaControlFocus, this.props.overrideMediaControlHandler).then(() => {
                LogUtils_1.LogUtil.info("Video: setMediaControlFocus is configured..");
            });
        }
        else {
            LogUtils_1.LogUtil.warn("Video: setMediaControlFocus is not configured..");
        }
        this.initializeMediaPlayer().then(() => {
            LogUtils_1.LogUtil.info("Video: Media Player Initialized");
            this.isMediaPlayerInitialized = true;
            super.onMediaPlayerInitialized();
            // if surface was created before media player was initialized
            // pass the cached surface handle
            if (this.surfaceHandle !== null) {
                this.maybeSetSurface(this.surfaceHandle);
            }
        });
    }
    /**
     * @hidden
     */
    get videoWidth() {
        if (this.mediaPlayer_ !== null) {
            const videoPlayer = this.mediaPlayer_;
            return videoPlayer.videoWidth;
        }
        else {
            return NaN;
        }
    }
    /**
     * @hidden
     */
    get videoHeight() {
        if (this.mediaPlayer_ !== null) {
            const videoPlayer = this.mediaPlayer_;
            return videoPlayer.videoHeight;
        }
        else {
            return NaN;
        }
    }
    /**
     * @hidden
     */
    set width(w) {
        this.width_ = w;
    }
    /**
     * @hidden
     */
    get width() {
        if (this.width_ !== undefined) {
            return this.width_;
        }
        else {
            return 0;
        }
    }
    /**
     * @hidden
     */
    set height(h) {
        this.height_ = h;
    }
    /**
    * @hidden
    */
    get height() {
        if (this.height_ !== undefined) {
            return this.height_;
        }
        else {
            return 0;
        }
    }
    /**
     * @hidden
     */
    get poster() {
        return "emptyposter";
    }
    /**
     * @hidden
     */
    set poster(img) {
        this.poster_ = img;
    }
    /**
     * @hidden
     */
    get playsInline() {
        return true;
    }
    /**
     * @hidden
     */
    set playsInline(b) {
        // Video element is not expected to handle playsInline.
        // Main document element should handle this.
        // this.playsInline_ = b;
    }
    /**
     * @hidden
     */
    componentDidMount() {
        LogUtils_1.LogUtil.info("Video: componentDidMount");
    }
    /**
     * @hidden
     */
    componentWillUnmount() {
        LogUtils_1.LogUtil.info("Video: componentWillUnmount");
        this.destroyMediaPlayer().then(() => {
            LogUtils_1.LogUtil.info("Video: media player destroyed");
        });
    }
    /**
     * @hidden
     */
    componentDidUpdate(prevProps) {
        LogUtils_1.LogUtil.info("Video: componentDidUpdate");
        if (prevProps.src !== this.props.src) {
            if (this.props.src !== undefined) {
                super.src = this.props.src;
            }
        }
        /* check if seek interval updated by app */
        if (prevProps.defaultSeekIntervalInSec !== this.props.defaultSeekIntervalInSec) {
            super.defaultSeekIntervalInSec = this.props.defaultSeekIntervalInSec;
        }
    }
    /**
     * @hidden
     */
    onSurfaceMounted() {
        LogUtils_1.LogUtil.info("Video: onSurfaceMounted");
    }
    /**
     * @hidden
     */
    onSurfaceUnmounted() {
        LogUtils_1.LogUtil.info("Video: onSurfaceUnmounted");
    }
    /**
     * @hidden
     */
    maybeSetSurface(surfaceHandle) {
        LogUtils_1.LogUtil.info(`Video: maybeSetSurface: ${surfaceHandle}`);
        if (this.isMediaPlayerInitialized) {
            LogUtils_1.LogUtil.info(`maybeSetSurface: Media Player initialized.`);
            const videoPlayer = this.mediaPlayer_;
            videoPlayer.setSurfaceHandle(surfaceHandle);
        }
        // cache it
        this.surfaceHandle = surfaceHandle;
    }
    /**
     * @hidden
     */
    onSurfaceCreated(surfaceHandle) {
        LogUtils_1.LogUtil.info(`Video: onSurfaceCreated: ${surfaceHandle}`);
        if (surfaceHandle.length > 0) {
            this.maybeSetSurface(surfaceHandle);
        }
    }
    /**
    * @hidden
    */
    onSurfaceDestroyed(surfaceHandle) {
        LogUtils_1.LogUtil.info(`Video: onSurfaceDestroyed: ${surfaceHandle}`);
        if (this.mediaPlayer_ !== null) {
            const videoPlayer = this.mediaPlayer_;
            videoPlayer.clearSurfaceHandle(surfaceHandle);
        }
    }
    /**
     * @hidden
     */
    render() {
        LogUtils_1.LogUtil.info('render Video');
        return (<react_native_1.View style={{ width: this.width, height: this.height }}>
                <KeplerVideoSurfaceView_1.KeplerVideoSurfaceView style={{ zIndex: 0 }} scalingmode={this.props.scalingmode} onComponentDidMount={this.onSurfaceMounted} onSurfaceViewCreated={this.onSurfaceCreated} onSurfaceViewDestroyed={this.onSurfaceDestroyed} onComponentWillUnmount={this.onSurfaceUnmounted}>
                </KeplerVideoSurfaceView_1.KeplerVideoSurfaceView>
                <KeplerCaptionsView_1.KeplerCaptionsView onCaptionViewCreated={this.onCaptionViewCreated} show={this.props.showCaptions ?? true} style={{ width: '100%',
                height: '100%',
                top: 0,
                left: 0,
                position: 'absolute',
                backgroundColor: 'transparent',
                flexDirection: 'column',
                alignItems: 'center', zIndex: 2 }}/>
                  {this.props.controls ? <MediaControls_1.default initFocus={true} mediaPlayer={this.mediaPlayer_} style={{ zIndex: 2 }}/> : null}
            </react_native_1.View>);
    }
    /**
     * @hidden
     */
    handleEvent(event) {
        if (this.mediaPlayer_ !== null) {
            const videoPlayer = this.mediaPlayer_;
            videoPlayer.handleEvent(event);
        }
    }
    /**
     * @hidden
     * @brief Fetches video playback quality attributes.
     */
    getVideoPlaybackQuality() {
        if (this.mediaPlayer_ !== null) {
            const videoPlayer = this.mediaPlayer_;
            return videoPlayer.getVideoPlaybackQuality();
        }
        return new VideoPlayer_1.VideoPlaybackQualityImpl(Date.now(), NaN, NaN);
    }
    /**
     * @brief DeInitiailzes the player synchronously
     * @param timeoutInMs timeout threshold set by client
     * @returns MediaPlayerDeInitStatus
     */
    destroyMediaPlayerSync(timeoutInMs) {
        return super.destroyMediaPlayerSync(timeoutInMs);
    }
}
exports.Video = Video;
;
