// @amazon-devices/react-native-kepler/react-native.config.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

/**
 * CHANGELOG:
 * 74bc58c6 - Updated react-native to support kepler platform
 */

'use strict';

const ios = require('@react-native-community/cli-platform-ios');
const android = require('@react-native-community/cli-platform-android');

// amznmod_react - adding kepler support
const path = require('path');
const fs = require('fs');

function findConfigFiles(directory) {
  let configFiles = [];

  fs.readdirSync(directory).forEach(file => {
    // file is the path to the current file
    file = path.join(directory, file);
    // if file is directory, recur inside it and concat the returned value
    // otherwise, push the file to the return array (if it's a config file)
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      configFiles = configFiles.concat(findConfigFiles(file));
    } else if (file.split(/(\\|\/)/g).pop() === 'react-native.config.js') {
      configFiles.push(file);
    }
  });

  return configFiles;
}

function prepareDependency(dependency) {
  return dependency
    ? Object.fromEntries(
        Object.entries(dependency).map(entry => {
          var uiComponentName = entry[1]['uiComponentName'];
          if (entry[1]['libraryName'] && entry[1]['uiComponentName']) {
            throw new Error(
              "Invalid configuration. Both the fields 'libraryName' & 'uiComponentName' are configured, while only one is expected",
            );
          }
          if (entry[1]['uiComponentName']) {
            if (entry[1]['uiComponentName'][0] != '/')
              uiComponentName = '/' + uiComponentName;
          }
          return [
            entry[0],
            {
              libraryName: entry[1]['libraryName']
                ? entry[1]['libraryName']
                : '',
              uiComponentName: entry[1]['uiComponentName']
                ? uiComponentName
                : '',
              linkDynamic: entry[1]['linkDynamic']
                ? entry[1]['linkDynamic']
                : false,
              turbomodules: entry[1]['turbomodules']
                ? entry[1]['turbomodules']
                : [],
              components: entry[1]['components'] ? entry[1]['components'] : [],
              provider: entry[1]['provider']
                ? entry[1]['provider']
                : `custom`,
            },
          ];
        }),
      )
    : null;
}

const isReactNativeKepler = path.basename(__dirname) === 'react-native-kepler';
const iosCommands = isReactNativeKepler ? [] : ios.commands;
const androidCommands = isReactNativeKepler ? [] : android.commands;
// amznmod_react

module.exports = {
  commands: [...ios.commands, ...android.commands],
  platforms: {
    ios: {
      projectConfig: ios.projectConfig,
      dependencyConfig: ios.dependencyConfig,
    },
    android: {
      projectConfig: android.projectConfig,
      dependencyConfig: android.dependencyConfig,
    },
    // amznmod_react - adding kepler support
    kepler: {
      npmPackageName: '@amazon-devices/react-native-kepler',
      linkConfig: () => null,
      projectConfig: (projectRoot, projectParams) => {
        let fullDependencies = {};
        const configFiles = findConfigFiles(
          path.join(projectRoot, '/node_modules'),
        );
        configFiles.forEach(element => {
          const fileExport = require(element);
          fullDependencies = {
            ...fullDependencies,
            ...fileExport.dependency?.platforms?.kepler?.autolink,
          };
        });
        return {
          mainBundle:
            projectParams && projectParams['mainBundle']
              ? projectParams['mainBundle']
              : 'index.hermes.bundle',
          initModule:
            projectParams && projectParams['initModule']
              ? projectParams['initModule']
              : '',
          developmentEnabled:
            projectParams && projectParams['developmentEnabled']
              ? projectParams['developmentEnabled']
              : false,
          runtimeFeatures:
            projectParams && projectParams['runtimeFeatures']
              ? projectParams['runtimeFeatures']
              : {},
          fullDependencies: prepareDependency(fullDependencies),
          frameworkRenderOptimizationsEnabled:
            projectParams &&
            projectParams['frameworkRenderOptimizationsEnabled']
              ? projectParams['frameworkRenderOptimizationsEnabled']
              : false,
        };
      },
      dependencyConfig: (projectRoot, dependencyParams) => {
        return {
          autolink:
            dependencyParams && dependencyParams['autolink']
              ? prepareDependency(dependencyParams['autolink'])
              : null,
        };
      },
    },
    // amznmod_react
  },
  /**
   * Used when running RNTester (with React Native from source)
   */
  reactNativePath: '.',
  project: {
    ios: {
      sourceDir: '../packages/rn-tester',
    },
    android: {
      sourceDir: '../packages/rn-tester',
    },
    kepler: {
      prebuilt: {
        path: '{{{sdkPath}}}/component-sdks/VegaUIReact/{{arch}}/lib/libreact-kepler-prebuilt-067.so',
      },
    },
  },
};
