// @amzn/react-native-kepler/Libraries/PermissionsKepler/PermissionsKepler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.usePermissionsKepler = exports.PERMISSIONS = void 0;
/**
 * Copyright (c) 2025 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @format
 */
const react_1 = require("react");
const NativePermissionsKepler_1 = __importDefault(require("./NativePermissionsKepler"));
const Platform = require("../Utilities/Platform");
/**
 * Enum for system permissions.
 * Each entry represents a permission defined within the system.
 */
var PERMISSIONS;
(function (PERMISSIONS) {
    PERMISSIONS["NET_INFO"] = "com.amazon.network.privilege.net-info";
    PERMISSIONS["ACCESS_ACCESSORY_CONTROL_ACCOUNT"] = "com.amazon.acc.privilege.access";
    PERMISSIONS["CONNECT_ACCESSORY_CONTROL_ACCOUNT"] = "com.amazon.acc.privilege.connect";
    PERMISSIONS["DISCOVER_ACCESSORY_CONTROL_ACCOUNT"] = "com.amazon.acc.privilege.discover";
    PERMISSIONS["ACCESS_COARSE_LOCATION"] = "com.amazon.location.privilege.access.coarse";
    PERMISSIONS["ACCESS_FINE_LOCATION"] = "com.amazon.location.privilege.access.fine";
    PERMISSIONS["RECORD_AUDIO"] = "com.amazon.audio.privilege.microphone.access";
    PERMISSIONS["DEVICE_FRIENDLY_NAME"] = "com.amazon.devconf.privilege.identifiers.device-friendly-name";
    PERMISSIONS["BLUETOOTH_ADVERTISE"] = "com.amazon.bluetooth.privilege.advertise";
    PERMISSIONS["BLUETOOTH_DISCOVERABLE"] = "com.amazon.bluetooth.privilege.discoverable";
    PERMISSIONS["READ_EXTERNAL_STORAGE"] = "com.amazon.external-disk.privilege.readonly";
    PERMISSIONS["DISCOVERABLE_ACCESSORY_CONTROL_ACCOUNT"] = "com.amazon.acc.privilege.discoverable";
    PERMISSIONS["CAMERA"] = "com.amazon.camera.privilege.access";
    PERMISSIONS["READ_CALENDAR"] = "com.amazon.calendar.privilege.read";
    PERMISSIONS["WRITE_CALENDAR"] = "com.amazon.calendar.privilege.write";
})(PERMISSIONS = exports.PERMISSIONS || (exports.PERMISSIONS = {}));
const usePermissionsKepler = () => {
    /**
     * Checks if a specific permission has been granted.
     *
     * @param permission - The permission to check.
     * @returns A promise that resolves to a boolean value indicating whether the specified permission has been granted.
     */
    const check = react_1.useCallback((permission) => {
        if (Platform.OS !== "kepler") {
            console.warn('"PermissionsKepler" module works only for Kepler platform.');
            return Promise.resolve(false);
        }
        return NativePermissionsKepler_1.default.check(permission).catch(() => {
            return Promise.reject(new Error(`check() failed for ${permission}`));
        });
    }, []);
    /**
     * Requests a specific permission from the user.
     *
     * The function checks whether the permission is already granted:
     * - If the permission is already granted, it automatically resolves to `granted` without showing any dialog.
     * - If the permission is not granted, a dialog appears prompting the user with Allow/Deny options.
     *   - Resolves to `granted` if the user selects Allow.
     *   - Resolves to `denied` if the user selects Deny.
     *
     * @param permission - The permission to request.
     * @returns A promise that resolves to a PermissionStatus indicating whether the permission was granted or denied.
     */
    const request = react_1.useCallback((permission) => {
        if (Platform.OS !== "kepler") {
            console.warn('"PermissionsKepler" module works only for Kepler platform.');
            return Promise.resolve("denied");
        }
        return NativePermissionsKepler_1.default.request(permission)
            .then((result) => {
            return result === true ? "granted" : "denied";
        })
            .catch(() => {
            return Promise.reject(new Error(`request() failed for ${permission}`));
        });
    }, []);
    return [check, request];
};
exports.usePermissionsKepler = usePermissionsKepler;
