// @amazon-devices/react-native-w3cmedia/dist/components/mediacontrols/SeekBarUI/Slider/tokenResolver.slider.js
"use strict";
/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveProgressBarTokens = exports.resolveMarkerTokens = exports.resolveCenteredNodeTokens = exports.resolveTimeShiftIndicatorTokens = exports.resolveBarContainerTokens = exports.resolveThumbDotTokens = exports.resolveThumbContainerTokens = exports.thumbRippleRadius = exports.thumbRippleSize = void 0;
exports.thumbRippleSize = 80;
exports.thumbRippleRadius = exports.thumbRippleSize / 2;
const thumbSize = 24;
const thumbColor = 'red';
const resolveThumbContainerTokens = () => {
    return {
        zIndex: 1,
        position: 'absolute',
        width: 1850,
        alignItems: 'flex-start',
        marginLeft: 16
    };
};
exports.resolveThumbContainerTokens = resolveThumbContainerTokens;
const resolveThumbDotTokens = () => {
    return {
        width: thumbSize,
        height: thumbSize,
        borderRadius: thumbSize / 2,
        position: 'absolute',
        backgroundColor: thumbColor,
        opacity: 1.0,
        shadowColor: '#00000035',
        shadowOffset: {
            width: 0,
            height: 0
        },
        shadowRadius: 6
    };
};
exports.resolveThumbDotTokens = resolveThumbDotTokens;
const resolveBarContainerTokens = () => {
    return {
        width: 1920,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-start',
        paddingLeft: 32
    };
};
exports.resolveBarContainerTokens = resolveBarContainerTokens;
const resolveTimeShiftIndicatorTokens = () => {
    return {
        position: 'absolute',
        backgroundColor: '#808080',
        width: 1850,
        justifyContent: 'center',
    };
};
exports.resolveTimeShiftIndicatorTokens = resolveTimeShiftIndicatorTokens;
const resolveCenteredNodeTokens = () => {
    return {
        width: 0,
        justifyContent: 'center',
        alignItems: 'center'
    };
};
exports.resolveCenteredNodeTokens = resolveCenteredNodeTokens;
const resolveMarkerTokens = () => {
    return {
        zIndex: 1,
        position: 'absolute',
        overflow: 'visible',
        width: 1850,
    };
};
exports.resolveMarkerTokens = resolveMarkerTokens;
const resolveProgressBarTokens = () => {
    return {
        width: 1850,
        flexDirection: "row",
        alignItems: "center",
        justifyContent: 'flex-start'
    };
};
exports.resolveProgressBarTokens = resolveProgressBarTokens;
