// @amzn/react-native-kepler/Libraries/Components/Touchable/TouchableWithoutFeedback.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

/**
 * CHANGELOG:
 * Added isTouchable/isPressable
 * Added isTouchable import
 * Added accessibilityDescribedBy, accessibilityOrientationText
 * Added VoiceOrdinal, VoiceExternalId, VoiceEntityType, VoiceAlternativeNames
 * Added ariaPosInSet, alexaEntityType, keplerAtlLabels, ariaLabel
 * Added alexaExternalIds
 * Pass onPress from Pressability
 * Added enableSynchronousFocusEventsKepler
 * Rename keplerAtlLabels to keplerAltLabels
 * Added acccessibilityLabelledBy
 */

import type {
  AccessibilityActionEvent,
  AccessibilityActionInfo,
  AccessibilityRole,
  AccessibilityState,
  AccessibilityValue,
  VoiceEntityTypeValue,
} from '../../Components/View/ViewAccessibility';
import type {EdgeInsetsOrSizeProp} from '../../StyleSheet/EdgeInsetsPropType';
import type {
  BlurEvent,
  FocusEvent,
  LayoutEvent,
  PressEvent,
} from '../../Types/CoreEventTypes';

import View from '../../Components/View/View';
import Pressability, {
  type PressabilityConfig,
  //amznmod_react - Add isTouchable import
  isTouchable,
} from '../../Pressability/Pressability';
import {PressabilityDebugView} from '../../Pressability/PressabilityDebug';
import * as React from 'react';
//amznmod_react - Adding kepler platform support
import Platform from '../../Utilities/Platform';

type Props = $ReadOnly<{|
  accessibilityActions?: ?$ReadOnlyArray<AccessibilityActionInfo>,
  accessibilityElementsHidden?: ?boolean,
  accessibilityHint?: ?Stringish,
  accessibilityLanguage?: ?Stringish,
  accessibilityIgnoresInvertColors?: ?boolean,
  accessibilityLabel?: ?Stringish,
  // amznmod_react: add accessibilityLabelledBy
  accessibilityLabelledBy?: ?Stringish,
  accessibilityLiveRegion?: ?('none' | 'polite' | 'assertive'),
  accessibilityRole?: ?AccessibilityRole,
  accessibilityState?: ?AccessibilityState,
  accessibilityValue?: ?AccessibilityValue,
  'aria-valuemax'?: AccessibilityValue['max'],
  'aria-valuemin'?: AccessibilityValue['min'],
  'aria-valuenow'?: AccessibilityValue['now'],
  'aria-valuetext'?: AccessibilityValue['text'],
  accessibilityViewIsModal?: ?boolean,
  'aria-modal'?: ?boolean,
  accessible?: ?boolean,
  // amznmod_react - add accessibilityDescribedBy, accessibilityOrientationText
  accessibilityDescribedBy?: ?$ReadOnlyArray<Stringish>,
  accessibilityDescribedByKepler?: ?$ReadOnlyArray<Stringish>, // TODO: to be deprecated
  accessibilityOrientationText?: ?Stringish,
  accessibilityOrientationTextKepler?: ?Stringish, // TODO: to be deprecated
  /**
   * alias for accessibilityState
   *
   * see https://reactnative.dev/docs/accessibility#accessibilitystate
   */
  'aria-busy'?: ?boolean,
  'aria-checked'?: ?boolean | 'mixed',
  'aria-disabled'?: ?boolean,
  'aria-expanded'?: ?boolean,
  'aria-selected'?: ?boolean,
  'aria-hidden'?: ?boolean,
  'aria-live'?: ?('polite' | 'assertive' | 'off'),
  'aria-label'?: ?Stringish,
  // amznmod_react: add accessibilityLabelledBy, aria-labelledby
  'aria-labelledby'?: ?Stringish,
  children?: ?React.Node,
  delayLongPress?: ?number,
  delayPressIn?: ?number,
  delayPressOut?: ?number,
  disabled?: ?boolean,
  // amznmod_react - add isTouchable/isPressable
  isPressable?: ?boolean,
  // amznmod_react Add isLongPressRegistered
  isLongPressRegistered?: ?boolean,
  // amznmod_react - Add voiceOrdinal
  voiceOrdinal?: ?number,
  // amznmod_react - Add voiceEntityType
  voiceEntityType?: ?VoiceEntityTypeValue,
  // amznmod_react - Add voiceExternalIds
  voiceExternalIds?: object,
  // amznmod_react - Add voiceAlternativeNames
  voiceAlternativeNames?: ?$ReadOnlyArray<string>,
  // amznmod_react - Add ariaPosInset
  ariaPosInSet?: ?number,
  // amznmod_react - Add alexaEntityType
  alexaEntityType?: ?VoiceEntityTypeValue,
  // amznmod_react - Add keplerAtlLabels (Deprecated, use keplerAltLabels)
  keplerAtlLabels?: ?$ReadOnlyArray<string>,
  // amznmod_react - Add keplerAltLabels
  keplerAltLabels?: ?$ReadOnlyArray<string>,
  // amznmod_react - Add alexaExternalIds
  alexaExternalIds?: object,
  // amznmod_react - Add ariaLabel
  ariaLabel?: ?string,
  focusable?: ?boolean,
  hitSlop?: ?EdgeInsetsOrSizeProp,
  id?: string,
  importantForAccessibility?: ?('auto' | 'yes' | 'no' | 'no-hide-descendants'),
  nativeID?: ?string,
  onAccessibilityAction?: ?(event: AccessibilityActionEvent) => mixed,
  onBlur?: ?(event: BlurEvent) => mixed,
  onFocus?: ?(event: FocusEvent) => mixed,
  onLayout?: ?(event: LayoutEvent) => mixed,
  onLongPress?: ?(event: PressEvent) => mixed,
  onPress?: ?(event: PressEvent) => mixed,
  onPressIn?: ?(event: PressEvent) => mixed,
  onPressOut?: ?(event: PressEvent) => mixed,
  pressRetentionOffset?: ?EdgeInsetsOrSizeProp,
  rejectResponderTermination?: ?boolean,
  testID?: ?string,
  touchSoundDisabled?: ?boolean,
  // amznmod_react - Add enableSynchronousFocusEventsKepler
  enableSynchronousFocusEventsKepler?: ?boolean,
|}>;

type State = $ReadOnly<{|
  pressability: Pressability,
|}>;

const PASSTHROUGH_PROPS = [
  'accessibilityActions',
  'accessibilityElementsHidden',
  'accessibilityHint',
  'accessibilityDescribedBy',
  'accessibilityDescribedByKepler', // TODO: to be deprecated
  'accessibilityOrientationText',
  'accessibilityOrientationTextKepler', // TODO: to be deprecated
  'accessibilityLanguage',
  'accessibilityIgnoresInvertColors',
  'accessibilityLabel',
  'accessibilityLiveRegion',
  'accessibilityRole',
  'accessibilityValue',
  'aria-valuemax',
  'aria-valuemin',
  'aria-valuenow',
  'aria-valuetext',
  // amznmod_react: Support aria-label props in TouchableWithoutFeedback
  'aria-label',
  'accessibilityViewIsModal',
  'aria-modal',
  'hitSlop',
  'importantForAccessibility',
  'nativeID',
  'onAccessibilityAction',
  'onBlur',
  'onFocus',
  'onLayout',
  'testID',
  'voiceOrdinal',
  'voiceEntityType',
  'voiceExternalIds',
  'voiceAlternativeNames',
  // amznmod_react: provide backward compability to "ariaLabel(ariaLabelVoice) prop" (PROJVEGA-153671)
  'ariaLabel',
  // amznmod_react - Add enableSynchronousFocusEventsKepler
  'enableSynchronousFocusEventsKepler'
];

class TouchableWithoutFeedback extends React.Component<Props, State> {
  state: State = {
    pressability: new Pressability(createPressabilityConfig(this.props)),
  };

  render(): React.Node {
    const element = React.Children.only<$FlowFixMe>(this.props.children);
    const children: Array<React.Node> = [element.props.children];
    const ariaLive = this.props['aria-live'];

    if (__DEV__) {
      if (element.type === View) {
        children.push(
          <PressabilityDebugView color="red" hitSlop={this.props.hitSlop} />,
        );
      }
    }

    let _accessibilityState = {
      busy: this.props['aria-busy'] ?? this.props.accessibilityState?.busy,
      checked:
        this.props['aria-checked'] ?? this.props.accessibilityState?.checked,
      disabled:
        this.props['aria-disabled'] ?? this.props.accessibilityState?.disabled,
      expanded:
        this.props['aria-expanded'] ?? this.props.accessibilityState?.expanded,
      selected:
        this.props['aria-selected'] ?? this.props.accessibilityState?.selected,
    };


    // amznmod_react: add accessibilityLabelledBy
    const _accessibilityLabelledBy =
      this.props['aria-labelledby'] ?? this.props.accessibilityLabelledBy;

    // BACKWARD-COMPATIBILITY: Focus and blur events were never supported before
    // adopting `Pressability`, so preserve that behavior.
    const {onBlur, onFocus, ...eventHandlersWithoutBlurAndFocus} =
      this.state.pressability.getEventHandlers();

    // amznmod_react - add isTouchable/isPressable
    const isTouchHandlersRegistered = isTouchable(this.props);
    const elementProps: {[string]: mixed, ...} = {
      ...eventHandlersWithoutBlurAndFocus,
      accessible: this.props.accessible !== false,
      // amznmod_react - add isTouchable/isPressable
      isPressable: isTouchHandlersRegistered,

      accessibilityState:
        this.props.disabled != null
          ? {
              ..._accessibilityState,
              disabled: this.props.disabled,
            }
          : _accessibilityState,
      focusable:
        this.props.focusable !== false && this.props.onPress !== undefined,

      // amznmod_react Add isLongPressRegistered
      isLongPressRegistered: this.props.onLongPress !== undefined,

      accessibilityElementsHidden:
        this.props['aria-hidden'] ?? this.props.accessibilityElementsHidden,
      // amznmod_react: add accessibilityLabelledBy
      accessibilityLabelledBy: _accessibilityLabelledBy,
      importantForAccessibility:
        this.props['aria-hidden'] === true
          ? 'no-hide-descendants'
          : this.props.importantForAccessibility,
      accessibilityLiveRegion:
        ariaLive === 'off'
          ? 'none'
          : ariaLive ?? this.props.accessibilityLiveRegion,
      nativeID: this.props.id ?? this.props.nativeID,
      // amznmod_react - Add ariaPosInSet
      ariaPosInSet: this.props.ariaPosInSet ?? this.props.voiceOrdinal,
      // amznmod_react - Add alexaEntityType
      alexaEntityType: this.props.alexaEntityType ?? this.props.voiceEntityType,
      // amznmod_react - Add keplerAltLabels
      keplerAltLabels:
        this.props.keplerAltLabels ?? this.props.keplerAtlLabels ?? this.props.voiceAlternativeNames,
      // amznmod_react - Add alexaExternalIds
      alexaExternalIds:
        this.props.alexaExternalIds ?? this.props.voiceExternalIds,
    };
    for (const prop of PASSTHROUGH_PROPS) {
      if (this.props[prop] !== undefined) {
        elementProps[prop] = this.props[prop];
      }
    }

    // amznmod_react: add warning log for deprecated props
    if (this.props.keplerAtlLabels) {
      console.warn('keplerAtlLabels has been replaced by keplerAltLabels. please update your code to use keplerAltLabels')
    }
    return React.cloneElement(element, elementProps, ...children);
  }

  componentDidUpdate(): void {
    this.state.pressability.configure(createPressabilityConfig(this.props));
  }

  componentWillUnmount(): void {
    this.state.pressability.reset();
  }
}

function createPressabilityConfig({
  'aria-disabled': ariaDisabled,
  ...props
}: Props): PressabilityConfig {
  const accessibilityStateDisabled =
    ariaDisabled ?? props.accessibilityState?.disabled;
  return {
    cancelable: !props.rejectResponderTermination,
    disabled:
      props.disabled !== null ? props.disabled : accessibilityStateDisabled,
    hitSlop: props.hitSlop,
    delayLongPress: props.delayLongPress,
    delayPressIn: props.delayPressIn,
    delayPressOut: props.delayPressOut,
    minPressDuration: 0,
    pressRectOffset: props.pressRetentionOffset,
    android_disableSound: props.touchSoundDisabled,
    onBlur: props.onBlur,
    onFocus: props.onFocus,
    onLongPress: props.onLongPress,
    onPress: props.onPress,
    onPressIn: props.onPressIn,
    onPressOut: props.onPressOut,
  };
}

TouchableWithoutFeedback.displayName = 'TouchableWithoutFeedback';

module.exports = TouchableWithoutFeedback;
