// @amazon-devices/kepler-media-controls/dist/MediaControlEndpointLocatorAsync.js
"use strict";
/*
 * Copyright 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaControlEndpointLocatorAsync = void 0;
const KeplerMediaControlsClientTM_1 = __importDefault(require("./turbo-modules/KeplerMediaControlsClientTM"));
const MediaControlClientAsync_1 = require("./MediaControlClientAsync");
if (KeplerMediaControlsClientTM_1.default === undefined || KeplerMediaControlsClientTM_1.default === null) {
    throw new Error('Media Controls Client TurboModule is null');
}
/**
 * Class to represent a subscription to a listener from the endpoint locator changes.
 * Used to remove a listener.
 */
class Subscription {
    token;
    constructor(token) {
        this.token = token;
    }
    /**
     * called to unsubscribe the listener from the endpoint locator changes
     */
    unsubscribe() {
        KeplerMediaControlsClientTM_1.default.removeChangeListener(this.token);
    }
}
/**
 * MediaControlEndpointLocatorAsync class used to locate media control endpoints
 */
class MediaControlEndpointLocatorAsync {
    /**
     * Call to obtain a list of media control clients
     * @returns @type {Promise} of @type {MediaControlTypes.IMediaControlClientAsync[]} containing
     * clients to interact with all available media control endpoints
     */
    async getMediaControlEndpoints() {
        return await KeplerMediaControlsClientTM_1.default.getMediaControlEndpoints().then((endpoints) => {
            const clients = endpoints.map((endpoint) => new MediaControlClientAsync_1.MediaControlClientAsync(endpoint));
            return clients;
        });
    }
    /**
     * Subscribes a listener to any changes in the list of control endpoints
     * @param listener @type {MediaControlTypes.IMediaControlEndpointLocatorListener}
     * @returns @type {Promise} of @type {MediaControlTypes.ISubscription} which can be used
     * to remove the listener
     */
    async addChangeListener(listener) {
        return await new Promise((resolve) => {
            KeplerMediaControlsClientTM_1.default.addChangeListener(listener.onEndpointsChanged).then((token) => {
                resolve(new Subscription(token));
            }).catch((error) => error);
        });
    }
}
exports.MediaControlEndpointLocatorAsync = MediaControlEndpointLocatorAsync;
