// @amazon-devices/kepler-media-controls/dist/MediaControlServer.js
"use strict";
/*
 * Copyright 2024-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaControlServer = void 0;
const KeplerMediaControlsServerTM_1 = __importDefault(require("./turbo-modules/KeplerMediaControlsServerTM"));
if (KeplerMediaControlsServerTM_1.default === undefined || KeplerMediaControlsServerTM_1.default === null) {
    throw new Error('Media Controls TurboModule is null');
}
function normalizeError(error) {
    if (error instanceof Error) {
        return {
            type: error.name,
            message: error.message
        };
    }
    return {
        type: 'BaseError',
        message: 'Call rejected'
    };
}
/**
 * The entry point for KMC server side APIs. Provides the ability to register handlers
 * and update media session states
 */
class MediaControlServer {
    /**
     * Get the major version number of the TM interface
     * @returns @type {number} the major version
     */
    static getMajorVersion() {
        return KeplerMediaControlsServerTM_1.default.getMajorVersion();
    }
    /**
     * Get the minor version number of the TM interface
     * @returns @type {number} the minor version
     */
    static getMinorVersion() {
        return KeplerMediaControlsServerTM_1.default.getMinorVersion();
    }
    /**
     * Get the patch version number of the TM interface
     * @returns @type {number} the patch version
     */
    static getPatchVersion() {
        return KeplerMediaControlsServerTM_1.default.getPatchVersion();
    }
    /**
     * Set the event handlers associated with the provided component
     * @param handler @type {KeplerMediaControlsTypes.IMediaControlHandlerAsync} Object containing
     * the event handlers
     * @param componentInstance @type {IComponentInstance} The component instance for which
     * the handlers are being set
     */
    setHandlerForComponent(handler, componentInstance) {
        KeplerMediaControlsServerTM_1.default.setHandlerForComponent(componentInstance, (transactionId, sessionId) => {
            handler.handlePlay(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, context, sessionId) => {
            handler.handlePause(sessionId, context).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleTogglePlayPause(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleStop(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleStartOver(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleFastForward(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleRewind(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, speed, sessionId) => {
            handler.handleSetPlaybackSpeed(speed, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, delta, sessionId) => {
            handler.handleSkipForward(delta, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, delta, sessionId) => {
            handler.handleSkipBackward(delta, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, position, sessionId) => {
            handler.handleSeek(position, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, volume, sessionId) => {
            handler.handleSetAudioVolume(volume, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, audioTrack, sessionId) => {
            handler.handleSetAudioTrack(audioTrack, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, textTrack, sessionId) => {
            handler.handleEnableTextTrack(textTrack, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleDisableTextTrack(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleNext(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handlePrevious(sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, enable, sessionId) => {
            handler.handleEnableShuffle(enable, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, mode, sessionId) => {
            handler.handleSetRepeatMode(mode, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, id, rating, sessionId) => {
            handler.handleSetRating(id, rating, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, id) => {
            handler.handleGetMetadataInfo(id).then((metadata) => {
                KeplerMediaControlsServerTM_1.default.getMetadataResponse(transactionId, metadata);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, action, sessionId) => {
            handler.handleCustomAction(action, sessionId).then(() => {
                KeplerMediaControlsServerTM_1.default.transactionSuccess(transactionId);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }, (transactionId, sessionId) => {
            handler.handleGetSessionState(sessionId).then((states) => {
                KeplerMediaControlsServerTM_1.default.getSessionStateResponse(transactionId, states);
            }).catch((reason) => {
                KeplerMediaControlsServerTM_1.default.transactionError(transactionId, normalizeError(reason));
            });
        }).catch((error) => error);
    }
    /**
     * Update the media session states
     * @param states @type {KeplerMediaControlsTypes.MediaSessionState[]} Array of media session states
     */
    updateMediaSessionStates(states) {
        KeplerMediaControlsServerTM_1.default.updateMediaSessionStates(states);
    }
}
exports.MediaControlServer = MediaControlServer;
