// @amzn/react-native-kepler/Libraries/AppState/NativeAppState.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @flow
 * @format
 */

import type {TurboModule} from '../TurboModule/RCTExport';

import * as TurboModuleRegistry from '../TurboModule/TurboModuleRegistry';

// amznmod_react - Add Kepler support for module states
export type ModuleStateKepler = {|
    module_name: string,
    module_state: string
|};

//amznmod_react - extends KeplerTurboModule
export interface Spec extends KeplerTurboModule {
  +getConstants: () => {|
    initialAppState: string,
  |};
  +getCurrentAppState: (
      // amznmod_react - Add Kepler support for module states
      success: (appState: {|app_state: string, module_states: ModuleStateKepler[]|}) => void,
    error: (error: Object) => void,
  ) => void;

  // Events
  +addListener: (eventName: string) => void;
  +removeListeners: (count: number) => void;
}

export default (TurboModuleRegistry.getEnforcing<Spec>('AppState'): Spec);
