// @amzn/react-native-kepler/Libraries/Components/View/ReactNativeViewAttributes.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * Portions of this file are copyright (c) 2021 - 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 * 
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

/**
 * CHANGELOG:
 * Added VoiceOrdinal, VoiceExternalIds, VoiceEntityTypes, VoicePreferredNames
 * Added accessibilityDescribedBy, accessibilityOrientationText
 * Added transportControls
 * Added ariaPosInSet, alexaEntityType, keplerAtlLabels, ariaLabel
 * Added alexaExternalIds
 * Added enableSynchronousFocusEventsKepler
 * Rename keplerAtlLabels to keplerAltLabels
 */

'use strict';
import ReactNativeStyleAttributes from './ReactNativeStyleAttributes';

const UIView = {
  pointerEvents: true,
  accessible: true,
  // amznmod_react Add isPressable
  isPressable: true,
  // amznmod_react Add isLongPressRegistered
  isLongPressRegistered: true,
  // amznmod_react Add voiceOrdinal
  voiceOrdinal: true,
  // amznmod_react Add voiceExternalIds
  voiceExternalIds: true,
  // amznmod_react Add voiceEntityType
  voiceEntityType: true,
  // amznmod_react Add voiceAlternativeNames
  voiceAlternativeNames: true,
  // amznmod_react - Add ariaPosInset
  ariaPosInSet: true,
  // amznmod_react Add alexaEntityType
  alexaEntityType: true,
  // amznmod_react - Add keplerAtlLabels (Deprecated, use keplerAltLabels)
  keplerAtlLabels: true,
  // amznmod_react - Add keplerAltLabels
  keplerAltLabels: true,
  // amznmod_react - Add alexaExternalIds
  alexaExternalIds: true,
  // amznmod_react - Add ariaLabel
  ariaLabel: true,
  accessibilityActions: true,
  accessibilityLabel: true,
  accessibilityLiveRegion: true,
  accessibilityRole: true,
  accessibilityState: true,
  // amznmod_react - Add transportControls
  transportControls: true,
  accessibilityValue: true,
  accessibilityHint: true,
  // amznmod_react - add accessibilityDescribedBy, accessibilityOrientationText
  accessibilityDescribedBy: true,
  accessibilityDescribedByKepler: true, // TODO: to be deprecated
  accessibilityOrientationText: true,
  accessibilityOrientationTextKepler: true, // TODO: to be deprecated
  accessibilityLanguage: true,
  importantForAccessibility: true,
  nativeID: true,
  testID: true,
  renderToHardwareTextureAndroid: true,
  // amznmod_react - Add texture cache flag on Kepler
  renderToHardwareTextureKepler: true,
  shouldRasterizeIOS: true,
  onLayout: true,
  onAccessibilityAction: true,
  onAccessibilityTap: true,
  onMagicTap: true,
  onAccessibilityEscape: true,
  collapsable: true,
  needsOffscreenAlphaCompositing: true,
  // amznmod_react - Add enableSynchronousFocusEventsKepler
  enableSynchronousFocusEventsKepler: true,
  style: ReactNativeStyleAttributes,
};

const RCTView = {
  ...UIView,

  // This is a special performance property exposed by RCTView and useful for
  // scrolling content when there are many subviews, most of which are offscreen.
  // For this property to be effective, it must be applied to a view that contains
  // many subviews that extend outside its bound. The subviews must also have
  // overflow: hidden, as should the containing view (or one of its superviews).
  removeClippedSubviews: true,
};

const ReactNativeViewAttributes = {
  UIView: UIView,
  RCTView: RCTView,
};

module.exports = ReactNativeViewAttributes;
