// @amzn/react-native-kepler/Libraries/Components/View/ReactNativeViewViewConfigAndroid.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Portions of this file are copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

/**
 * CHANGELOG:
 * Added isFocusableInTouchMode
 * enable TVFocusGuide props/methods
 */

'use strict';

const ReactNativeViewViewConfigAndroid = {
  uiViewClassName: 'RCTView',
  bubblingEventTypes: {
    topSelect: {
      phasedRegistrationNames: {
        bubbled: 'onSelect',
        captured: 'onSelectCapture',
      },
    },
    topAssetDidLoad: {
      phasedRegistrationNames: {
        bubbled: 'onAssetDidLoad',
        captured: 'onAssetDidLoadCapture',
      },
    },
  },
  directEventTypes: {
    topClick: {
      registrationName: 'onClick',
    },
    topContentSizeChange: {
      registrationName: 'onContentSizeChange',
    },
    topLoadingError: {
      registrationName: 'onLoadingError',
    },
    topLoadingFinish: {
      registrationName: 'onLoadingFinish',
    },
    topLoadingStart: {
      registrationName: 'onLoadingStart',
    },
    topMessage: {
      registrationName: 'onMessage',
    },
    topMomentumScrollBegin: {
      registrationName: 'onMomentumScrollBegin',
    },
    topMomentumScrollEnd: {
      registrationName: 'onMomentumScrollEnd',
    },
    topScroll: {
      registrationName: 'onScroll',
    },
    topScrollBeginDrag: {
      registrationName: 'onScrollBeginDrag',
    },
    topScrollEndDrag: {
      registrationName: 'onScrollEndDrag',
    },
    topSelectionChange: {
      registrationName: 'onSelectionChange',
    },
    onAssetDidLoad: {
      registrationName: 'onAssetDidLoad',
    },
  },
  validAttributes: {
    hasTVPreferredFocus: true,
    focusable: true,
    // amznmod_react - Add isFocusableInTouchMode
    isFocusableInTouchMode: true,
    nativeBackgroundAndroid: true,
    nativeForegroundAndroid: true,
    nextFocusDown: true,
    nextFocusForward: true,
    nextFocusLeft: true,
    nextFocusRight: true,
    nextFocusUp: true,
    // amznmod_react - enable TVFocusGuide props/methods
    autoFocus: true,
    trapFocusUp: true,
    trapFocusDown: true,
    trapFocusLeft: true,
    trapFocusRight: true,
    destinations: true, //amznmod_react - Using destinations as prop instead of method
  },
};

module.exports = ReactNativeViewViewConfigAndroid;
