// @amazon-devices/kepler-cli-platform/dist/common/packageJsonUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2025 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
const path = __importStar(require("path"));
const common_1 = require("../common");
const merge_1 = __importDefault(require("lodash/merge"));
class ConfigLoader {
    constructor(ctx) {
        this.ctx = ctx;
        this.config = this.loadConfig();
    }
    loadConfig() {
        var _a, _b;
        const packageJsonPath = path.join(this.ctx.root, 'package.json');
        let userConfig = require(packageJsonPath);
        // Supports legacy disableSplitBundle config until .20
        if (((_a = userConfig === null || userConfig === void 0 ? void 0 : userConfig.kepler) === null || _a === void 0 ? void 0 : _a.disableSplitBundle) !== undefined) {
            common_1.logger.warn(`
⚠️ Deprecation Warning:
The "disableSplitBundle" option is deprecated and will be removed in next release.
Please update your configuration to use "useSystemJsBundles" instead.
Current configuration in package.json:
{
  "kepler": {
    "disableSplitBundle": ${userConfig.kepler.disableSplitBundle}
  }
}

Should be changed to:
{
  "kepler": {
    "useSystemJsBundles": ${!userConfig.kepler.disableSplitBundle}
  }
}
      `);
            if (((_b = userConfig === null || userConfig === void 0 ? void 0 : userConfig.kepler) === null || _b === void 0 ? void 0 : _b.useSystemJsBundles) === undefined) {
                userConfig = {
                    ...userConfig,
                    kepler: {
                        ...userConfig.kepler,
                        useSystemJsBundles: !userConfig.kepler.disableSplitBundle,
                    },
                };
            }
        }
        // Use lodash merge to deeply merge configs
        // User config takes precedence over default config
        return (0, merge_1.default)({}, ConfigLoader.DEFAULT_CONFIG, userConfig);
    }
    getUseSystemJsBundlesConfig() {
        var _a, _b;
        return (_b = (_a = this.config.kepler) === null || _a === void 0 ? void 0 : _a.useSystemJsBundles) !== null && _b !== void 0 ? _b : true;
    }
    getUseSystemJsBundlesForHeadlessJsConfig() {
        var _a, _b;
        return (_b = (_a = this.config.kepler) === null || _a === void 0 ? void 0 : _a.useSystemJsBundlesForHeadlessJs) !== null && _b !== void 0 ? _b : true;
    }
}
ConfigLoader.DEFAULT_CONFIG = {
    kepler: {
        useSystemJsBundles: true,
        useSystemJsBundlesForHeadlessJs: true,
    },
};
exports.default = ConfigLoader;
