// @amazon-devices/react-native-kepler/Libraries/PlayerSession/api/PlayerSessionTypes.js
"use strict";
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioFocusMode = exports.ErrorCode = exports.PlayerState = exports.PlayerCaptionMode = exports.PlayerMediaOps = exports.PlayerSpeedMode = exports.PlayerShuffleMode = exports.PlayerRepeatMode = exports.PlayerTrackRating = void 0;
/**
 * @file PlayerSessionTypes.ts
 * @description Common types, enums and interfaces used in PlayerSession / PlayerSessionCallback modules
 * @todo Check if some of the enums can be string enums
 */
/**
 * @enum PlayerTrackRating
 * @description An enum to represent Player Track Ratings
 */
var PlayerTrackRating;
(function (PlayerTrackRating) {
    PlayerTrackRating["PLAYER_RATING_NONE"] = "PlayerRatingNone";
    PlayerTrackRating["PLAYER_RATING_1_STARS"] = "PlayerRatingOneStars";
    PlayerTrackRating["PLAYER_RATING_2_STARS"] = "PlayerRatingTwoStars";
    PlayerTrackRating["PLAYER_RATING_3_STARS"] = "PlayerRatingThreeStars";
    PlayerTrackRating["PLAYER_RATING_4_STARS"] = "PlayerRatingFourStars";
    PlayerTrackRating["PLAYER_RATING_5_STARS"] = "PlayerRatingFiveStars";
    PlayerTrackRating["PLAYER_RATING_5_THUMB_DOWN"] = "PlayerRatingFiveThumbDown";
    PlayerTrackRating["PLAYER_RATING_5_THUMB_UP"] = "PlayerRatingFiveThumbUp";
    PlayerTrackRating["PLAYER_RATING_MAX"] = "PlayerRatingMax";
})(PlayerTrackRating = exports.PlayerTrackRating || (exports.PlayerTrackRating = {}));
;
/**
 * @enum PlayerRepeatMode
 * @description An enum to represent Player Repeat mode
 */
var PlayerRepeatMode;
(function (PlayerRepeatMode) {
    PlayerRepeatMode["PLAYER_REPEAT_MODE_NONE"] = "PlayerRepeatModeNone";
    PlayerRepeatMode["PLAYER_REPEAT_MODE_CURRENT"] = "PlayerRepeatModeCurrent";
    PlayerRepeatMode["PLAYER_REPEAT_MODE_PLAYLIST"] = "PlayerRepeatModePlaylist";
    PlayerRepeatMode["PLAYER_REPEAT_MODE_INVALID"] = "PlayerRepeatModeInvalid";
})(PlayerRepeatMode = exports.PlayerRepeatMode || (exports.PlayerRepeatMode = {}));
;
/**
 * @enum PlayerShuffleMode
 * @description An enum to represent Player Shuffle mode
 */
var PlayerShuffleMode;
(function (PlayerShuffleMode) {
    PlayerShuffleMode["PLAYER_SHUFFLE_MODE_NONE"] = "PlayerShuffleModeNone";
    PlayerShuffleMode["PLAYER_SHUFFLE_MODE_PLAYLIST_OFF"] = "PlayerShuffleModePlaylistOff";
    PlayerShuffleMode["PLAYER_SHUFFLE_MODE_PLAYLIST_ON"] = "PlayerShuffleModePlaylistOn";
    PlayerShuffleMode["PLAYER_SHUFFLE_MODE_INVALID"] = "PlayerShuffleModeInvalid";
})(PlayerShuffleMode = exports.PlayerShuffleMode || (exports.PlayerShuffleMode = {}));
;
/**
 * @enum PlayerSpeedMode
 * @description An enum to represent Playback Repeat mode
 */
var PlayerSpeedMode;
(function (PlayerSpeedMode) {
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_4X"] = "PlayerSpeedModeMinus_4X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_3_75X"] = "PlayerSpeedModeMinus_3_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_3_5X"] = "PlayerSpeedModeMinus_3_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_3_25X"] = "PlayerSpeedModeMinus_3_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_3X"] = "PlayerSpeedModeMinus_3X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_2_75X"] = "PlayerSpeedModeMinus_2_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_2_5X"] = "PlayerSpeedModeMinus_2_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_2_25X"] = "PlayerSpeedModeMinus_2_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_2X"] = "PlayerSpeedModeMinus_2X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_1_75X"] = "PlayerSpeedModeMinus_1_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_1_5X"] = "PlayerSpeedModeMinus_1_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_1_25X"] = "PlayerSpeedModeMinus_1_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_0_75X"] = "PlayerSpeedModeMinus_0_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_0_5X"] = "PlayerSpeedModeMinus_0_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_MINUS_0_25X"] = "PlayerSpeedModeMinus_0_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_0_25X"] = "PlayerSpeedMode_0_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_0_5X"] = "PlayerSpeedMode_0_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_0_75X"] = "PlayerSpeedMode_0_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_1X"] = "PlayerSpeedMode_1X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_1_25X"] = "PlayerSpeedMode_1_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_1_5X"] = "PlayerSpeedMode_1_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_1_75X"] = "PlayerSpeedMode_1_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_2X"] = "PlayerSpeedMode_2X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_2_25X"] = "PlayerSpeedMode_2_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_2_5X"] = "PlayerSpeedMode_2_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_2_75X"] = "PlayerSpeedMode_2_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_3X"] = "PlayerSpeedMode_3X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_3_25X"] = "PlayerSpeedMode_3_25X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_3_5X"] = "PlayerSpeedMode_3_5X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_3_75X"] = "PlayerSpeedMode_3_75X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_4X"] = "PlayerSpeedMode_4X";
    PlayerSpeedMode["PLAYER_SPEED_MODE_INVALID"] = "PlayerSpeedModeInvalid";
})(PlayerSpeedMode = exports.PlayerSpeedMode || (exports.PlayerSpeedMode = {}));
;
/**
 * @enum PlayerMediaOps
 * @description An enum to represent Player supported Media operations
 */
var PlayerMediaOps;
(function (PlayerMediaOps) {
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_PLAY"] = 1] = "PLAYER_MEDIA_OP_PLAY";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_PAUSE"] = 2] = "PLAYER_MEDIA_OP_PAUSE";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_PREPARE"] = 4] = "PLAYER_MEDIA_OP_PREPARE";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_REWIND"] = 8] = "PLAYER_MEDIA_OP_REWIND";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SEEK_TO"] = 16] = "PLAYER_MEDIA_OP_SEEK_TO";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SET_VOL"] = 32] = "PLAYER_MEDIA_OP_SET_VOL";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SHUFFLE"] = 64] = "PLAYER_MEDIA_OP_SHUFFLE";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_FF"] = 128] = "PLAYER_MEDIA_OP_FF";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_STOP"] = 256] = "PLAYER_MEDIA_OP_STOP";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SKIP_NEXT"] = 512] = "PLAYER_MEDIA_OP_SKIP_NEXT";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SKIP_PREV"] = 1024] = "PLAYER_MEDIA_OP_SKIP_PREV";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_ADD_PLAY"] = 2048] = "PLAYER_MEDIA_OP_ADD_PLAY";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_PLAY_URI"] = 4096] = "PLAYER_MEDIA_OP_PLAY_URI";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_PREPARE_URI"] = 8192] = "PLAYER_MEDIA_OP_PREPARE_URI";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_INVALID"] = 16384] = "PLAYER_MEDIA_OP_INVALID";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_START_OVER"] = 32768] = "PLAYER_MEDIA_OP_START_OVER";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_REPEAT"] = 65536] = "PLAYER_MEDIA_OP_REPEAT";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_REPEAT_ONE"] = 131072] = "PLAYER_MEDIA_OP_REPEAT_ONE";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_FAVORITE"] = 262144] = "PLAYER_MEDIA_OP_FAVORITE";
    PlayerMediaOps[PlayerMediaOps["PLAYER_MEDIA_OP_SET_CAP_MODE"] = 524288] = "PLAYER_MEDIA_OP_SET_CAP_MODE";
})(PlayerMediaOps = exports.PlayerMediaOps || (exports.PlayerMediaOps = {}));
;
/**
 * @enum PlayerCaptionMode
 * @description An enum to represent Player Caption mode
 */
var PlayerCaptionMode;
(function (PlayerCaptionMode) {
    PlayerCaptionMode[PlayerCaptionMode["PLAYER_CAPTION_MODE_DISABLE"] = 0] = "PLAYER_CAPTION_MODE_DISABLE";
    PlayerCaptionMode[PlayerCaptionMode["PLAYER_CAPTION_MODE_ENABLE"] = 1] = "PLAYER_CAPTION_MODE_ENABLE";
})(PlayerCaptionMode = exports.PlayerCaptionMode || (exports.PlayerCaptionMode = {}));
;
/**
 * @enum PlayerState
 * @description An enum to represent Player State
 */
var PlayerState;
(function (PlayerState) {
    PlayerState["PLAYER_STATE_NONE"] = "PlayerStateNone";
    PlayerState["PLAYER_STATE_BUFFERING"] = "PlayerStateBuffering";
    PlayerState["PLAYER_STATE_CONNECTING"] = "PlayerStateConnecting";
    PlayerState["PLAYER_STATE_FAST_FORWARDING"] = "PlayerStateFastForwarding";
    PlayerState["PLAYER_STATE_PAUSED"] = "PlayerStatePaused";
    PlayerState["PLAYER_STATE_PLAYING"] = "PlayerStatePlaying";
    PlayerState["PLAYER_STATE_REWINDING"] = "PlayerStateRewinding";
    PlayerState["PLAYER_STATE_SKIPPING_TO_NEXT"] = "PlayerStateSkippingToNext";
    PlayerState["PLAYER_STATE_SKIPPING_TO_PREVIOUS"] = "PlayerStateSkippingToPrevious";
    PlayerState["PLAYER_STATE_STOPPED"] = "PlayerStateStopped";
    PlayerState["PLAYER_STATE_SESSION_RELEASE"] = "PlayerStateSessionRelease";
    PlayerState["PLAYER_STATE_ERROR"] = "PlayerStateError";
    PlayerState["PLAYER_STATE_INVALID"] = "PlayerStateInvalid";
})(PlayerState = exports.PlayerState || (exports.PlayerState = {}));
;
/**
 * @enum ErrorCode
 * @description An enum to represent Error codes
 * @todo Check with Media Player team why two different types of error codes are needed
 */
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ERROR_APP_NONE"] = "ErrorAppNone";
    ErrorCode["ERROR_APP_ERROR"] = "ErrorAppError";
    ErrorCode["ERROR_CONCURRENT_STREAM_LIMIT"] = "ErrorConcurrentStreamLimit";
    ErrorCode["ERROR_END_OF_QUEUE"] = "ErrorEndOfQueue";
    ErrorCode["ERROR_PARENTAL_CONTROL_RESTRICTED"] = "ErrorParentalControlRestricted";
    ErrorCode["ERROR_UNKNOWN_ERROR"] = "ErrorUnknownError";
    ErrorCode["ERROR_UNPLAYABLE_BY_AUTHORIZATION"] = "ErrorUnplayableByAuthorization";
    ErrorCode["ERROR_OPERATION_REJECTED_UNINTERRUPTIBLE"] = "ErrorOperationRejectedUninterruptible";
    ErrorCode["ERROR_UNPLAYABLE_BY_REGION"] = "ErrorUnplayableByRegion";
    ErrorCode["ERROR_UNPLAYABLE_BY_SUBSCRIPTION"] = "ErrorUnplayableBySubscription";
    ErrorCode["ERROR_OPERATION_REJECTED_SKIP_LIMIT"] = "ErrorOperationRejectedSkipLimit";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
;
/**
 * @enum AudioFocusMode
 * @description An enum to represent Audio Focus mode
 */
var AudioFocusMode;
(function (AudioFocusMode) {
    AudioFocusMode["AUDIO_FOCUS_NONE"] = "AudioFocusNone";
    AudioFocusMode["AUDIO_FOCUS_PRIMARY"] = "AudioFocusPrimary";
    AudioFocusMode["AUDIO_FOCUS_PAUSED"] = "AudioFocusPaused";
    AudioFocusMode["AUDIO_FOCUS_PAUSED_FOR_UNMIXABLE"] = "AudioFocusPausedOrUnMixable";
    AudioFocusMode["AUDIO_FOCUS_DUCK"] = "AudioFocusDuck";
    AudioFocusMode["AUDIO_FOCUS_INVALID"] = "AudioFocusInvalid";
})(AudioFocusMode = exports.AudioFocusMode || (exports.AudioFocusMode = {}));
;
;
;
;
;
