// @amazon-devices/react-native-kepler/Libraries/WindowManager/WindowManager.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWindowManager = void 0;
const react_1 = require("react");
const RootTag_1 = require("../ReactNative/RootTag");
const WindowManagerSpec_1 = __importDefault(require("./WindowManagerSpec"));
const Dimensions = require('../Utilities/Dimensions').default;
function useWindowManager() {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    const resizeWindow = (targetWidth, targetHeight) => {
        const screenDimensions = Dimensions.get('screen');
        const windowDimensions = Dimensions.get('window');
        let adjustedHeight = targetHeight;
        let adjustedWidth = targetWidth;
        const warnings = [];
        if (targetHeight <= 0 || targetHeight > screenDimensions.height) {
            adjustedHeight = windowDimensions.height;
            warnings.push(`Invalid height. Adjusted from ${targetHeight} to ${windowDimensions.height}`);
        }
        if (targetWidth <= 0 || targetWidth > screenDimensions.width) {
            adjustedWidth = windowDimensions.width;
            warnings.push(`Invalid width. Adjusted from ${targetWidth} to ${windowDimensions.width}`);
        }
        if (warnings.length > 0) {
            console.warn(`Window resize adjustments: ${warnings.join(', ')}`);
        }
        if (adjustedHeight !== windowDimensions.height || adjustedWidth !== windowDimensions.width) {
            WindowManagerSpec_1.default.resizeWindow(adjustedWidth, adjustedHeight, rootTag);
        }
    };
    const setWindowTo = (x, y) => {
        const screenDimensions = Dimensions.get('screen');
        const warnings = [];
        if (x < 0 || x > screenDimensions.width) {
            warnings.push(`Invalid x position.`);
        }
        if (y < 0 || y > screenDimensions.height) {
            warnings.push(`Invalid y position.`);
        }
        if (warnings.length > 0) {
            console.warn(`${warnings.join(', ')}`);
            return;
        }
        WindowManagerSpec_1.default.setWindowPosition(Number(x), Number(y), rootTag);
    };
    return react_1.useMemo(() => ({
        resizeWindowTo: (width, height) => {
            resizeWindow(width, height);
        },
        resizeWindowBy: (deltaWidth, deltaHeight) => {
            const windowDimensions = Dimensions.get('window');
            resizeWindow(windowDimensions.width + deltaWidth, windowDimensions.height + deltaHeight);
        },
        moveTo: (x, y) => {
            setWindowTo(x, y);
        },
        moveBy: (deltaX, deltaY) => {
            const currentPosition = WindowManagerSpec_1.default.getWindowPosition(rootTag);
            setWindowTo(Number(currentPosition["x"]) + Number(deltaX), Number(currentPosition["y"]) + Number(deltaY));
        },
        getPosition: () => {
            return WindowManagerSpec_1.default.getWindowPosition(rootTag);
        }
    }), [rootTag]);
}
exports.useWindowManager = useWindowManager;
