// @amazon-devices/kepler-media-controls/dist/turbo-modules/types/KeplerMediaControlsTypes.js
"use strict";
/*
 * Copyright 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotAvailableError = exports.SeekOutOfRangeError = exports.SpeedOutOfRangeError = exports.NotAllowedError = exports.InvalidStateError = exports.CommandSource = exports.RadioBand = exports.RepeatMode = exports.Action = exports.PlaybackStatus = void 0;
/**
 * Specifies the playback status.
 */
var PlaybackStatus;
(function (PlaybackStatus) {
    /* The media player is currently playing. */
    PlaybackStatus["PLAYING"] = "PLAYING";
    /* Playback is paused. */
    PlaybackStatus["PAUSED"] = "PAUSED";
    /* There is no playback in process. */
    PlaybackStatus["NOT_PLAYING"] = "NOT_PLAYING";
    /* Playback is buffering. */
    PlaybackStatus["BUFFERING"] = "BUFFERING";
})(PlaybackStatus = exports.PlaybackStatus || (exports.PlaybackStatus = {}));
/**
 * Represents the actions that you can invoke using the media control client.
 */
class Action {
    name;
    constructor(name) {
        this.name = name;
    }
    static PLAY = new Action('PLAY');
    static PAUSE = new Action('PAUSE');
    static TOGGLE_PLAY_PAUSE = new Action('TOGGLE_PLAY_PAUSE');
    static STOP = new Action('STOP');
    static START_OVER = new Action('START_OVER');
    static FAST_FORWARD = new Action('FAST_FORWARD');
    static REWIND = new Action('REWIND');
    static SET_PLAYBACK_SPEED = new Action('SET_PLAYBACK_SPEED');
    static SKIP_FORWARD = new Action('SKIP_FORWARD');
    static SKIP_BACKWARD = new Action('SKIP_BACKWARD');
    static SEEK = new Action('SEEK');
    static SET_AUDIO_VOLUME = new Action('SET_AUDIO_VOLUME');
    static SET_AUDIO_TRACK = new Action('SET_AUDIO_TRACK');
    static ENABLE_TEXT_TRACK = new Action('ENABLE_TEXT_TRACK');
    static DISBLE_TEXT_TRACK = new Action('DISABLE_TEXT_TRACK');
    static NEXT = new Action('NEXT');
    static PREVIOUS = new Action('PREVIOUS');
    static ENABLE_SHUFFLE = new Action('ENABLE_SHUFFLE');
    static SET_REPEAT_MODE = new Action('SET_REPEAT_MODE');
    static SET_RATING = new Action('SET_RATING');
    static GET_META_DATA = new Action('GET_META_DATA');
}
exports.Action = Action;
/**
 * Specifies the repeat mode for the media player.
 */
var RepeatMode;
(function (RepeatMode) {
    /* Disable repeat mode. */
    RepeatMode[RepeatMode["OFF"] = 0] = "OFF";
    /* Repeat the current track. */
    RepeatMode[RepeatMode["REPEAT_TRACK"] = 1] = "REPEAT_TRACK";
    /* Repeat the current playlist. */
    RepeatMode[RepeatMode["REPEAT_PLAYLIST"] = 2] = "REPEAT_PLAYLIST";
})(RepeatMode = exports.RepeatMode || (exports.RepeatMode = {}));
/**
 * Specifies the radio band for the media player.
 */
var RadioBand;
(function (RadioBand) {
    /* No radio band selected. */
    RadioBand["NONE"] = "NONE";
    /* The radio band is unknown. */
    RadioBand["UNKNOWN"] = "UNKNOWN";
    /* Specifies the AM band. */
    RadioBand["AM"] = "AM";
    /* Specifies the FM band. */
    RadioBand["FM"] = "FM";
    /* Specifies the DAB band. */
    RadioBand["DAB"] = "DAB";
})(RadioBand = exports.RadioBand || (exports.RadioBand = {}));
/**
 * Specifies the source of the media command.
 */
var CommandSource;
(function (CommandSource) {
    /* Command was initiated by the system in response to the display being turned off. */
    CommandSource["SYSTEM_DISPLAY_OFF"] = "SYSTEM_DISPLAY_OFF";
    /* Command was initiated by the user, e.g., from a voice command */
    CommandSource["USER"] = "USER";
    /* Command was initiated by neither the user nor the system when display is turned off. */
    CommandSource["OTHER"] = "OTHER";
})(CommandSource = exports.CommandSource || (exports.CommandSource = {}));
/**
 * Error occurs when a request cannot be processed due to being in the
 * wrong state to process it.
 */
class InvalidStateError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidStateError';
    }
}
exports.InvalidStateError = InvalidStateError;
/**
 * Error occurs when a request cannot be handled.
 */
class NotAllowedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'NotAllowedError';
    }
}
exports.NotAllowedError = NotAllowedError;
/**
 * Error occurs when the requested playback speed is not supported.
 */
class SpeedOutOfRangeError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SpeedOutOfRangeError';
    }
}
exports.SpeedOutOfRangeError = SpeedOutOfRangeError;
/**
 * Error occurs when the requested seek position is out of the advertised range.
 */
class SeekOutOfRangeError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SeekOutOfRangeError';
    }
}
exports.SeekOutOfRangeError = SeekOutOfRangeError;
/**
 * Error occurs when requested data is not available.
 */
class NotAvailableError extends Error {
    constructor(message) {
        super(message);
        this.name = 'NotAvailableError';
    }
}
exports.NotAvailableError = NotAvailableError;
