// @amazon-devices/react-native-kepler/Libraries/Navigable/NavigableComponentSpec.js
"use strict";
/**
 * Copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const registerGeneratedViewConfig_1 = __importDefault(require("../Utilities/registerGeneratedViewConfig"));
const NativeComponent = 'NativeNavigable';
const viewConfig = {
    validAttributes: {
        blockEdges: true,
        defaultFocusItem: true,
        loopFocus: true,
        navigationStrategy: true,
        retainLastFocus: true,
    },
    bubblingEventTypes: {
        change: {
            phasedRegistrationNames: {
                bubbled: 'onChange',
                captured: 'onChangeCapture'
            }
        }
    }
};
registerGeneratedViewConfig_1.default(NativeComponent, {
    uiViewClassName: NativeComponent,
    ...viewConfig
});
exports.default = NativeComponent;
