// @amazon-devices/kepler-module-manifest-builder/dist/test/utils/compatibility-utils.test.js
"use strict";
// compatibility-utils.test.ts
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const compatibility_utils_1 = require("../../src/utils/compatibility-utils");
const logger_1 = require("../../src/utils/logger");
// Mock all fs operations
jest.mock('fs', () => ({
    existsSync: jest.fn(),
    readFileSync: jest.fn(),
}));
// Mock path operations
jest.mock('path', () => ({
    join: jest.fn((...args) => args.join('/')),
}));
describe('compatibility-utils', () => {
    // Mock console operations
    const mockLogger = {
        error: jest.spyOn(logger_1.logger, 'error').mockImplementation(),
    };
    beforeEach(() => {
        // Clear all mocks before each test
        jest.clearAllMocks();
    });
    describe('getKeplerNodeModulesData', () => {
        const mockLibraryPath = '/path/to/library';
        const mockCompatibilityPath = '/path/to/library/kepler-compatibility.json';
        it('should return null when compatibility file does not exist', () => {
            fs.existsSync.mockReturnValue(false);
            const result = (0, compatibility_utils_1.getKeplerNodeModulesData)(mockLibraryPath);
            expect(result).toBeNull();
            expect(fs.existsSync).toHaveBeenCalledWith(mockCompatibilityPath);
        });
        it('should return correct module info for valid compatibility file', () => {
            const mockCompatibility = {
                versions: {
                    '2.0.0': 'com.amazon.kepler.test@ITest_0',
                    '2.1.0': 'com.amazon.kepler.test@ITest_1',
                    '2.2.0': 'com.amazon.kepler.test@ITest_2',
                },
            };
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify(mockCompatibility));
            const result = (0, compatibility_utils_1.getKeplerNodeModulesData)(mockLibraryPath);
            expect(result).toEqual({
                needsModule: 'com.amazon.kepler.test@ITest_2',
                versionMap: mockCompatibility.versions,
                privilegeMap: {},
            });
            expect(fs.readFileSync).toHaveBeenCalledWith(mockCompatibilityPath, 'utf-8');
        });
        it('should handle invalid JSON in compatibility file', () => {
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue('invalid json');
            const result = (0, compatibility_utils_1.getKeplerNodeModulesData)(mockLibraryPath);
            expect(result).toBeNull();
            expect(mockLogger.error).toHaveBeenCalledWith('Failed to read /path/to/library/kepler-compatibility.json:', expect.any(Error));
        });
        it('should handle empty compatibility file', () => {
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify({}));
            const result = (0, compatibility_utils_1.getKeplerNodeModulesData)(mockLibraryPath);
            expect(result).toBeNull();
            expect(mockLogger.error).toHaveBeenCalledWith('Error processing compatibility mapping:', expect.any(Error));
        });
    });
    describe('parseCompatibilityInfo', () => {
        it('should parse valid compatibility info', () => {
            const mockContent = JSON.stringify({
                versions: {
                    '2.0.0': 'com.amazon.kepler.test@ITest_0',
                    '2.1.0': 'com.amazon.kepler.test@ITest_1',
                },
            });
            const result = (0, compatibility_utils_1.parseCompatibilityInfo)(mockContent, '@amazon-devices/react-native-gesture-handler');
            expect(result).toEqual({
                includesModules: [
                    'com.amazon.kepler.test@ITest_0',
                    'com.amazon.kepler.test@ITest_1',
                ],
                versionMap: {
                    '2.0.0': 'com.amazon.kepler.test@ITest_0',
                    '2.1.0': 'com.amazon.kepler.test@ITest_1',
                },
            });
        });
        it('should handle invalid JSON content', () => {
            const result = (0, compatibility_utils_1.parseCompatibilityInfo)('invalid json', '@amazon-devices/react-native-gesture-handler');
            expect(result).toBeNull();
            expect(mockLogger.error).toHaveBeenCalledWith('Error parsing compatibility info for @amazon-devices/react-native-gesture-handler:', expect.any(Error));
        });
        it('should handle empty versions', () => {
            const result = (0, compatibility_utils_1.parseCompatibilityInfo)('{"versions": {}}', '@amazon-devices/react-native-gesture-handler');
            expect(result).toEqual({
                includesModules: [],
                versionMap: {},
            });
        });
        it('should handle empty content', () => {
            const result = (0, compatibility_utils_1.parseCompatibilityInfo)('{}', '@amazon-devices/react-native-gesture-handler');
            expect(result).toEqual({
                includesModules: [],
                versionMap: {},
            });
        });
        it('should handle string arrays in kepler-compatibility input', () => {
            const mockCompatibility = {
                versions: {
                    '2.0.0': [
                        '/com.amazon.kepler.experimental.features_2025_02@IFeature',
                        '/com.amazon.kepler.security@ISecurityModule',
                    ],
                    '2.1.0': [
                        '/com.amazon.kepler.experimental.features_2025_03@IFeature',
                        '/com.amazon.kepler.security@ISecurityModule_2',
                    ],
                },
            };
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify(mockCompatibility));
            const result = (0, compatibility_utils_1.getKeplerNodeModulesData)('/path/to/library');
            expect(result).toEqual({
                needsModule: [
                    '/com.amazon.kepler.experimental.features_2025_03@IFeature',
                    '/com.amazon.kepler.security@ISecurityModule_2',
                ],
                versionMap: mockCompatibility.versions,
                privilegeMap: {},
            });
        });
    });
});
