// @amazon-devices/kepler-module-manifest-builder/dist/test/utils/file-utils.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const fs = __importStar(require("fs"));
const file_utils_1 = require("../../src/utils/file-utils");
const logger_1 = require("../../src/utils/logger");
// Mock all external dependencies
jest.mock('child_process', () => ({
    execSync: jest.fn(),
}));
jest.mock('fs', () => ({
    existsSync: jest.fn(),
    readFileSync: jest.fn(),
    readdirSync: jest.fn(),
}));
jest.mock('path', () => ({
    join: jest.fn((...args) => args.join('/')),
}));
describe('file-utils', () => {
    // Setup common mocks
    const mockLogger = {
        error: jest.spyOn(logger_1.logger, 'error').mockImplementation(),
    };
    beforeEach(() => {
        jest.clearAllMocks();
    });
    describe('extractTarball', () => {
        const mockConfig = {
            npmManager: 'npm',
            packageName: '@amazon-devices/react-native-gesture-handler',
            version: '2.0.0',
            tempDir: '/tmp/test',
        };
        it('should successfully extract a package tarball', () => {
            // Given
            child_process_1.execSync
                .mockReturnValueOnce('') // npm pack
                .mockReturnValueOnce(''); // tar extract
            // Mock finding the tarball
            fs.readdirSync.mockReturnValue([
                'amzn__react-native-gesture-handler-2.0.0.tgz',
            ]);
            // When
            (0, file_utils_1.extractTarball)(mockConfig.npmManager, mockConfig.packageName, mockConfig.version, mockConfig.tempDir);
            // Then
            expect(child_process_1.execSync).toHaveBeenNthCalledWith(1, `npm pack @amazon-devices/react-native-gesture-handler@2.0.0 --pack-destination /tmp/test`, expect.any(Object));
            expect(child_process_1.execSync).toHaveBeenNthCalledWith(2, expect.stringContaining('tar -xzf'), expect.any(Object));
        });
        it('should throw error when npm pack fails', () => {
            // Given
            child_process_1.execSync.mockImplementationOnce(() => {
                throw new Error('npm pack failed');
            });
            // Then
            expect(() => (0, file_utils_1.extractTarball)(mockConfig.npmManager, mockConfig.packageName, mockConfig.version, mockConfig.tempDir)).toThrow('Failed to extract NPM package');
        });
        it('should support different package managers', () => {
            // Given
            child_process_1.execSync
                .mockReturnValueOnce('')
                .mockReturnValueOnce('');
            // When
            (0, file_utils_1.extractTarball)('yarn', mockConfig.packageName, mockConfig.version, mockConfig.tempDir);
            // Then
            expect(child_process_1.execSync).toHaveBeenCalledWith(expect.stringContaining('yarn pack'), expect.any(Object));
        });
    });
    describe('parseJsonFile', () => {
        const mockPackageJsonPath = '/path/to/package.json';
        it('should successfully read and parse package.json', () => {
            // Given
            const mockPackageJson = {
                name: '@amazon-devices/react-native-gesture-handler',
                version: '2.0.0',
            };
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify(mockPackageJson));
            // When
            const result = (0, file_utils_1.parseJsonFile)(mockPackageJsonPath);
            // Then
            expect(result).toEqual(mockPackageJson);
            expect(fs.readFileSync).toHaveBeenCalledWith(mockPackageJsonPath, 'utf-8');
        });
        it('should throw error when package.json does not exist', () => {
            // Given
            fs.existsSync.mockReturnValue(false);
            // Then
            expect(() => (0, file_utils_1.parseJsonFile)(mockPackageJsonPath)).toThrow(`Provided file not found: ${mockPackageJsonPath}`);
            expect(mockLogger.error).toHaveBeenCalled();
        });
        it('should throw error when package.json contains invalid JSON', () => {
            // Given
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue('invalid json');
            // Then
            expect(() => (0, file_utils_1.parseJsonFile)(mockPackageJsonPath)).toThrow();
            expect(mockLogger.error).toHaveBeenCalled();
        });
    });
    describe('getCompatibilityForLibraryName', () => {
        const mockConfig = {
            nodeModulesPath: '/path/to/node_modules',
            libraryName: '@amazon-devices/react-native-gesture-handler',
        };
        it('should return compatibility info when file exists', () => {
            // Given
            const mockCompatibility = {
                '2.0.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_0',
                '2.1.0': '/com.amazon.kepler.gesture_handler_2@IGestureHandler_1',
            };
            fs.existsSync.mockReturnValue(true);
            fs.readFileSync.mockReturnValue(JSON.stringify(mockCompatibility));
            // When
            const result = (0, file_utils_1.getCompatibilityForLibraryName)(mockConfig.nodeModulesPath, mockConfig.libraryName);
            // Then
            expect(result).toBe(JSON.stringify(mockCompatibility));
            expect(fs.readFileSync).toHaveBeenCalledWith(expect.stringContaining('kepler-compatibility.json'), 'utf-8');
        });
        it('should return null when compatibility file does not exist', () => {
            // Given
            fs.existsSync.mockReturnValue(false);
            // When
            const result = (0, file_utils_1.getCompatibilityForLibraryName)(mockConfig.nodeModulesPath, mockConfig.libraryName);
            // Then
            expect(result).toBeNull();
            expect(fs.readFileSync).not.toHaveBeenCalled();
        });
    });
});
