// @amazon-devices/kepler-module-manifest-builder/dist/src/utils/compatibility-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCompatibilityInfo = exports.processCompatibilityMapping = exports.getPrivilegesFromManifest = exports.getKeplerNodeModulesData = void 0;
const fs = __importStar(require("fs"));
const lodash_1 = require("lodash");
const path = __importStar(require("path"));
const semver = __importStar(require("semver"));
const toml = __importStar(require("toml"));
const file_utils_1 = require("./file-utils");
const logger_1 = require("./logger");
function getKeplerNodeModulesData(jsRootPath) {
    try {
        const keplerCompatibility = (0, file_utils_1.parseJsonFile)(path.join(jsRootPath, 'kepler-compatibility.json'));
        if (!keplerCompatibility) {
            logger_1.logger.warn(`File kepler-compatibility.json is invalid or missing from dir: ${jsRootPath}:`);
            return null;
        }
        return processCompatibilityMapping(keplerCompatibility);
    }
    catch (error) {
        if (error instanceof Error &&
            error.message.startsWith('Provided file not found')) {
            logger_1.logger.info(`No kepler-compatibility.json found in dir: ${jsRootPath}`);
            return null;
        }
        logger_1.logger.warn(`Error parsing kepler-compatibility.json in dir: ${jsRootPath}:`, error);
        return null;
    }
}
exports.getKeplerNodeModulesData = getKeplerNodeModulesData;
function getPrivilegesFromManifest(manifestToml) {
    try {
        const manifestTomlContent = fs.readFileSync(manifestToml, {
            encoding: 'utf8',
        });
        const parsed = toml.parse(manifestTomlContent);
        const privileges = [];
        if (parsed.needs?.privilege) {
            parsed.needs.privilege.forEach((p) => {
                privileges.push(p.id);
            });
        }
        return privileges;
    }
    catch (error) {
        throw new Error(`Error parsing toml in file ${manifestToml}: ${error}`);
    }
}
exports.getPrivilegesFromManifest = getPrivilegesFromManifest;
/**
 * Processes compatibility mapping to determine module information
 * @param versionMap The parsed kepler-compatibility version mapping
 * @returns Module information or null if invalid
 */
function processCompatibilityMapping(keplerCompatibility) {
    try {
        const { privileges, versions } = keplerCompatibility;
        const versionList = Object.keys(versions);
        const highestVersion = (0, lodash_1.isEmpty)(versionList)
            ? ''
            : versionList.reduce((highest, current) => semver.gt(current, highest) ? current : highest, versionList[0]);
        return {
            needsModule: versions[highestVersion],
            versionMap: versions,
            privilegeMap: privileges || {},
        };
    }
    catch (error) {
        logger_1.logger.error(`Error processing compatibility mapping:`, error);
        return null;
    }
}
exports.processCompatibilityMapping = processCompatibilityMapping;
/**
 * Parses the compatibility info for a library, using the format:
 * {
 *      "2.0.0" : "com.amazon.kepler.test_library_2@ILibrary_0",
 *      "2.1.0" : "com.amazon.kepler.test_library_2@ILibrary_1",
 *      "2.2.0" : "com.amazon.kepler.test_library_2@ILibrary_2",
 * }
 * @param content
 * @param libraryName
 * @returns
 */
function parseCompatibilityInfo(content, libraryName) {
    try {
        const jsonData = JSON.parse(content);
        const versionMap = jsonData.versions || {};
        const includesModules = Object.values(versionMap).reduce((acc, moduleId) => {
            if (Array.isArray(moduleId)) {
                acc.push(...moduleId);
            }
            else {
                acc.push(moduleId);
            }
            return acc;
        }, []);
        return {
            includesModules,
            versionMap,
        };
    }
    catch (error) {
        logger_1.logger.error(`Error parsing compatibility info for ${libraryName}:`, error);
        return null;
    }
}
exports.parseCompatibilityInfo = parseCompatibilityInfo;
