// @amazon-devices/kepler-module-manifest-builder/dist/src/utils/version-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHighestVersionInMap = exports.setNeedsForLibrary = exports.isExactVersion = exports.isVersionEqual = exports.isVersionLargerThan = exports.getIntersectionRange = exports.isVersionInRange = void 0;
const semver = __importStar(require("semver"));
const logger_1 = require("./logger");
/**
 * Checks if a version is included in a given NPM version range.
 * @param version - The version to check (e.g., "2.1.0")
 * @param range - The NPM version range (e.g., "^2.0.0", "~2.13.0")
 * @returns True if the version is included in the range, false otherwise
 */
function isVersionInRange(version, range) {
    try {
        return semver.satisfies(version, range);
    }
    catch (error) {
        logger_1.logger.error(`Error checking version: ${error}`);
        return false;
    }
}
exports.isVersionInRange = isVersionInRange;
/**
 * Finds the intersection of two NPM version ranges.
 * @param range1 - The first NPM version range (e.g., "^2.0.0")
 * @param range2 - The second NPM version range (e.g., "~2.13.0")
 * @returns The intersection of the two ranges (e.g., "~2.13.0")
 */
function getIntersectionRange(range1, range2) {
    if (!semver.validRange(range1)) {
        throw new Error(`Range ${range1} is invalid.`);
    }
    if (!semver.validRange(range2)) {
        throw new Error(`Range ${range2} is invalid.`);
    }
    if (!semver.intersects(range1, range2)) {
        throw new Error(`Ranges ${range1} and ${range2} do not intersect.`);
    }
    // ASSERT: From here, we know that range1 and range2 are valid intersections
    // Handle exact versions first
    if (semver.valid(range1)) {
        return range1;
    }
    if (semver.valid(range2)) {
        return range2;
    }
    // Handle less-than ranges
    if (range1.startsWith('<') && range2.startsWith('<')) {
        // For two less-than ranges, return the one with the smaller version
        const version1 = semver.clean(range1.substring(1)) || '';
        const version2 = semver.clean(range2.substring(1)) || '';
        // If versions are equal, prefer strict less-than over less-than-equal
        if (version1 === version2) {
            return range1.startsWith('<=') ? range2 : range1;
        }
        return semver.lt(version1, version2) ? range1 : range2;
    }
    if (range1.startsWith('<')) {
        return range1;
    }
    if (range2.startsWith('<')) {
        return range2;
    }
    const min1 = semver.minVersion(range1) || '0.0.0';
    const min2 = semver.minVersion(range2) || '0.0.0';
    // Handle ~ vs ^ ranges
    if (range1.startsWith('~') && range2.startsWith('^')) {
        return range1;
    }
    if (range2.startsWith('~') && range1.startsWith('^')) {
        return range2;
    }
    // If both ranges are the same type (^ or ~), return the one with the higher minimum version
    return semver.gt(min2, min1) ? range2 : range1;
}
exports.getIntersectionRange = getIntersectionRange;
/**
 * Checks if a version is larger than some other version.
 * @param version - The version to check (e.g., "2.1.0")
 * @param version2 - The other version to check (e.g., "2.4.0")
 * @returns True if the version is greater than version2, false otherwise
 */
function isVersionLargerThan(version, version2) {
    try {
        return semver.gt(version, version2);
    }
    catch (error) {
        logger_1.logger.error(`Error checking version: ${error}`);
        return false;
    }
}
exports.isVersionLargerThan = isVersionLargerThan;
/**
 * Checks if a two versions are equal.
 * @param version - The version to check (e.g., "2.1.0")
 * @param version2 - The other version to check (e.g., "2.4.0")
 * @returns True if the version is equal to version2, false otherwise
 */
function isVersionEqual(version, version2) {
    try {
        return semver.eq(version, version2);
    }
    catch (error) {
        logger_1.logger.error(`Error checking version: ${error}`);
        return false;
    }
}
exports.isVersionEqual = isVersionEqual;
/**
 * Check if a version is an exact version (not a range).
 * @param versionOrRange
 * @returns True if the version is an exact version, false otherwise
 */
function isExactVersion(versionOrRange) {
    return semver.valid(versionOrRange) !== null;
}
exports.isExactVersion = isExactVersion;
function setNeedsForLibrary(libraryName, manifestMap, appDefinedNeed) {
    let result = {
        version: appDefinedNeed,
        source: libraryName, // Default to self as source
    };
    // Check each library's requirements
    Object.entries(manifestMap).forEach(([libName, libInfo]) => {
        if (!libInfo.libraryDependencies) {
            return;
        }
        // Find requirements that target our library
        libInfo.libraryDependencies.forEach((req) => {
            if (req.libraryName === libraryName && // Requirement is for our target library
                isVersionLargerThan(req.needsVersion, result.version)) {
                result = {
                    version: req.needsVersion,
                    source: libName, // The library that has this requirement
                };
                logger_1.logger.info(`Found higher needs version ${req.needsVersion} for ${libraryName} from ${libName}`);
            }
        });
    });
    return result;
}
exports.setNeedsForLibrary = setNeedsForLibrary;
function getHighestVersionInMap(libraryName, versionMap) {
    return Object.keys(versionMap).reduce((highestKnown, currentVersion) => isVersionLargerThan(highestKnown, currentVersion)
        ? highestKnown
        : currentVersion, '0.0.0');
}
exports.getHighestVersionInMap = getHighestVersionInMap;
