// @amazon-devices/react-native-kepler/Libraries/SplashScreenManager/SplashScreenManager.js
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplashScreenManager = exports.useHideSplashScreenCallback = exports.usePreventHideSplashScreen = void 0;
const react_1 = require("react");
const NativeSplashScreenManager_1 = __importDefault(require("./NativeSplashScreenManager"));
const RootTag_1 = require("../ReactNative/RootTag");
class SplashScreenManagerImpl {
    /**
     * @brief keeps the splash screen visible until hide is called
     *
     * @param moduleName: name for the screen, the value should be identical with screenKey
     * that users register in app's main.cpp create() method
     *
     * @discussion users of this API are recommended to call it in global scope. Calling this
     * in a React component or hook could lead to bugs if it causes this API to be called
     * after the splash screen is already dismissed.
     */
    async preventAutoHideAsync(moduleName) {
        return NativeSplashScreenManager_1.default.preventAutoHide(moduleName);
    }
    /**
     * @brief hides the splash screen
     *
     * @param moduleName: name for the screen, the value should be identical with screenKey
     * that users register in app's main.cpp create() method
     *
     * @description users should be careful that this API is only called after the app's
     * content is ready to be displayed.
     */
    async hideAsync(moduleName) {
        return NativeSplashScreenManager_1.default.hide(moduleName);
    }
}
/**
 * @brief a custom hook to keep the splash screen visible until useHideSplashScreenCallback() is called.
 */
const usePreventHideSplashScreen = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    const ref = react_1.useRef(true);
    if (ref.current) {
        ref.current = false;
        NativeSplashScreenManager_1.default.preventAutoHideV2(rootTag);
    }
};
exports.usePreventHideSplashScreen = usePreventHideSplashScreen;
/**
 * @brief a custom hook to manually dismiss the splash screen.
 *
 * @returns a callback function to invoke when splash screen is ready to be dismissed.
 */
const useHideSplashScreenCallback = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    return react_1.useCallback(() => NativeSplashScreenManager_1.default.hideV2(rootTag), [rootTag]);
};
exports.useHideSplashScreenCallback = useHideSplashScreenCallback;
exports.SplashScreenManager = new SplashScreenManagerImpl();
