// @amazon-devices/react-native-kepler/Libraries/Components/Switch/SwitchSpec.js
"use strict";
/**
 * Copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 *
 * @format
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// Below util file is removed in RN 0.67, need to tweak when fix is ready and merging to 67 branches
const registerGeneratedViewConfig_1 = __importDefault(require("../../Utilities/registerGeneratedViewConfig"));
const NativeSwitchComponent = 'NativeSwitch';
const viewConfig = {
    value: true,
    disabled: true,
    thumbColor: true,
    trackColorTrue: true,
    trackColorFalse: true,
    bubblingEventTypes: {
        change: {
            phasedRegistrationNames: {
                bubbled: 'onChange',
                captured: 'onChangeCapture'
            }
        }
    }
};
registerGeneratedViewConfig_1.default(NativeSwitchComponent, {
    uiViewClassName: 'NativeSwitch',
    validAttributes: viewConfig
});
exports.default = NativeSwitchComponent;
