// @amzn/react-native-kepler/Libraries/PlayerSession/PlayerSession.js
"use strict";
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerSession = exports.PlayerSessionClient = void 0;
const Logger_1 = require("../Logger/Logger");
const PlayerSessionSpec_1 = __importDefault(require("./spec/PlayerSessionSpec"));
const PLAYER_SESSION_CLIENT_TAG = 'PlayerSession';
class PlayerSessionClient {
    constructor() {
        this.playerSessionEventSubscriptions_ = [];
    }
    create() {
        Logger_1.KPLOG().debug(`${PLAYER_SESSION_CLIENT_TAG}: Creating Player Session..`);
        return PlayerSessionSpec_1.default.create();
    }
    destroy() {
        for (const subscription of this.playerSessionEventSubscriptions_) {
            subscription.remove();
        }
        this.playerSessionEventSubscriptions_ = [];
        PlayerSessionSpec_1.default.destroy();
        Logger_1.KPLOG().debug(`${PLAYER_SESSION_CLIENT_TAG}: Player Session is destroyed`);
    }
    setPlayerSessionCallback(playerSessionCallback) {
        // Iterate over all the call backs and bind to player session turbo module
        for (const property in playerSessionCallback) {
            if (typeof PlayerSessionSpec_1.default.addEventListener !== "undefined") {
                this.playerSessionEventSubscriptions_.push(PlayerSessionSpec_1.default.addEventListener(property, playerSessionCallback[property]));
            }
            else {
                PlayerSessionSpec_1.default.addListener(property, playerSessionCallback[property]);
            }
        }
        Logger_1.KPLOG().debug(`${PLAYER_SESSION_CLIENT_TAG}: Callbacks are set in player session turbo module`);
    }
    setPlaybackState(playbackState) {
        Logger_1.KPLOG().debug(`${PLAYER_SESSION_CLIENT_TAG}: Sending Playback state to native ..`);
        return PlayerSessionSpec_1.default.setPlaybackState(playbackState);
    }
}
exports.PlayerSessionClient = PlayerSessionClient;
exports.PlayerSession = new PlayerSessionClient();
