// @amzn/react-native-kepler/Libraries/Components/ToastKepler/NotificationManagerUtils.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToastKind = exports.AppLaunchTarget = void 0;
/*
 * Types and classes below are taken fron NotificationManager implementation to
 * ensure type safety
 */
/**
 * Type of launch target used to describe an interactive app.
 * This target will be launched when a notification is
 * interacted with.
 */
class AppLaunchTarget {
    constructor(appUri, title, iconUri) {
        this.appUri = appUri;
        this.title = title;
        this.iconUri = iconUri;
        return this;
    }
    /**
     * @param visitor Implementation of a visitor
     * implementing functions that perform different tasks
     * depending on the kind of launch target
     */
    visit(visitor) {
        visitor.visitApp(this);
    }
}
exports.AppLaunchTarget = AppLaunchTarget;
class ToastKind {
    constructor() {
        this.kindDescriptor = 5;
    }
    /**
     * Used to visit the 'kind' in a type safe way
     * @param visitor Implementation of a visitor
     * implementing functions that perform different tasks
     * depending on the type of 'kind'
     */
    visit(visitor) {
        visitor.visitToast(this);
    }
}
exports.ToastKind = ToastKind;
