// @amazon-devices/kepler-cli-platform/dist/options.js
"use strict";
/*
 * Copyright (c) 2023 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.systemBundleOptionsArray = exports.hermesKeplerOptionsArray = exports.bundleKeplerOptionsArray = exports.buildKeplerOptionsArray = void 0;
// Hermes options
const hermesKeplerOptions = {
    out: {
        name: '--out <output-path>',
        description: 'Path at which the compiled Hermes bundle would be generated. Should be path to a file name for the bundle. Intermediate directories in the path would be created if they don\'t exist',
    },
    bundlePath: {
        name: '--bundle-path <bundle_path>',
        description: 'Path to the bundle file to be compiled',
    },
    hermesFlags: {
        name: '--hermes-flags <opts>',
        description: 'Options to forward to Hermes compiler for processing the bundle file. Options should be wrapped in quotes and comma separated to prevent parsing errors with the RN CLI command. For RN 0.71 and earlier, also escape ("\\") the first option',
        default: '-O,-w',
    },
};
// Bundle options
const bundleKeplerOptions = {
    buildType: {
        name: '--build-type <variant>',
        description: 'Set the build variant, either "Debug" or "Release"',
    },
    /** @deprecated Use useSystemJsBundles instead */
    splitBundle: {
        name: '--split-bundle <boolean>',
        description: '[DEPRECATED] Please use --use-system-js-bundles instead. This option will be removed in next release.',
        parse: (val) => val !== 'false',
    },
    /** Whether to split system library dependencies out of the JS bundle, take precedence over splitBundle option */
    useSystemJsBundles: {
        name: '--use-system-js-bundles <boolean>',
        description: 'Whether to split system library dependencies out of the JS bundle. When this flag is not' +
            ' specified, kepler will look at "useSystemJsBundles" property under "kepler" property of project' +
            ' package.json. If that property is also undefined, kepler will not split the bundle and proceed as usual.' +
            ' This behaviour may change in future versions to split bundle by default.',
        parse: (val) => val !== 'false',
    },
    maxWorkers: {
        name: '--max-workers <number>',
        description: 'Specifies the maximum number of workers the worker-pool will spawn for transforming files. This defaults to the number of the cores available on your machine.',
        parse: (workers) => Number(workers),
    },
    resetCache: {
        name: '--reset-cache',
        description: 'Removes cached files',
        default: false,
    },
    config: {
        name: '--config <string>',
        description: 'Path to the CLI configuration file',
    },
    verbose: {
        name: '--verbose',
        description: 'Enables verbose logging',
        default: false,
    },
    hermesFlags: hermesKeplerOptions.hermesFlags, // Reuse from bundle options
};
// Build options (superset of bundle kepler)
const buildKeplerOptions = {
    // Include all bundle options
    ...bundleKeplerOptions,
    ...hermesKeplerOptions,
    // Add build-specific options
    includeJsBundle: {
        name: '--include-js-bundle <boolean>',
        description: 'Whether to include the js bundle in the vpkg archive or not. This enables vpkgs submitted to the App Store to be upgraded/optimized automatically. Default is true.',
        parse: (val) => val !== 'false',
    },
    skipBundling: {
        name: '--skip-bundling',
        description: 'Skip creating a compiled Hermes bundle',
        default: false,
    },
    target: {
        name: '--target <target_name>',
        description: 'Run "kepler build --help" to find out the set of valid targets',
    },
    buildNumber: {
        name: '--build-number <number>',
        description: 'Set the build number for the current build',
    },
    buildVersion: {
        name: '--build-version <number>',
        description: 'Set the build version for the current build',
    },
    checkManifest: {
        name: '--check-manifest <boolean>',
        description: 'Validates app manifest file and fails command if invalid. Default is true',
        parse: (val) => val !== 'false',
    },
    additionalNativeOptions: {
        name: '--additional-native-options <opts>',
        description: 'Options to forward to the underlying native build system for Kepler CLI. Options should be wrapped' +
            ' in double quotes and comma separated to prevent parsing errors with the RN CLI command. Eg.' +
            ' "--opt1,--opt2=val"',
    },
    // Add remaining CommandLineArgs options that weren't omitted
    resetGlobalCache: {
        name: '--reset-global-cache',
        description: 'Removes global cache',
        default: false,
    },
    transformer: {
        name: '--transformer <string>',
        description: 'Specify a custom transformer to be used',
    },
    bundleEncoding: {
        name: '--bundle-encoding <string>',
        description: 'Encoding the bundle should be written in (https://nodejs.org/api/buffer.html#buffer_buffer).',
        default: 'utf8',
    },
    sourcemapSourcesRoot: {
        name: '--sourcemap-sources-root <string>',
        description: 'Path to make sourcemap\'s sources entries relative to, ex. /root/dir',
    },
    sourcemapUseAbsolutePath: {
        name: '--sourcemap-use-absolute-path',
        description: 'Report SourceMapURL using its full path',
        default: false,
    },
};
const SystemBundleOptions = {
    hermesFlags: hermesKeplerOptions.hermesFlags, // Reuse from bundle options
};
exports.buildKeplerOptionsArray = Object.values(buildKeplerOptions);
exports.bundleKeplerOptionsArray = Object.values(bundleKeplerOptions);
exports.hermesKeplerOptionsArray = Object.values(hermesKeplerOptions);
exports.systemBundleOptionsArray = Object.values(SystemBundleOptions);
