// @amzn/keplermediadescriptor/dist/turbo-modules/LevelValues.js
"use strict";
/*
 * Copyright 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.H263Level60 = exports.H263Level50 = exports.H263Level40 = exports.H263Level30 = exports.H263Level20 = exports.H263Level10 = exports.H263LevelUnknown = exports.HEVCHighTierLevel62 = exports.HEVCMainTierLevel62 = exports.HEVCHighTierLevel61 = exports.HEVCMainTierLevel61 = exports.HEVCHighTierLevel6 = exports.HEVCMainTierLevel6 = exports.HEVCHighTierLevel52 = exports.HEVCMainTierLevel52 = exports.HEVCHighTierLevel51 = exports.HEVCMainTierLevel51 = exports.HEVCHighTierLevel5 = exports.HEVCMainTierLevel5 = exports.HEVCHighTierLevel41 = exports.HEVCMainTierLevel41 = exports.HEVCHighTierLevel4 = exports.HEVCMainTierLevel4 = exports.HEVCMainTierLevel31 = exports.HEVCMainTierLevel3 = exports.HEVCMainTierLevel21 = exports.HEVCMainTierLevel2 = exports.HEVCMainTierLevel1 = exports.HEVCLevelUnknown = exports.AVCLevel62 = exports.AVCLevel61 = exports.AVCLevel6 = exports.AVCLevel52 = exports.AVCLevel51 = exports.AVCLevel5 = exports.AVCLevel42 = exports.AVCLevel41 = exports.AVCLevel4 = exports.AVCLevel32 = exports.AVCLevel31 = exports.AVCLevel3 = exports.AVCLevel22 = exports.AVCLevel21 = exports.AVCLevel2 = exports.AVCLevel13 = exports.AVCLevel12 = exports.AVCLevel11 = exports.AVCLevel1b = exports.AVCLevel1 = exports.AVCLevelUnknown = void 0;
exports.AV1Level42 = exports.AV1Level41 = exports.AV1Level4 = exports.AV1Level33 = exports.AV1Level32 = exports.AV1Level31 = exports.AV1Level3 = exports.AV1Level23 = exports.AV1Level22 = exports.AV1Level21 = exports.AV1Level2 = exports.AV1LevelUnknown = exports.VP8LevelVersion3 = exports.VP8LevelVersion2 = exports.VP8LevelVersion1 = exports.VP8LevelVersion0 = exports.VP8LevelUnknown = exports.MPEG2LevelHP = exports.MPEG2LevelHL = exports.MPEG2LevelH14 = exports.MPEG2LevelML = exports.MPEG2LevelLL = exports.MPEG2LevelUnknown = exports.MPEG4Level6 = exports.MPEG4Level5 = exports.MPEG4Level4a = exports.MPEG4Level4 = exports.MPEG4Level3b = exports.MPEG4Level3 = exports.MPEG4Level2 = exports.MPEG4Level1 = exports.MPEG4Level0b = exports.MPEG4Level0 = exports.MPEG4LevelUnknown = exports.VP9Level62 = exports.VP9Level61 = exports.VP9Level6 = exports.VP9Level52 = exports.VP9Level51 = exports.VP9Level5 = exports.VP9Level41 = exports.VP9Level4 = exports.VP9Level31 = exports.VP9Level3 = exports.VP9Level21 = exports.VP9Level2 = exports.VP9Level11 = exports.VP9Level1 = exports.VP9LevelUnknown = exports.H263Level70 = void 0;
exports.DolbyVisionLevelUhd60 = exports.DolbyVisionLevelUhd48 = exports.DolbyVisionLevelUhd30 = exports.DolbyVisionLevelUhd24 = exports.DolbyVisionLevelFhd60 = exports.DolbyVisionLevelFhd30 = exports.DolbyVisionLevelFhd24 = exports.DolbyVisionLevelHd30 = exports.DolbyVisionLevelHd24 = exports.DolbyVisionLevelUnknown = exports.AV1Level73 = exports.AV1Level72 = exports.AV1Level71 = exports.AV1Level7 = exports.AV1Level63 = exports.AV1Level62 = exports.AV1Level61 = exports.AV1Level6 = exports.AV1Level53 = exports.AV1Level52 = exports.AV1Level51 = exports.AV1Level5 = exports.AV1Level43 = void 0;
/* AVC Profiles */
/**
 * unknown AVC level
 */
exports.AVCLevelUnknown = 0;
/**
 * AVC level 1
 */
exports.AVCLevel1 = 1;
/**
 * AVC level 1b
 */
exports.AVCLevel1b = 2;
/**
 * AVC level 1.1
 */
exports.AVCLevel11 = 3;
/**
 * AVC level 1.2
 */
exports.AVCLevel12 = 4;
/**
 * AVC level 1.3
 */
exports.AVCLevel13 = 5;
/**
 * AVC level 2
 */
exports.AVCLevel2 = 6;
/**
 * AVC level 2.1
 */
exports.AVCLevel21 = 7;
/**
 * AVC level 2.2
 */
exports.AVCLevel22 = 8;
/**
 * AVC level 3
 */
exports.AVCLevel3 = 9;
/**
 * AVC level 3.1
 */
exports.AVCLevel31 = 10;
/**
 * AVC level 3.2
 */
exports.AVCLevel32 = 11;
/**
 * AVC level 4
 */
exports.AVCLevel4 = 12;
/**
 * AVC level 4.1
 */
exports.AVCLevel41 = 13;
/**
 * AVC level 4.2
 */
exports.AVCLevel42 = 14;
/**
 * AVC level 5
 */
exports.AVCLevel5 = 15;
/**
 * AVC level 5.1
 */
exports.AVCLevel51 = 16;
/**
 * AVC level 5.2
 */
exports.AVCLevel52 = 17;
/**
 * AVC level 6
 */
exports.AVCLevel6 = 18;
/**
 * AVC level 6.1
 */
exports.AVCLevel61 = 19;
/**
 * AVC level 6.2
 */
exports.AVCLevel62 = 20;
/* HEVC Levels*/
/**
 * Unknown HEVC level
 */
exports.HEVCLevelUnknown = 100;
/**
 * HEVC main tier level 1
 */
exports.HEVCMainTierLevel1 = 101;
/**
 * HEVC main tier level 2
 */
exports.HEVCMainTierLevel2 = 102;
/**
 * HEVC main tier level 2.1
 */
exports.HEVCMainTierLevel21 = 103;
/**
 * HEVC main tier level 3
 */
exports.HEVCMainTierLevel3 = 104;
/**
 * HEVC main tier level 3.1
 */
exports.HEVCMainTierLevel31 = 105;
/**
 * HEVC main tier level 4
 */
exports.HEVCMainTierLevel4 = 106;
/**
 * HEVC high tier level 4
 */
exports.HEVCHighTierLevel4 = 107;
/**
 * HEVC main tier level 4.1
 */
exports.HEVCMainTierLevel41 = 108;
/**
 * HEVC high tier level 4.1
 */
exports.HEVCHighTierLevel41 = 109;
/**
 * HEVC main tier level 5
 */
exports.HEVCMainTierLevel5 = 110;
/**
 * HEVC high tier level 5
 */
exports.HEVCHighTierLevel5 = 111;
/**
 * HEVC main tier level 5.1
 */
exports.HEVCMainTierLevel51 = 112;
/**
 * HEVC high tier level 5.1
 */
exports.HEVCHighTierLevel51 = 113;
/**
 * HEVC main tier level 5.2
 */
exports.HEVCMainTierLevel52 = 114;
/**
 * HEVC high tier level 5.2
 */
exports.HEVCHighTierLevel52 = 115;
/**
 * HEVC main tier level 6
 */
exports.HEVCMainTierLevel6 = 116;
/**
 * HEVC high tier level 6
 */
exports.HEVCHighTierLevel6 = 117;
/**
 * HEVC main tier level 6.1
 */
exports.HEVCMainTierLevel61 = 118;
/**
 * HEVC high tier level 6.1
 */
exports.HEVCHighTierLevel61 = 119;
/**
 * HEVC main tier level 6.2
 */
exports.HEVCMainTierLevel62 = 120;
/**
 * HEVC high tier level 6.2
 */
exports.HEVCHighTierLevel62 = 121;
/* H263 Levels*/
/**
 * Unknown h263 level
 */
exports.H263LevelUnknown = 200;
/**
 * H263 level 1.0
 */
exports.H263Level10 = 201;
/**
 * H263 level 2.0
 */
exports.H263Level20 = 202;
/**
 * H263 level 3.0
 */
exports.H263Level30 = 203;
/**
 * H263 level 4.0
 */
exports.H263Level40 = 204;
/**
 * H263 level 5.0
 */
exports.H263Level50 = 205;
/**
 * H263 level 6.0
 */
exports.H263Level60 = 206;
/**
 * H263 level 7.0
 */
exports.H263Level70 = 207;
/* VP9 Levels*/
/**
 * Unknown VP9 level
 */
exports.VP9LevelUnknown = 300;
/**
 * VP9 Level 1
 */
exports.VP9Level1 = 301;
/**
 * VP9 Level 1.1
 */
exports.VP9Level11 = 302;
/**
 * VP9 Level 2
 */
exports.VP9Level2 = 303;
/**
 * VP9 Level 2.1
 */
exports.VP9Level21 = 304;
/**
 * VP9 Level 3
 */
exports.VP9Level3 = 305;
/**
 * VP9 Level 3.1
 */
exports.VP9Level31 = 306;
/**
 * VP9 Level 4
 */
exports.VP9Level4 = 307;
/**
 * VP9 Level 4.1
 */
exports.VP9Level41 = 308;
/**
 * VP9 Level 5
 */
exports.VP9Level5 = 309;
/**
 * VP9 Level 5.1
 */
exports.VP9Level51 = 310;
/**
 * VP9 Level 5.2
 */
exports.VP9Level52 = 311;
/**
 * VP9 Level 6
 */
exports.VP9Level6 = 312;
/**
 * VP9 Level 6.1
 */
exports.VP9Level61 = 313;
/**
 * VP9 Level 6.2
 */
exports.VP9Level62 = 314;
/* MPEG4 Levels*/
/**
 * Unknown MPEG4 level
 */
exports.MPEG4LevelUnknown = 400;
/**
 * MPEG4 level 0
 */
exports.MPEG4Level0 = 401;
/**
 * MPEG4 level 0b
 */
exports.MPEG4Level0b = 402;
/**
 * MPEG4 level 1
 */
exports.MPEG4Level1 = 403;
/**
 * MPEG4 level 2
 */
exports.MPEG4Level2 = 404;
/**
 * MPEG4 level 3
 */
exports.MPEG4Level3 = 405;
/**
 * MPEG4 level 3b
 */
exports.MPEG4Level3b = 406;
/**
 * MPEG4 level 4
 */
exports.MPEG4Level4 = 407;
/**
 * MPEG4 level 4a
 */
exports.MPEG4Level4a = 408;
/**
 * MPEG4 level 5
 */
exports.MPEG4Level5 = 409;
/**
 * MPEG4 level 6
 */
exports.MPEG4Level6 = 410;
/* MPEG2 Levels*/
/**
 * Unknown MPEG2 level
 */
exports.MPEG2LevelUnknown = 500;
/**
 * MPEG2 low level
 */
exports.MPEG2LevelLL = 501;
/**
 * MPEG2 main level
 */
exports.MPEG2LevelML = 502;
/**
 * MPEG2 high 1440
 */
exports.MPEG2LevelH14 = 503;
/**
 * MPEG2 high level
 */
exports.MPEG2LevelHL = 504;
/**
 * MPEG2 level HP
 */
exports.MPEG2LevelHP = 505;
/* VP8 Levels*/
/**
 * Unknown VP8 level
 */
exports.VP8LevelUnknown = 600;
/**
 * VP8 level version 0
 */
exports.VP8LevelVersion0 = 601;
/**
 * VP8 level version 1
 */
exports.VP8LevelVersion1 = 602;
/**
 * VP8 level version 2
 */
exports.VP8LevelVersion2 = 603;
/**
 * VP8 level version 3
 */
exports.VP8LevelVersion3 = 604;
/* AV1 Levels*/
/**
 * unknown AV1 level
 */
exports.AV1LevelUnknown = 700;
/**
 * AV1 level 2
 */
exports.AV1Level2 = 701;
/**
 * AV1 level 2.1
 */
exports.AV1Level21 = 702;
/**
 * AV1 level 2.2
 */
exports.AV1Level22 = 703;
/**
 * AV1 level 2.3
 */
exports.AV1Level23 = 704;
/**
 * AV1 level 3
 */
exports.AV1Level3 = 705;
/**
 * AV1 level 3.1
 */
exports.AV1Level31 = 706;
/**
 * AV1 level 3.2
 */
exports.AV1Level32 = 707;
/**
 * AV1 level 3.3
 */
exports.AV1Level33 = 708;
/**
 * AV1 level 4
 */
exports.AV1Level4 = 709;
/**
 * AV1 level 4.1
 */
exports.AV1Level41 = 710;
/**
 * AV1 level 4.2
 */
exports.AV1Level42 = 711;
/**
 * AV1 level 4.3
 */
exports.AV1Level43 = 712;
/**
 * AV1 level 5
 */
exports.AV1Level5 = 713;
/**
 * AV1 level 5.1
 */
exports.AV1Level51 = 714;
/**
 * AV1 level 5.2
 */
exports.AV1Level52 = 715;
/**
 * AV1 level 5.3
 */
exports.AV1Level53 = 716;
/**
 * AV1 level 6
 */
exports.AV1Level6 = 717;
/**
 * AV1 level 6.1
 */
exports.AV1Level61 = 718;
/**
 * AV1 level 6.2
 */
exports.AV1Level62 = 719;
/**
 * AV1 level 6.3
 */
exports.AV1Level63 = 720;
/**
 * AV1 level 7
 */
exports.AV1Level7 = 721;
/**
 * AV1 level 7.1
 */
exports.AV1Level71 = 722;
/**
 * AV1 level 7.2
 */
exports.AV1Level72 = 723;
/**
 * AV1 level 7.3
 */
exports.AV1Level73 = 724;
/* DolbyVision Levels*/
/**
 * Unknown dolby vision level
 */
exports.DolbyVisionLevelUnknown = 800;
/**
 * Dolby vision level HD 2.4
 */
exports.DolbyVisionLevelHd24 = 801;
/**
 * Dolby vision level HD 3.0
 */
exports.DolbyVisionLevelHd30 = 802;
/**
 * Dolby vision level FHD 2.4
 */
exports.DolbyVisionLevelFhd24 = 803;
/**
 * Dolby vision level FHD 3.0
 */
exports.DolbyVisionLevelFhd30 = 804;
/**
 * Dolby vision level FHD 6.0
 */
exports.DolbyVisionLevelFhd60 = 805;
/**
 * Dolby vision level UHD 2.4
 */
exports.DolbyVisionLevelUhd24 = 806;
/**
 * Dolby vision level UHD 3.0
 */
exports.DolbyVisionLevelUhd30 = 807;
/**
 * Dolby vision level UHD 4.8
 */
exports.DolbyVisionLevelUhd48 = 808;
/**
 * Dolby vision level UHD 6.0
 */
exports.DolbyVisionLevelUhd60 = 809;
