// @amzn/keplermediadescriptor/dist/turbo-modules/ProfileValues.js
"use strict";
/*
 * Copyright 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MPEG4ProfileAdvancedCore = exports.MPEG4ProfileAdvancedCoding = exports.MPEG4ProfileCoreScalable = exports.MPEG4ProfileAdvancedRealTime = exports.MPEG4ProfileHybrid = exports.MPEG4ProfileBasicAnimated = exports.MPEG4ProfileSimpleFBA = exports.MPEG4ProfileSimpleFace = exports.MPEG4ProfileScalableTexture = exports.MPEG4ProfileNbit = exports.MPEG4ProfileMain = exports.MPEG4ProfileCore = exports.MPEG4ProfileSimpleScalable = exports.MPEG4ProfileSimple = exports.MPEG4ProfileUnknown = exports.VP9Profile3HDR10Plus = exports.VP9Profile2HDR10Plus = exports.VP9Profile3HDR = exports.VP9Profile2HDR = exports.VP9Profile3 = exports.VP9Profile2 = exports.VP9Profile1 = exports.VP9Profile0 = exports.VP9ProfileUnknown = exports.H263ProfileHighLatency = exports.H263ProfileInterlace = exports.H263ProfileInternet = exports.H263ProfileHighCompression = exports.H263ProfileISWV3 = exports.H263ProfileISWV2 = exports.H263ProfileBackwardCompatible = exports.H263ProfileH320Coding = exports.H263ProfileBaseline = exports.H263ProfileUnknown = exports.HEVCProfileMain10HDR10Plus = exports.HEVCProfileMain10HDR10 = exports.HEVCProfileMainStill = exports.HEVCProfileMain10 = exports.HEVCProfileMain = exports.HEVCProfileUnknown = exports.AVCProfileConstrainedHigh = exports.AVCProfileConstrainedBaseline = exports.AVCProfileHigh444 = exports.AVCProfileHigh422 = exports.AVCProfileHigh10 = exports.AVCProfileHigh = exports.AVCProfileExtended = exports.AVCProfileMain = exports.AVCProfileBaseline = exports.AVCProfileUnknown = void 0;
exports.DolbyVisionProfileDvavSe = exports.DolbyVisionProfileDvheSt = exports.DolbyVisionProfileDvheDtb = exports.DolbyVisionProfileDvheDth = exports.DolbyVisionProfileDvheStn = exports.DolbyVisionProfileDvheDtr = exports.DolbyVisionProfileDvheDen = exports.DolbyVisionProfileDvheDer = exports.DolbyVisionProfileDvavPen = exports.DolbyVisionProfileDvavPer = exports.DolbyVisionProfileUnknown = exports.AV1ProfileMain10HDR10Plus = exports.AV1ProfileMain10HDR10 = exports.AV1ProfileMain10 = exports.AV1ProfileMain8 = exports.AV1ProfileUnknown = exports.VP8ProfileMain = exports.VP8ProfileUnknown = exports.MPEG2ProfileHigh = exports.MPEG2ProfileSpatial = exports.MPEG2ProfileSNR = exports.MPEG2Profile422 = exports.MPEG2ProfileMain = exports.MPEG2ProfileSimple = exports.MPEG2ProfileUnknown = exports.MPEG4ProfileAdvancedSimple = exports.MPEG4ProfileAdvancedScalable = void 0;
/* AVC Profiles */
/**
 * Unknown AVC profile
 */
exports.AVCProfileUnknown = 0;
/**
 * AVC baseline profile
 */
exports.AVCProfileBaseline = 1;
/**
 * AVC main profile
 */
exports.AVCProfileMain = 2;
/**
 * AVC extended profile
 */
exports.AVCProfileExtended = 3;
/**
 * AVC high profile
 */
exports.AVCProfileHigh = 4;
/**
 * AVC high profile 10
 */
exports.AVCProfileHigh10 = 5;
/**
 * AVC high profile  4:2:2
 */
exports.AVCProfileHigh422 = 6;
/**
 * AVC high profile  4:4:4
 */
exports.AVCProfileHigh444 = 7;
/**
 * AVC Constrained Baseline Profile
 */
exports.AVCProfileConstrainedBaseline = 8;
/**
 * AVC Constrained High Profile
 */
exports.AVCProfileConstrainedHigh = 9;
/* HEVC Profiles */
/**
 * Unknown HEVC profile
 */
exports.HEVCProfileUnknown = 100;
/**
 * HEVC main profile
 */
exports.HEVCProfileMain = 101;
/**
 * HEVC main 10 proile
 */
exports.HEVCProfileMain10 = 102;
/**
 * HEVC Main Still Picture profile
 */
exports.HEVCProfileMainStill = 103;
/**
 * HEVC HDR main 10 profile
 */
exports.HEVCProfileMain10HDR10 = 104;
/**
 * HEVC HDR10+ main 10 profile
 */
exports.HEVCProfileMain10HDR10Plus = 105;
/* H263 Profiles */
/**
 * Unknown H263 profile
 */
exports.H263ProfileUnknown = 200;
/**
 * H263 baseline profile
 */
exports.H263ProfileBaseline = 201;
/**
 * H263 H320Coding profile
 */
exports.H263ProfileH320Coding = 202;
/**
 * H263 BackwardCompatible profile
 */
exports.H263ProfileBackwardCompatible = 203;
/**
 * H263 ISWV2 profile
 */
exports.H263ProfileISWV2 = 204;
/**
 * H263 ISWV3 profile
 */
exports.H263ProfileISWV3 = 205;
/**
 * H263 HighCompression profile
 */
exports.H263ProfileHighCompression = 206;
/**
 * H263 Internet profile
 */
exports.H263ProfileInternet = 207;
/**
 * H263 Interlace profile
 */
exports.H263ProfileInterlace = 208;
/**
 * H263 HighLatency profile
 */
exports.H263ProfileHighLatency = 209;
/* VP9 Profiles */
/**
 * Unknown VP9 profile
 */
exports.VP9ProfileUnknown = 300;
/**
 * VP9 profile 0
 */
exports.VP9Profile0 = 301;
/**
 * VP9 profile 1
 */
exports.VP9Profile1 = 302;
/**
 * VP9 profile 2
 */
exports.VP9Profile2 = 303;
/**
 * VP9 profile 3
 */
exports.VP9Profile3 = 304;
/**
 * VP9 profile 2 for HDR
 */
exports.VP9Profile2HDR = 305;
/**
 * VP9 profile 3 for HDR
 */
exports.VP9Profile3HDR = 306;
/**
 * VP9 profile 2 for HDR10+
 */
exports.VP9Profile2HDR10Plus = 307;
/**
 * VP9 profile 3 for HDR10+
 */
exports.VP9Profile3HDR10Plus = 308;
/* MPEG4 Profiles */
/**
 * unknown MPEG4 profile
 */
exports.MPEG4ProfileUnknown = 400;
/**
 * MPEG4 simple profile
 */
exports.MPEG4ProfileSimple = 401;
/**
 * MPEG4 simple scalable profile
 */
exports.MPEG4ProfileSimpleScalable = 402;
/**
 * MPEG4 core profile
 */
exports.MPEG4ProfileCore = 403;
/**
 * MPEG4 main profile
 */
exports.MPEG4ProfileMain = 404;
/**
 * MPEG4 Nbit profile
 */
exports.MPEG4ProfileNbit = 405;
/**
 * MPEG4 scalable texture profile
 */
exports.MPEG4ProfileScalableTexture = 406;
/**
 * MPEG4 simple face profile
 */
exports.MPEG4ProfileSimpleFace = 407;
/**
 * MPEG4 simple FBA profile
 */
exports.MPEG4ProfileSimpleFBA = 408;
/**
 * MPEG4 basic animated profile
 */
exports.MPEG4ProfileBasicAnimated = 409;
/**
 * MPEG4 hybrid profile
 */
exports.MPEG4ProfileHybrid = 410;
/**
 * MPEG4 advanced realtime profile
 */
exports.MPEG4ProfileAdvancedRealTime = 411;
/**
 * MPEG4 core scalable profile
 */
exports.MPEG4ProfileCoreScalable = 412;
/**
 * MPEG4 advanced coding profile
 */
exports.MPEG4ProfileAdvancedCoding = 413;
/**
 * MPEG4 advanced core profile
 */
exports.MPEG4ProfileAdvancedCore = 414;
/**
 * MPEG4 advanced scalable profile
 */
exports.MPEG4ProfileAdvancedScalable = 415;
/**
 * MPEG4 advanced simple profile
 */
exports.MPEG4ProfileAdvancedSimple = 416;
/* MPEG2 Profiles */
/**
 * unknown MPEG2 profile
 */
exports.MPEG2ProfileUnknown = 500;
/**
 * MPEG2 simple profile
 */
exports.MPEG2ProfileSimple = 501;
/**
 * MPEG2 main profile
 */
exports.MPEG2ProfileMain = 502;
/**
 * MPEG2 4:2:2 profile
 */
exports.MPEG2Profile422 = 503;
/**
 * MPEG2 SNR scalable profile
 */
exports.MPEG2ProfileSNR = 504;
/**
 * MPEG2 spatially scalable profile
 */
exports.MPEG2ProfileSpatial = 505;
/**
 * MPEG2 high profile
 */
exports.MPEG2ProfileHigh = 506;
/* VP8 Profiles */
/**
 * Unknown VP8 profile
 */
exports.VP8ProfileUnknown = 600;
/**
 * VP8 main profile
 */
exports.VP8ProfileMain = 601;
/* AV1 Profiles */
/**
 * Unknown AV1 profile
 */
exports.AV1ProfileUnknown = 700;
/**
 * AV1 main profile 8 bit depth
 */
exports.AV1ProfileMain8 = 701;
/**
 * AV1 main profile 10 bit depth
 */
exports.AV1ProfileMain10 = 702;
/**
 * AV1 main profile 10 bit depth for HDR
 */
exports.AV1ProfileMain10HDR10 = 703;
/**
 * AV1 main profile 10 bit depth for HDR10+
 */
exports.AV1ProfileMain10HDR10Plus = 704;
/* Dolby Vision Profiles */
/**
 * Unknwon Dolby vision profile
 */
exports.DolbyVisionProfileUnknown = 800;
/**
 * Dolby vision profile based on AV1 codec.\
 * Dolby profile ID 0 - BL signal cross-compatibility ID 2
 */
exports.DolbyVisionProfileDvavPer = 801;
/**
 * Dolby vision profile based on AV1 codec.\
 * Dolby profile ID 1 - BL signal cross-compatibility ID 0
 */
exports.DolbyVisionProfileDvavPen = 802;
/**
 * Dolby vision profile based on 8 bit HEVC codec.\
 * Dolby profile ID 2 - BL signal cross-compatibility ID 2
 */
exports.DolbyVisionProfileDvheDer = 803;
/**
 * Dolby vision profile based on 8 bit HEVC codec.\
 * Dolby profile ID 2 - BL signal cross-compatibility ID 2
 */
exports.DolbyVisionProfileDvheDen = 804;
/**
 * Dolby vision profile based on HEVC codec.
 * Dolby profile ID 4
 */
exports.DolbyVisionProfileDvheDtr = 805;
/**
 * Dolby vision profile based on HEVC codec.
 * Dolby profile ID 5
 */
exports.DolbyVisionProfileDvheStn = 806;
/**
 * Dolby vision profile based on 10bit HEVC codec\
 * Dolby profile ID 6 - BL signal cross-compatibility ID 1
 */
exports.DolbyVisionProfileDvheDth = 807;
/**
 * Dolby vision profile based on HEVC codec.\
 * Dolby profile ID 7
 */
exports.DolbyVisionProfileDvheDtb = 808;
/**
 * Dolby vision profile based on HEVC codec.\
 * Dolby profile ID 8
 */
exports.DolbyVisionProfileDvheSt = 809;
/**
 * Dolby vision profile based on AVC codec.\
 * Dolby profile ID 9
 */
exports.DolbyVisionProfileDvavSe = 810;
