// @amzn/react-native-kepler/Libraries/AppStateManager/AppStateManager.js
/*
* Copyright 2023 Amazon.com, Inc. or its affiliates. All rights reserved.
*
* AMAZON PROPRIETARY/CONFIDENTIAL
*
* You may not use this file except in compliance with the terms and
* conditions set forth in the accompanying LICENSE.TXT file.
*
* THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
* DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
* IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
*/
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppStateManager = void 0;
const AppStateManagerTurboModuleSpec_1 = __importDefault(require("./AppStateManagerTurboModuleSpec"));
/**
 * `AppStateManager` provides a way to save and restore the UI States for the app component
 */
class AppStateManagerImpl {
    /**
     * Fetches the data app developer stored.
     * @param {number} rootTag - RootTag of the app component.
     * @return {JsonObject | undefined} - Data that App developer saved.
     */
    getSavedData(rootTag) {
        const savedData = AppStateManagerTurboModuleSpec_1.default.getSavedData(rootTag);
        if (savedData == "" || savedData == null) {
            return undefined;
        }
        return JSON.parse(savedData);
    }
    /**
     * @deprecated use addSaveEventListener instead
     *
     * Register a listener for Saving the state. Each listener for saving the state is tagged to the root tag. Multiple listeners cannot be added for a given rootTag
     * The listener can be removed either from the application side, if it is not removed from application side then it gets removed when screen gets destroyed
     * @param {number} rootTag - RootTag of the app component.
     * @param {OnSaveStateCallback} callback - Callback to be invoked when an event for saving the data arrives.
     */
    addSaveListener(rootTag, callback) {
        console.warn("AppStateManager::addSaveListener is deprecated. Use addSaveEventListener instead.");
        const callbackInternal = () => {
            try {
                const saveData = callback();
                AppStateManagerTurboModuleSpec_1.default.setSaveState(JSON.stringify(saveData), rootTag);
            }
            catch (error) {
                console.error("SaveState - error in callback: ", error);
            }
        };
        AppStateManagerTurboModuleSpec_1.default.addSaveStateListener(rootTag.toString(), callbackInternal);
    }
    /**
     * Register a listener for Saving the state. Each listener for saving the state is tagged to the root tag. Multiple listeners cannot be added for a given rootTag
     * The listener can be removed either from the application side, if it is not removed from application side then it gets removed when screen gets destroyed
     * @param {number} rootTag - RootTag of the app component.
     * @param {OnSaveStateCallback} callback - Callback to be invoked when an event for saving the data arrives.
     *
     * @returns pointer to the created EventSubscription
     */
    addSaveEventListener(rootTag, callback) {
        const callbackInternal = () => {
            try {
                const saveData = callback();
                AppStateManagerTurboModuleSpec_1.default.setSaveState(JSON.stringify(saveData), rootTag);
            }
            catch (error) {
                console.error("SaveState - error in callback: ", error);
            }
        };
        return AppStateManagerTurboModuleSpec_1.default.addSaveStateEventListener(rootTag.toString(), callbackInternal);
    }
    /**
     * Removes listener for Saving the state.
     * @param {number} rootTag - RootTag of the app component.
     */
    removeSaveListener(rootTag) {
        AppStateManagerTurboModuleSpec_1.default.removeSaveStateListener(rootTag.toString());
    }
}
exports.AppStateManager = new AppStateManagerImpl();
