// @amazon-devices/react-native-kepler/Libraries/AppState/AppState.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Portions of this file are copyright (c) 2021 Amazon.com, Inc. or its
 * affiliates.  All rights reserved.
 *
 * PORTIONS OF THIS FILE ARE AMAZON PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO
 * LICENSE TERMS.
 *
 * Amazon modifications are indicated by [amznmod_* comments].
 *
 * @flow strict-local
 * @format
 */

/**
 * CHANGELOG:
 * e326dc898e - Altering KeplerTurboModule to allow for using JS Sub management.
 */

 import {type EventSubscription} from '../vendor/emitter/EventEmitter';
 import NativeEventEmitter from '../EventEmitter/NativeEventEmitter';
 import logError from '../Utilities/logError';
 import Platform from '../Utilities/Platform';

 // amznmod_react - headless support by excluding NativeAppState
 let NativeAppState;
 let ModuleStateKepler;
 if (typeof __KEPLER_HEADLESS__ === 'undefined') {
   const module = require('./NativeAppState');
   NativeAppState = module.default;
   ModuleStateKepler = module.ModuleStateKepler;
 } else {
   console.debug(
     'AppState is not supported within a headless service or task.',
   );
 }
 
 
 export type AppStateValues = 'inactive' | 'background' | 'active';
 
 type AppStateEventDefinitions = {
   change: [AppStateValues],
   // amznmod_react - Add Kepler support for module states
   moduleChange: [AppStateValues],
   memoryWarning: [],
   blur: [],
   focus: [],
 };
 
 type NativeAppStateEventDefinitions = {
   // amznmod_react - Add Kepler support for module states
   appStateDidChange: [{app_state: AppStateValues},{module_states: AppStateValues}],
   appStateFocusChange: [boolean],
   memoryWarning: [],
 };
 
 
 /**
  * `AppState` can tell you if the app is in the foreground or background,
  * and notify you when the state changes.
  *
  * See https://reactnative.dev/docs/appstate.html
  */
 
 class AppState {
   currentState: ?string = null;
     // amznmod_react - Add Kepler support for module states and reconfigure
     _supportedEvents = ['change', 'moduleChange', 'memoryWarning', 'blur', 'focus', 'reconfigure'];
     currentModuleStates: ?ModuleStateKepler[];
     currentState: ?string;
   isAvailable: boolean;
 
   _emitter: ?NativeEventEmitter<NativeAppStateEventDefinitions>;
 
   constructor() {
     if (NativeAppState == null) {
       this.isAvailable = false;
       // amznmod_react - headless support by excluding NativeAppState
       // we are following the iOS-specific contract which includes 'unknown'
       this.currentState = 'unknown';
     } else {
       this.isAvailable = true;
 
       const emitter: NativeEventEmitter<NativeAppStateEventDefinitions> = new NativeEventEmitter(
         // T88715063: NativeEventEmitter only used this parameter on iOS. Now it uses it on all platforms, so this code was modified automatically to preserve its behavior
         // If you want to use the native module on other platforms, please remove this condition and test its behavior
         // amznmod_react change platform to be kepler applicable
         Platform.OS !== 'kepler' ? null : NativeAppState,
       );
       this._emitter = emitter;
       this.currentState = NativeAppState.getConstants().initialAppState;
       // amznmod_react - Add Kepler support for module states
       this.currentModuleStates = [];
       let eventUpdated = false;
 
       // TODO: this is a terrible solution - in order to ensure `currentState`
       // prop is up to date, we have to register an observer that updates it
       // whenever the state changes, even if nobody cares. We should just
       // deprecate the `currentState` property and get rid of this.
       emitter.addListener('appStateDidChange', appStateData => {
         eventUpdated = true;
         this.currentState = appStateData.app_state;
         // amznmod_react - Add Kepler support for module states
         this.currentModuleStates = appStateData.module_states;
       });
 
       // TODO: see above - this request just populates the value of `currentState`
       // when the module is first initialized. Would be better to get rid of the
       // prop and expose `getCurrentAppState` method directly.
       // $FlowExpectedError[incompatible-call]
       NativeAppState.getCurrentAppState(appStateData => {
         // It's possible that the state will have changed here & listeners need to be notified
         // amznmod_react - Add Kepler support for module states
         // Removing check to see if current state is different from what was returned
         // since a module state may have changed and it's a complex comparison
         // to determine if the currentModuleStates contains the same modules
         // with the same states as appStateData.module_states.
         // Will just send the update
         if (!eventUpdated /*&& this.currentState !== appStateData.app_state*/) {
           this.currentState = appStateData.app_state;
           this.currentModuleStates = appStateData.module_states;
           emitter.emit('appStateDidChange', appStateData);
         }
       }, logError);
     }
   }
 
   /**
    * Add a handler to AppState changes by listening to the `change` event type
    * and providing the handler.
    *
    * See https://reactnative.dev/docs/appstate.html#addeventlistener
    */
   addEventListener<K: $Keys<AppStateEventDefinitions>>(
     type: K,
     handler: (...$ElementType<AppStateEventDefinitions, K>) => void,
   ): EventSubscription {
     // amznmod_react - headless support, dummy return
     if (typeof __KEPLER_HEADLESS__ !== 'undefined' && __KEPLER_HEADLESS__ == true) {
       console.warn(
         'AppState is not supported within a headless service or task.',
       );
       return {
         remove: () => {},
       };
     }

     const emitter = this._emitter;
     if (emitter == null) {
       throw new Error('Cannot use AppState when `isAvailable` is false.');
     }
     switch (type) {
       case 'change':
         // $FlowIssue[invalid-tuple-arity] Flow cannot refine handler based on the event type
         const changeHandler: AppStateValues => void = handler;
         return emitter.addListener('appStateDidChange', appStateData => {
           changeHandler(appStateData.app_state);
         });
         // amznmod_react - Add Kepler support for module states
       case 'moduleChange':
         // $FlowIssue[invalid-tuple-arity] Flow cannot refine handler based on the event type
         const moduleChangeHandler: AppStateValues => void = handler;
         return emitter.addListener('appStateDidChange', appStateData => {
           moduleChangeHandler(appStateData.module_states);
         });
       case 'memoryWarning':
         // $FlowIssue[invalid-tuple-arity] Flow cannot refine handler based on the event type
         const memoryWarningHandler: () => void = handler;
         return emitter.addListener('memoryWarning', memoryWarningHandler);
       // amznmod_react - add support for reconfigure
       case 'reconfigure':
         const reconfigureHandler: () => void = handler;
         return emitter.addListener('reconfigure', reconfigureData => {
            reconfigureHandler(reconfigureData);
          });
       case 'blur':
       case 'focus':
         // $FlowIssue[invalid-tuple-arity] Flow cannot refine handler based on the event type
         const focusOrBlurHandler: () => void = handler;
         return emitter.addListener('appStateFocusChange', hasFocus => {
           if (type === 'blur' && !hasFocus) {
             focusOrBlurHandler();
           }
           if (type === 'focus' && hasFocus) {
             focusOrBlurHandler();
           }
         });
     }
     throw new Error('Trying to subscribe to unknown event: ' + type);
   }
 }
 
 module.exports = (new AppState(): AppState);
