// @amazon-devices/react-native-kepler/Libraries/KeplerAppState/KeplerAppStateManager.js
"use strict";
/*
 * Copyright 2021-24 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useComponentInstance = exports.useAddKeplerAppStateListenerWithReplayCallback = exports.useAddKeplerAppStateListenerCallback = exports.useGetCurrentKeplerAppStateCallback = exports.useKeplerAppStateManager = void 0;
const react_1 = require("react");
const KeplerAppStateManagerSpec_1 = __importDefault(require("./KeplerAppStateManagerSpec"));
const RootTag_1 = require("../ReactNative/RootTag");
class KeplerAppStateManager {
    constructor(rootTag) {
        /**
         * @brief Returns the current app state of your interactive component.
         */
        this.getCurrentState = () => {
            return KeplerAppStateManagerSpec_1.default.getCurrentState(this.rootTag);
        };
        /**
         * @brief Adds a listener for the given event.
         *
         * @param {KeplerAppStateEvent} eventName - The event name to subscribe to.
         * @param {OnEventCallback} callback - Callback to be invoked when an event arrives.
         * @returns {EventSubscription} - A handle to the subscription. Call `remove()` to remove the subscription.
         */
        this.addAppStateListener = (eventName, callback) => {
            return KeplerAppStateManagerSpec_1.default.addEventListener(this.rootTag, eventName, callback);
        };
        /**
         * @brief Adds a listener for the given event and this will capture and replay the events that came before it was attached.
         *
         * @param {KeplerAppStateEvent} eventName - The event name to subscribe to.
         * @param {OnEventWithReplayCallback} callback - Callback to be invoked when an event arrives.
         * @returns {EventSubscription} - A handle to the subscription. Call `remove()` to remove the subscription.
         */
        this.addEventListenerWithReplay = (eventName, callback) => {
            return KeplerAppStateManagerSpec_1.default.addEventListenerWithReplay(this.rootTag, eventName, callback);
        };
        /**
         * @brief Returns an interface of type IComponentInstance
         */
        this.getComponentInstance = () => {
            return KeplerAppStateManagerSpec_1.default.getComponentInstance(this.rootTag);
        };
        this.rootTag = rootTag;
    }
}
/**
* A custom hook that returns an instance of IKeplerAppStateManager.
*/
const useKeplerAppStateManager = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    return react_1.useMemo(() => new KeplerAppStateManager(rootTag), [rootTag]);
};
exports.useKeplerAppStateManager = useKeplerAppStateManager;
/**
 * @deprecated Use getCurrentState from KeplerAppStateManager instead
 *
 * @brief A custom hook that returns a callback that, when invoked, returns the current app state of
 * your interactive component.
 */
const useGetCurrentKeplerAppStateCallback = () => {
    const keplerAppStateManager = exports.useKeplerAppStateManager();
    return react_1.useCallback(() => {
        return keplerAppStateManager.getCurrentState();
    }, [keplerAppStateManager]);
};
exports.useGetCurrentKeplerAppStateCallback = useGetCurrentKeplerAppStateCallback;
/**
 * @deprecated Use addAppStateListener from KeplerAppStateManager instead
 *
 * @brief A custom hook that returns a callback that, when invoked, adds a listener for the given event.
 *
 * @returns KeplerAppStateCallback: a callback function to invoke to register a listener.
 *     @param {KeplerAppStateEvent} eventName - The event name to subscribe to.
 *     @param {OnAppStateEventCallback} callback - Callback to be invoked when an event arrives
 *     @returns {EventSubscription} - A handle to the subscription. Call `remove()` to remove the subscription.
 */
const useAddKeplerAppStateListenerCallback = () => {
    const keplerAppStateManager = exports.useKeplerAppStateManager();
    return react_1.useCallback((eventName, callback) => {
        return keplerAppStateManager.addAppStateListener(eventName, callback);
    }, [keplerAppStateManager]);
};
exports.useAddKeplerAppStateListenerCallback = useAddKeplerAppStateListenerCallback;
/**
 * @deprecated Use addEventListenerWithReplay from KeplerAppStateManager instead
 *
 * @brief A custom hook that returns a callback that, when invoked, adds a listener for the given event and this will capture and replay the events that came before it was attached.
 *
 * @returns KeplerAppStateWithReplayCallback: a callback function to invoke to register a listener.
 *     @param {KeplerAppStateEvent} eventName - The event name to subscribe to.
 *     @param {OnAppStateEventWithReplayCallback} callback - Callback to be invoked when an event arrives
 *     @returns {EventSubscription} - A handle to the subscription. Call `remove()` to remove the subscription.
 */
const useAddKeplerAppStateListenerWithReplayCallback = () => {
    const keplerAppStateManager = exports.useKeplerAppStateManager();
    return react_1.useCallback((eventName, callback) => {
        return keplerAppStateManager.addEventListenerWithReplay(eventName, callback);
    }, [keplerAppStateManager]);
};
exports.useAddKeplerAppStateListenerWithReplayCallback = useAddKeplerAppStateListenerWithReplayCallback;
/**
 * @deprecated use getComponentInstance from KeplerAppStateManager instead
 *
 * @brief A custom hook that returns an interface of type IComponentInstance
 *
 * @returns IComponentInstance: an interface having name and type.
 */
const useComponentInstance = () => {
    const keplerAppStateManager = exports.useKeplerAppStateManager();
    return keplerAppStateManager.getComponentInstance();
};
exports.useComponentInstance = useComponentInstance;
