// @amzn/react-native-kepler/Libraries/TimeoutManager/TimeoutManager.js
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutManager = exports.useSetLifespanCallback = exports.useSetTimeoutCallback = exports.LIFESPAN_POLICY = void 0;
const react_1 = require("react");
const NativeTimeoutManager_1 = __importDefault(require("./NativeTimeoutManager"));
const RootTag_1 = require("../ReactNative/RootTag");
var NativeTimeoutManager_2 = require("./NativeTimeoutManager");
Object.defineProperty(exports, "LIFESPAN_POLICY", { enumerable: true, get: function () { return NativeTimeoutManager_2.LIFESPAN_POLICY; } });
/**
 * `TimeoutManager` provides a way to set timeout period and lifespan policy
 * for individual component
 *
 */
class TimeoutManagerLibrary {
    /**
     * @brief set a timeout period during app inactivity for the screen named `moduleName`
     *
     * @param moduleName: name for the screen, the value should be identical with screenKey
     * that users register in app's main.cpp create() method
     * @param timeout: period of time to keep inactive component in foreground
     *
     */
    setTimeout(moduleName, timeout) {
        NativeTimeoutManager_1.default.setTimeout(moduleName, timeout);
    }
    /**
     * @brief set a lifespan policy during app inactivity for the screen named `moduleName`
     *
     * @param moduleName: name for the screen, the value should be identical with screenKey
     * that users register in app's main.cpp create() method
     * @param policy: type of lifespan policy
     * can determine which app in backstack comes to foreground first
     *
     */
    setLifespan(moduleName, policy) {
        NativeTimeoutManager_1.default.setLifespan(moduleName, policy);
    }
}
/**
 * @brief set a timeout period during app inactivity for the screen.
 *
 * @description A custom hook that returns a callback that sets the timeout period
 * for your KeplerScript application.
 *
 * @returns A callback function to invoke to set the timeout of your application.
 * The callback takes the desired timeout as a parameter in seconds.
 *
 */
const useSetTimeoutCallback = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    return react_1.useCallback((timeout) => NativeTimeoutManager_1.default.setTimeoutV2(rootTag, timeout), [rootTag]);
};
exports.useSetTimeoutCallback = useSetTimeoutCallback;
/**
 * @brief set a lifespan policy during app inactivity for the screen.
 *
 * @description A custom hook that returns a callback that sets the lifespan policy
 * for your KeplerScript application.
 *
 * @returns A callback function to invoke to set the desired policy for your
 * application. he callback takes the desired lifespan policy as a parameter.
 * See the LIFESPAN_POLICY enum for valid values.
 *
 */
const useSetLifespanCallback = () => {
    const rootTag = react_1.useContext(RootTag_1.RootTagContext);
    return react_1.useCallback((policy) => NativeTimeoutManager_1.default.setLifespanV2(rootTag, policy), [rootTag]);
};
exports.useSetLifespanCallback = useSetLifespanCallback;
exports.TimeoutManager = new TimeoutManagerLibrary();
