// @amazon-devices/react-native-w3cmedia/dist/index.js
"use strict";
/*
 * Copyright 2022-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeplerMediaControlHandler = exports.VTTCue = exports.decodingInfo = exports.TextDecoder = exports.AudioContentType = exports.AudioUsageType = exports.KeplerVideoView = exports.KeplerShareableSurfaceView = exports.KeplerVideoSurfaceView = exports.KeplerCaptionsView = exports.WebCrypto = exports.W3CMediaTurboModule = exports.QuotaExceededError = exports.InvalidStateError = exports.NotSupportedError = exports.W3CLibPolyfill = exports.requestMediaKeysSystemAccess = exports.requestMediaKeySystemAccess = exports.TrackEvent = exports.Event = exports.HTMLAudioElement = exports.HTMLVideoElement = exports.HTMLMediaElement = exports.EndOfStreamError = exports.ReadyState = exports.AppendMode = exports.SourceBufferImpl = exports.MediaSource = exports.VideoPlayer = exports.AudioPlayer = exports.Video = exports.Audio = exports.Media = void 0;
const KeplerCaptionsView_1 = require("./interface/KeplerCaptionsView");
Object.defineProperty(exports, "KeplerCaptionsView", { enumerable: true, get: function () { return KeplerCaptionsView_1.KeplerCaptionsView; } });
const KeplerVideoSurfaceView_1 = require("./interface/KeplerVideoSurfaceView");
Object.defineProperty(exports, "KeplerVideoSurfaceView", { enumerable: true, get: function () { return KeplerVideoSurfaceView_1.KeplerVideoSurfaceView; } });
const KeplerShareableSurfaceView_1 = require("./interface/KeplerShareableSurfaceView");
Object.defineProperty(exports, "KeplerShareableSurfaceView", { enumerable: true, get: function () { return KeplerShareableSurfaceView_1.KeplerShareableSurfaceView; } });
const KeplerVideoView_1 = require("./interface/KeplerVideoView");
Object.defineProperty(exports, "KeplerVideoView", { enumerable: true, get: function () { return KeplerVideoView_1.KeplerVideoView; } });
const EncryptedMediaImpl_1 = require("./EncryptedMediaImpl");
Object.defineProperty(exports, "requestMediaKeysSystemAccess", { enumerable: true, get: function () { return EncryptedMediaImpl_1.requestMediaKeysSystemAccess; } });
Object.defineProperty(exports, "requestMediaKeySystemAccess", { enumerable: true, get: function () { return EncryptedMediaImpl_1.requestMediaKeySystemAccess; } });
const Audio_1 = require("./interface/Audio");
Object.defineProperty(exports, "Audio", { enumerable: true, get: function () { return Audio_1.Audio; } });
const AudioPlayer_1 = require("./interface/AudioPlayer");
Object.defineProperty(exports, "AudioPlayer", { enumerable: true, get: function () { return AudioPlayer_1.AudioPlayer; } });
const SourceBufferImpl_1 = require("./SourceBufferImpl");
Object.defineProperty(exports, "SourceBufferImpl", { enumerable: true, get: function () { return SourceBufferImpl_1.SourceBufferImpl; } });
const MediaErrorImpl_1 = require("./MediaErrorImpl");
Object.defineProperty(exports, "InvalidStateError", { enumerable: true, get: function () { return MediaErrorImpl_1.InvalidStateError; } });
Object.defineProperty(exports, "NotSupportedError", { enumerable: true, get: function () { return MediaErrorImpl_1.NotSupportedError; } });
const EncryptedMediaInterface_1 = require("./interface/EncryptedMediaInterface");
Object.defineProperty(exports, "QuotaExceededError", { enumerable: true, get: function () { return EncryptedMediaInterface_1.QuotaExceededError; } });
const EventTargetInterface_1 = require("./interface/EventTargetInterface");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return EventTargetInterface_1.Event; } });
Object.defineProperty(exports, "TrackEvent", { enumerable: true, get: function () { return EventTargetInterface_1.TrackEvent; } });
const HTMLMediaElement_1 = require("./interface/HTMLMediaElement");
Object.defineProperty(exports, "HTMLAudioElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLAudioElement; } });
Object.defineProperty(exports, "HTMLMediaElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLMediaElement; } });
Object.defineProperty(exports, "HTMLVideoElement", { enumerable: true, get: function () { return HTMLMediaElement_1.HTMLVideoElement; } });
const MediaSource_1 = require("./interface/MediaSource");
Object.defineProperty(exports, "MediaSource", { enumerable: true, get: function () { return MediaSource_1.MediaSource; } });
const MediaSourceInterface_1 = require("./interface/MediaSourceInterface");
Object.defineProperty(exports, "AppendMode", { enumerable: true, get: function () { return MediaSourceInterface_1.AppendMode; } });
Object.defineProperty(exports, "EndOfStreamError", { enumerable: true, get: function () { return MediaSourceInterface_1.EndOfStreamError; } });
Object.defineProperty(exports, "ReadyState", { enumerable: true, get: function () { return MediaSourceInterface_1.ReadyState; } });
// temporary till we import from audio turbo module
const MediaTypes_1 = require("./interface/MediaTypes");
Object.defineProperty(exports, "AudioContentType", { enumerable: true, get: function () { return MediaTypes_1.AudioContentType; } });
Object.defineProperty(exports, "AudioUsageType", { enumerable: true, get: function () { return MediaTypes_1.AudioUsageType; } });
const Video_1 = require("./interface/Video");
Object.defineProperty(exports, "Video", { enumerable: true, get: function () { return Video_1.Video; } });
const VideoPlayer_1 = require("./interface/VideoPlayer");
Object.defineProperty(exports, "VideoPlayer", { enumerable: true, get: function () { return VideoPlayer_1.VideoPlayer; } });
const Media_1 = require("./Media");
Object.defineProperty(exports, "Media", { enumerable: true, get: function () { return Media_1.Media; } });
const W3CLibPolyfill_1 = require("./polyfills/W3CLibPolyfill");
Object.defineProperty(exports, "W3CLibPolyfill", { enumerable: true, get: function () { return W3CLibPolyfill_1.W3CLibPolyfill; } });
const W3CMediaTurboModule_1 = __importDefault(require("./turbo-modules/W3CMediaTurboModule"));
exports.W3CMediaTurboModule = W3CMediaTurboModule_1.default;
const WebCrypto_1 = require("./polyfills/WebCrypto");
Object.defineProperty(exports, "WebCrypto", { enumerable: true, get: function () { return WebCrypto_1.WebCrypto; } });
const TextDecoder_1 = require("./polyfills/TextDecoder");
Object.defineProperty(exports, "TextDecoder", { enumerable: true, get: function () { return TextDecoder_1.TextDecoder; } });
const KeplerMediaControlHandler_1 = require("./interface/KeplerMediaControlHandler");
Object.defineProperty(exports, "KeplerMediaControlHandler", { enumerable: true, get: function () { return KeplerMediaControlHandler_1.KeplerMediaControlHandler; } });
const W3CMediaCapabilitiesImpl_1 = require("./W3CMediaCapabilitiesImpl");
Object.defineProperty(exports, "decodingInfo", { enumerable: true, get: function () { return W3CMediaCapabilitiesImpl_1.decodingInfo; } });
const VTTCue_1 = require("./VTTCue");
Object.defineProperty(exports, "VTTCue", { enumerable: true, get: function () { return VTTCue_1.VTTCue; } });
