// @amazon-devices/react-native-w3cmedia/dist/MediaElementEventRouter.js
"use strict";
/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaElementEventRouter = void 0;
const W3CMediaTurboModule_1 = __importDefault(require("./turbo-modules/W3CMediaTurboModule"));
const LogUtils_1 = require("./LogUtils");
/**
 * NFP Turbo module sends MediaElementEvent as a callback. However,
 * W3C spec expects the events to be emitted from elements like
 * HTMLMediaElement, MediaSource and SourceBuffer.
 * This file receives the event from Turbo module and routes it
 * to registered objects (HTMLMediaElement, MediaSource and SourceBuffer)
 */
class MediaElementEventRouter {
    /**
     * MediaElement Event Sources
     */
    MEDIA_EVENT_SOURCE_HTMLMEDIAELEMENT;
    MEDIA_EVENT_SOURCE_MEDIASOURCE;
    MEDIA_EVENT_SOURCE_SOURCEBUFFER;
    MEDIA_EVENT_SOURCE_ACTIVESOURCEBUFFERS;
    MEDIA_EVENT_SOURCE_AUDIO_TRACKLIST;
    MEDIA_EVENT_SOURCE_VIDEO_TRACKLIST;
    MEDIA_EVENT_SOURCE_TEXT_TRACKLIST;
    mediaElement;
    mediaPlayer;
    mediaSource;
    sourceBuffers;
    constructor(mediaElement) {
        this.mediaElement = mediaElement;
        this.mediaPlayer = null;
        this.mediaSource = null;
        this.sourceBuffers = new Array();
        const mediaEventSource = W3CMediaTurboModule_1.default.getMediaEventSource();
        this.MEDIA_EVENT_SOURCE_ACTIVESOURCEBUFFERS = mediaEventSource.ACTIVESOURCEBUFFERS;
        this.MEDIA_EVENT_SOURCE_AUDIO_TRACKLIST = mediaEventSource.AUDIO_TRACKLIST;
        this.MEDIA_EVENT_SOURCE_HTMLMEDIAELEMENT = mediaEventSource.HTMLMEDIAELEMENT;
        this.MEDIA_EVENT_SOURCE_MEDIASOURCE = mediaEventSource.MEDIASOURCE;
        this.MEDIA_EVENT_SOURCE_SOURCEBUFFER = mediaEventSource.SOURCEBUFFER;
        this.MEDIA_EVENT_SOURCE_VIDEO_TRACKLIST = mediaEventSource.VIDEO_TRACKLIST;
        this.MEDIA_EVENT_SOURCE_TEXT_TRACKLIST = mediaEventSource.TEXT_TRACKLIST;
        this.onMediaElementEvent = this.onMediaElementEvent.bind(this);
        /**
         * Register the event callback to turbo module
         */
        W3CMediaTurboModule_1.default.setEventCallback(this.mediaElement, (event, buffer) => {
            event.buffer = buffer;
            return this.onMediaElementEvent(event);
        });
    }
    /**
     * Register the HTMLMediaElement object  to receive events
     * @param mediaPlayer HTMLMediaElement object that will receives the event
     */
    registerHTMLMediaElement(mediaPlayer) {
        this.mediaPlayer = mediaPlayer;
    }
    /**
     * Register  the MediaSource object to receive events
     * @param mediaSource MediaSource object that will receives the event
     */
    registerMediaSource(mediaSource) {
        this.mediaSource = mediaSource;
    }
    /**
     * Register  the  SourceBuffer object to receive events
     * @param sourceBuffer SourceBuffer object that will receives the event
     */
    registerSourceBuffer(sourceBuffer) {
        const bufferId = sourceBuffer.getId();
        this.sourceBuffers.push(sourceBuffer);
    }
    /**
     * Deregister MediaElement, MediaSource and SourceBuffer
     */
    deregisterMediaElement() {
        this.mediaPlayer = null;
        this.mediaSource = null;
        this.sourceBuffers.splice(0, this.sourceBuffers.length);
    }
    /**
     * Routes events to respective registered components.
     */
    onMediaElementEvent(event) {
        LogUtils_1.LogUtil.debug(`MediaElementEventRouter:onMediaElementEvent: src: ${event.src}, name: ${event.name} \
                trackId: ${event.track_id}, sourceBufferId: ${event.sourcebuffer_id}`);
        switch (event.src) {
            case this.MEDIA_EVENT_SOURCE_SOURCEBUFFER:
                // route to source buffers
                this.sourceBuffers.forEach(sourceBuffer => {
                    if (event.sourcebuffer_id === sourceBuffer.getId()) {
                        sourceBuffer.handleEvent(event);
                    }
                });
                break;
            case this.MEDIA_EVENT_SOURCE_ACTIVESOURCEBUFFERS:
                // route to media source
                this.mediaSource?.handleEvent(event.name);
                break;
            case this.MEDIA_EVENT_SOURCE_TEXT_TRACKLIST:
            case this.MEDIA_EVENT_SOURCE_AUDIO_TRACKLIST:
            case this.MEDIA_EVENT_SOURCE_VIDEO_TRACKLIST:
                // track list could be targeted to a source buffer or media element.
                // identify the target using source buffer id and route accordingly
                // currently native impl doesn't support media element target.
                // will be added when URL playback support is added.
                let eventHandled = false;
                this.sourceBuffers.forEach(sourceBuffer => {
                    if (event.sourcebuffer_id === sourceBuffer.getId()) {
                        sourceBuffer.handleEvent(event);
                        // also mirror this track event to HTMLMediaElement object
                        // Section 4.5.7 Initialization Segment Received
                        // Point: 5.1.2.9: Add new audio track to the audioTracks attribute on the HTMLMediaElement.
                        this.mediaPlayer?.handleEvent(event);
                        eventHandled = true;
                    }
                });
                // if not handled by source buffers above, it means it is for HTMLMediaElement Obj
                // for URL based playback
                if (!eventHandled) {
                    this.mediaPlayer?.handleEvent(event);
                }
                break;
            case this.MEDIA_EVENT_SOURCE_MEDIASOURCE:
                this.mediaSource?.handleEvent(event.name);
                break;
            case this.MEDIA_EVENT_SOURCE_HTMLMEDIAELEMENT:
            default:
                this.mediaPlayer?.handleEvent(event);
                break;
        }
    }
}
exports.MediaElementEventRouter = MediaElementEventRouter;
