// @amazon-devices/react-native-w3cmedia/dist/TextTrackListImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextTrackListImpl = void 0;
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const EventTargetInterface_1 = require("./interface/EventTargetInterface");
const MediaEventEmitterImpl_1 = require("./MediaEventEmitterImpl");
const LogUtils_1 = require("./LogUtils");
const TextTrackImpl_1 = require("./TextTrackImpl");
/**
 * Proxies to support [ ] operator on TextTrackList
 */
const proxyHandlers = {
    get(target, key) {
        if (!(key in target)) {
            return target.textTracks[key];
        }
        else {
            return Reflect.get(target, key);
        }
    }
};
class TextTrackListImpl {
    eventEmitter;
    textTracks = new Array();
    constructor() {
        this.eventEmitter = new MediaEventEmitterImpl_1.MediaEventEmitterImpl();
        return new Proxy(this, proxyHandlers);
    }
    // EventTarget API implementation
    addEventListener(type, listener, options) {
        LogUtils_1.LogUtil.info(`TextTrackListImpl:addEventListener ${type}`);
        this.eventEmitter.addEventListener(type, listener);
    }
    removeEventListener(type, listener, options) {
        LogUtils_1.LogUtil.info(`TextTrackListImpl:removeEventListener ${type}`);
        this.eventEmitter.removeEventListener(type, listener);
    }
    get length() {
        return this.textTracks.length;
    }
    /**
     * https://html.spec.whatwg.org/multipage/media.html#text-track-api
     */
    // FIXME: Seems like a wrong implementation of eventhandler
    set onchange(listener) {
        this.eventEmitter.addEventListener("change", listener);
    }
    // FIXME: Seems like a wrong implementation of eventhandler
    set onaddtrack(listener) {
        this.eventEmitter.addEventListener("addtrack", listener);
    }
    // FIXME: Seems like a wrong implementation of eventhandler
    set onremovetrack(listener) {
        this.eventEmitter.addEventListener("removetrack", listener);
    }
    getTrackById(id) {
        let idx = this.findTrackIndex(id);
        if (idx === -1) {
            LogUtils_1.LogUtil.warn(`TextTrackListImpl: getTrackById ${id} not found`);
            return undefined;
        }
        return this.textTracks[idx];
    }
    // other public APIs
    /**
     * Emits the Track list related events
     * @param eventName name of the track
     * @param textTrack track
     */
    emitEvent(eventName, textTrack) {
        LogUtils_1.LogUtil.debug(`TextTrackListImpl: emitEvent ${eventName} in track: ${textTrack} `);
        if (eventName === "change") {
            this.eventEmitter.emit(eventName, new EventTargetInterface_1.Event(eventName, this));
        }
        else {
            if (textTrack) {
                //@TODO PROJXXXX-192523
                this.eventEmitter.emit(eventName, new EventTargetInterface_1.Event(eventName, this));
                //this.eventEmitter.emit(eventName, new TrackEvent(eventName, this, textTrack));
            }
            else {
                this.eventEmitter.emit(eventName, new EventTargetInterface_1.Event(eventName, this));
            }
        }
    }
    /**
     * Adds a track to the tracks list
     */
    addTrack(track) {
        LogUtils_1.LogUtil.debug(`TextTrackListImpl: addTrack id ${track.id}`);
        this.textTracks.push(track);
        /**
         * List should fire the events when tracks are added
         * For Native tracks handleTextTrackListEvent does that
         */
        if (track.id.startsWith(TextTrackImpl_1.TextTrackImpl.JS_TRACK_ID_PREFIX)) {
            LogUtils_1.LogUtil.debug(`TextTrackListImpl: Inform listeners for TextTrack id ${track.id} addition`);
            this.emitEvent("addtrack", track);
        }
    }
    removeTrack(id) {
        LogUtils_1.LogUtil.debug(`TextTrackListImpl: removeTrack id ${id}`);
        let trackIndex = this.findTrackIndex(id);
        if (trackIndex === -1) {
            LogUtils_1.LogUtil.warn(`TextTrackListImpl: removeTrack ${id} not found`);
            return;
        }
        let removedTrack = this.textTracks.splice(trackIndex, 1); //remove element from array
        /**
         * List should fire the events when tracks are removed
         * For Native tracks handleTextTrackListEvent does that
         * For JS track this is unexpected.
         */
        if (removedTrack.length > 0 && removedTrack[0].id.startsWith(TextTrackImpl_1.TextTrackImpl.JS_TRACK_ID_PREFIX)) {
            LogUtils_1.LogUtil.warn(`TextTrackListImpl: unexpected removal of TextTrack id ${removedTrack[0].id}`);
            this.emitEvent("removetrack", removedTrack[0]);
        }
    }
    findTrackIndex(id) {
        return this.textTracks.findIndex(tt => tt.id === id); //find index in your array
    }
}
exports.TextTrackListImpl = TextTrackListImpl;
