// @amazon-devices/react-native-w3cmedia/dist/VideoTrackListImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoTrackListImpl = void 0;
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
const EventTargetInterface_1 = require("./interface/EventTargetInterface");
const LogUtils_1 = require("./LogUtils");
const MediaEventEmitterImpl_1 = require("./MediaEventEmitterImpl");
/**
 * Proxies to support [ ] operator on VideoTrackList
 */
const proxyHandlers = {
    get(target, key) {
        if (!(key in target)) {
            return target.videoTracks[key];
        }
        else {
            return Reflect.get(target, key);
        }
    }
};
/**
 * Implement VideoTrackList interface
 */
class VideoTrackListImpl {
    eventEmitter;
    videoTracks = new Array();
    constructor() {
        this.eventEmitter = new MediaEventEmitterImpl_1.MediaEventEmitterImpl();
        return new Proxy(this, proxyHandlers);
    }
    // EventTarget API implementation
    addEventListener(type, listener, options) {
        LogUtils_1.LogUtil.info(`VideoTrackListImpl:addEventListener ${type}`);
        this.eventEmitter.addEventListener(type, listener);
    }
    removeEventListener(type, listener, options) {
        LogUtils_1.LogUtil.info(`VideoTrackListImpl:removeEventListener ${type}`);
        this.eventEmitter.removeEventListener(type, listener);
    }
    // VideoTrackList interface API implementation
    get length() {
        return this.videoTracks.length;
    }
    get selectedIndex() {
        return 0;
    }
    set onchange(listener) {
        this.eventEmitter.addEventListener("change", listener);
    }
    set onaddtrack(listener) {
        this.eventEmitter.addEventListener("addtrack", listener);
    }
    set onremovetrack(listener) {
        this.eventEmitter.addEventListener("removetrack", listener);
    }
    getTrackById(id) {
        if (this.videoTracks.length === 0) {
            LogUtils_1.LogUtil.warn(`VideoTrackListImpl: no video tracks available`);
            return undefined;
        }
        let track = undefined;
        this.videoTracks.forEach((videoTrack) => {
            if (videoTrack.id === id) {
                track = videoTrack;
            }
        });
        return track;
    }
    // other public APIs
    /**
     * Emits the Track list related events
     * @param eventName name of the track
     * @param videoTrack track
     */
    emitEvent(eventName, videoTrack) {
        LogUtils_1.LogUtil.debug(`VideoTrackListImpl: emitEvent ${eventName}`);
        if (eventName === "change") {
            this.eventEmitter.emit(eventName, new EventTargetInterface_1.Event(eventName, this));
        }
        else {
            if (videoTrack) {
                this.eventEmitter.emit(eventName, new EventTargetInterface_1.TrackEvent(eventName, this, videoTrack));
            }
            else {
                this.eventEmitter.emit(eventName, new EventTargetInterface_1.Event(eventName, this));
            }
        }
    }
    /**
     * Adds a track to the tracks list
     */
    addTrack(track) {
        LogUtils_1.LogUtil.debug(`VideoTrackListImpl: addTrack id ${track.id}`);
        this.videoTracks.push(track);
    }
    removeTrack(id) {
        LogUtils_1.LogUtil.debug(`VideoTrackListImpl: removeTrack id ${id}`);
        // find the track
        let trackIndex = undefined;
        this.videoTracks.forEach((videoTrack, index) => {
            if (videoTrack.id === id) {
                trackIndex = index;
            }
        });
        // remove track if found
        if (trackIndex !== undefined) {
            this.videoTracks.splice(trackIndex, 1);
            LogUtils_1.LogUtil.debug(`video track id ${id} removed`);
        }
        else {
            LogUtils_1.LogUtil.warn(`video track id ${id} not found`);
        }
    }
}
exports.VideoTrackListImpl = VideoTrackListImpl;
