// @amazon-devices/react-native-w3cmedia/dist/VTTCue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VTTCue = void 0;
const LogUtils_1 = require("./LogUtils");
const EventTargetInterface_1 = require("./interface/EventTargetInterface");
const MediaEventEmitterImpl_1 = require("./MediaEventEmitterImpl");
// @CHECK should we move this to src/polyfills/VTTCue.ts ??
/* [Exposed=Window,
 Constructor(double startTime, double endTime, DOMString text)]
 */
/**
 * @hidden
 * https://www.w3.org/TR/webvtt1/#the-vttcue-interface
 */
class VTTCue {
    _track;
    _id;
    _startTime;
    _endTime;
    // @TODO implement pauseOnExit
    _pauseOnExit;
    /*
     * https://www.w3.org/TR/webvtt1/#vttregion
     */
    _region;
    _vertical;
    _snapToLines;
    _line;
    _lineAlign;
    _position;
    _positionAlign;
    _size;
    _align;
    _text;
    _isActive;
    eventEmitter;
    static cueId = 0n;
    /**
     * @track is a readonly property, can not define a setter.
     * Users do not set this while constructing the VTTCue.
     */
    constructor(startTime, endTime, text, track) {
        this._track = track || null;
        this._startTime = startTime;
        this._endTime = endTime;
        this._text = text;
        this._id = VTTCue.cueId.toString();
        VTTCue.cueId++;
        this._pauseOnExit = false;
        this._vertical = '';
        this._snapToLines = true;
        this._line = 'auto';
        this._lineAlign = 'start';
        this._position = 'auto';
        this._positionAlign = 'auto';
        this._size = 100;
        this._align = 'center';
        this._region = null;
        this._isActive = false;
        this.eventEmitter = new MediaEventEmitterImpl_1.MediaEventEmitterImpl();
        LogUtils_1.LogUtil.debug('VTTCue constructed with ' + startTime + ' - ' + endTime + ' : ' + text);
    }
    get track() {
        return this._track;
    }
    /**
     * Not a usual JS setter
     * @param track
     */
    setTrack(track) {
        this._track = track;
    }
    get id() {
        return this._id;
    }
    set id(cId) {
        this._id = cId;
    }
    get startTime() {
        return this._startTime;
    }
    set startTime(cStartTime) {
        this._startTime = cStartTime;
        this.track?.refresh();
    }
    get endTime() {
        return this._endTime;
    }
    set endTime(cEndTime) {
        this._endTime = cEndTime;
        this.track?.refresh();
    }
    get pauseOnExit() {
        return this._pauseOnExit;
    }
    set pauseOnExit(aPauseOnExit) {
        LogUtils_1.LogUtil.warn(`VTTCue: pauseOnExit is not supported.`);
        this._pauseOnExit = aPauseOnExit;
    }
    onenter = (e) => {
        LogUtils_1.LogUtil.debug(`VTTCue ${this._id}: Default handler event : ` + e?.type);
    };
    onexit = (e) => {
        LogUtils_1.LogUtil.debug(`VTTCue ${this._id}: Default handler event : ` + e?.type);
    };
    get vertical() {
        return this._vertical;
    }
    set vertical(aVerticalDirection) {
        this._vertical = aVerticalDirection;
    }
    get snapToLines() {
        return this._snapToLines;
    }
    set snapToLines(aSnapToLines) {
        this._snapToLines = aSnapToLines;
    }
    /*
      attribute  align;
      attribute DOMString text;*/
    /*
     * @TODO implement
     */
    // DocumentFragment getCueAsHTML();
    get line() {
        return this._line;
    }
    set line(aLineAndPositionSetting) {
        this._line = aLineAndPositionSetting;
    }
    get lineAlign() {
        return this._lineAlign;
    }
    set lineAlign(aLineAlign) {
        this._lineAlign = aLineAlign;
    }
    get position() {
        return this._position;
    }
    set position(aPosition) {
        this._position = aPosition;
    }
    get positionAlign() {
        return this._positionAlign;
    }
    set positionAlign(aPositionAlign) {
        this._positionAlign = aPositionAlign;
    }
    get size() {
        return this._size;
    }
    set size(aSize) {
        this._size = aSize;
    }
    get align() {
        return this._align;
    }
    set align(aAlign) {
        this._align = aAlign;
    }
    get text() {
        return this._text;
    }
    set text(aText) {
        this._text = aText;
    }
    get isActive() {
        return this._isActive;
    }
    set isActive(aActive) {
        this._isActive = aActive;
    }
    get region() {
        return this._region;
    }
    set region(aRegion) {
        this._region = aRegion;
    }
    // EventTarget API implementation
    addEventListener(type, listener, options) {
        LogUtils_1.LogUtil.debug(`VTTCue:addEventListener ${type}`);
        this.eventEmitter.addEventListener(type, listener);
    }
    removeEventListener(type, listener, options) {
        LogUtils_1.LogUtil.debug(`VTTCue:removeEventListener ${type}`);
        this.eventEmitter.removeEventListener(type, listener);
    }
    dispatchEvent(event) {
        LogUtils_1.LogUtil.debug(`AudioTrackListImpl: emitEvent ${event.type}`);
        /**
         * If we register these as listener with the emitter, we need not do this
         * @CHECK should we schedule these using setTimeout?
         * We dispatch these asynchronously from TMOA, so it seems safe.
         **/
        if (event.type === 'enter') {
            this.onenter(event);
        }
        else if (event.type === 'exit') {
            this.onexit(event);
        }
        return this.eventEmitter.emit(event.type, new EventTargetInterface_1.Event(event.type, this));
    }
}
exports.VTTCue = VTTCue;
