// @amzn/react-native-kepler/Libraries/AlexaManager/AlexaManagerTypes.js
"use strict";
/*
* Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
*
* AMAZON PROPRIETARY/CONFIDENTIAL
*
* You may not use this file except in compliance with the terms and
* conditions set forth in the accompanying LICENSE.TXT file.
*
* THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
* DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
* IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlexaErrorResponseType = exports.AlexaStateChangeCauseType = exports.AlexaManagerStatus = void 0;
// AlexaManager Status
var AlexaManagerStatus;
(function (AlexaManagerStatus) {
    // Request was successful
    AlexaManagerStatus[AlexaManagerStatus["SUCCESS"] = 0] = "SUCCESS";
    // One or more request parameters was invalid
    AlexaManagerStatus[AlexaManagerStatus["BAD_PARAM_ERROR"] = 1] = "BAD_PARAM_ERROR";
    // Internal error occured handling request
    AlexaManagerStatus[AlexaManagerStatus["GENERAL_ERROR"] = 2] = "GENERAL_ERROR";
})(AlexaManagerStatus = exports.AlexaManagerStatus || (exports.AlexaManagerStatus = {}));
var AlexaStateChangeCauseType;
(function (AlexaStateChangeCauseType) {
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["ALEXA_INTERACTION"] = 0] = "ALEXA_INTERACTION";
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["APP_INTERACTION"] = 1] = "APP_INTERACTION";
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["PHYSICAL_INTERACTION"] = 2] = "PHYSICAL_INTERACTION";
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["PERIODIC_POLL"] = 3] = "PERIODIC_POLL";
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["RULE_TRIGGER"] = 4] = "RULE_TRIGGER";
    AlexaStateChangeCauseType[AlexaStateChangeCauseType["VOICE_INTERACTION"] = 5] = "VOICE_INTERACTION";
})(AlexaStateChangeCauseType = exports.AlexaStateChangeCauseType || (exports.AlexaStateChangeCauseType = {}));
var AlexaErrorResponseType;
(function (AlexaErrorResponseType) {
    // The operation can't be performed because the endpoint is already in operation.
    AlexaErrorResponseType[AlexaErrorResponseType["ALREADY_IN_OPERATION"] = 0] = "ALREADY_IN_OPERATION";
    // The bridge is unreachable or offline.
    AlexaErrorResponseType[AlexaErrorResponseType["BRIDGE_UNREACHABLE"] = 1] = "BRIDGE_UNREACHABLE";
    // The endpoint can't handle the directive because it is performing another action.
    AlexaErrorResponseType[AlexaErrorResponseType["ENDPOINT_BUSY"] = 2] = "ENDPOINT_BUSY";
    // The endpoint can't handle the directive because the battery power is too low.
    AlexaErrorResponseType[AlexaErrorResponseType["ENDPOINT_LOW_POWER"] = 3] = "ENDPOINT_LOW_POWER";
    // The endpoint is unreachable or offline.
    AlexaErrorResponseType[AlexaErrorResponseType["ENDPOINT_UNREACHABLE"] = 4] = "ENDPOINT_UNREACHABLE";
    // The authorization credential provided by Alexa has expired.
    AlexaErrorResponseType[AlexaErrorResponseType["EXPIRED_AUTHORIZATION_CREDENTIAL"] = 5] = "EXPIRED_AUTHORIZATION_CREDENTIAL";
    // The endpoint can't handle the directive because it's firmware is out of date.
    AlexaErrorResponseType[AlexaErrorResponseType["FIRMWARE_OUT_OF_DATE"] = 6] = "FIRMWARE_OUT_OF_DATE";
    // The endpoint can't handle the directive because it has experienced a hardware malfunction.
    AlexaErrorResponseType[AlexaErrorResponseType["HARDWARE_MALFUNCTION"] = 7] = "HARDWARE_MALFUNCTION";
    // AVS does not have permissions to perform the specified action on the endpoint.
    AlexaErrorResponseType[AlexaErrorResponseType["INSUFFICIENT_PERMISSIONS"] = 8] = "INSUFFICIENT_PERMISSIONS";
    // An error occurred that can't be described by one of the other error types.
    AlexaErrorResponseType[AlexaErrorResponseType["INTERNAL_ERROR"] = 9] = "INTERNAL_ERROR";
    // The authorization credential provided by Alexa is invalid.
    AlexaErrorResponseType[AlexaErrorResponseType["INVALID_AUTHORIZATION_CREDENTIAL"] = 10] = "INVALID_AUTHORIZATION_CREDENTIAL";
    // The directive is not supported or is malformed.
    AlexaErrorResponseType[AlexaErrorResponseType["INVALID_DIRECTIVE"] = 11] = "INVALID_DIRECTIVE";
    // The directive contains a value that is not valid for the target endpoint.
    AlexaErrorResponseType[AlexaErrorResponseType["INVALID_VALUE"] = 12] = "INVALID_VALUE";
    // The endpoint does not exist, or no longer exists.
    AlexaErrorResponseType[AlexaErrorResponseType["NO_SUCH_ENDPOINT"] = 13] = "NO_SUCH_ENDPOINT";
    // The endpoint can't handle the directive because it is in a calibration phase, such as warming up.
    AlexaErrorResponseType[AlexaErrorResponseType["NOT_CALIBRATED"] = 14] = "NOT_CALIBRATED";
    // The endpoint can't be set to the specified value because of its current mode of operation.
    AlexaErrorResponseType[AlexaErrorResponseType["NOT_SUPPORTED_IN_CURRENT_MODE"] = 15] = "NOT_SUPPORTED_IN_CURRENT_MODE";
    // The endpoint is not in operation.
    AlexaErrorResponseType[AlexaErrorResponseType["NOT_IN_OPERATION"] = 16] = "NOT_IN_OPERATION";
    // The endpoint can't handle the directive because it doesn't support the requested power level.
    AlexaErrorResponseType[AlexaErrorResponseType["POWER_LEVEL_NOT_SUPPORTED"] = 17] = "POWER_LEVEL_NOT_SUPPORTED";
    // The maximum rate at which an endpoint or bridge can process directives has been exceeded.
    AlexaErrorResponseType[AlexaErrorResponseType["RATE_LIMIT_EXCEEDED"] = 18] = "RATE_LIMIT_EXCEEDED";
    // The endpoint can't be set to the specified value because it's outside the acceptable temperature range.
    AlexaErrorResponseType[AlexaErrorResponseType["TEMPERATURE_VALUE_OUT_OF_RANGE"] = 19] = "TEMPERATURE_VALUE_OUT_OF_RANGE";
    // The endpoint can't be set to the specified value because it's outside the acceptable range.
    AlexaErrorResponseType[AlexaErrorResponseType["VALUE_OUT_OF_RANGE"] = 20] = "VALUE_OUT_OF_RANGE";
})(AlexaErrorResponseType = exports.AlexaErrorResponseType || (exports.AlexaErrorResponseType = {}));
