// @amazon-devices/react-native-w3cmedia/dist/components/mediacontrols/SeekBarUI/BaseUtilities.js
"use strict";
/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.base = exports.baseTokensDark = exports.baseTokensLight = exports.baseColorTokensDark = exports.baseColorTokensLight = exports.baseSizes = void 0;
exports.baseSizes = {
    minimumTouchArea: 18,
    spacing: {
        xxs: 2,
        xs: 4,
        sm: 8,
        md: 12,
        lg: 16,
        xl: 20,
        xxl: 24
    }
};
exports.baseColorTokensLight = {
    black: '#000000',
    white: '#FFFFFF',
    shadow: '#000000',
    gray: {
        100: '#F4F4F5',
        200: '#E1E1E2',
        300: '#D0D0D4',
        400: '#A1A1AA',
        500: '#7A7A83',
        600: '#52525B',
        700: '#3F3F46',
        800: '#27272A',
        900: '#202024'
    },
    red: {
        100: '#F5BDB9',
        200: '#F19D97',
        300: '#EA7168',
        400: '#E6554A',
        500: '#E02B1D',
        600: '#CC271A',
        700: '#9F1F15',
        800: '#7B1810',
        900: '#5E120C'
    },
    darkgray: {
        100: '#BEBFC0',
        200: '#9EA0A2',
        300: '#727477',
        400: '#57595D',
        500: '#2D3034',
        600: '#292C2F',
        700: '#202225',
        800: '#191A1D',
        900: '#131416'
    },
    lightgray: {
        100: '#F9F9F9',
        200: '#F2F2F2',
        300: '#ECECEC',
        400: '#E5E5E6',
        500: '#DFDFE0',
        600: '#D8D9D9',
        700: '#D2D2D3',
        800: '#CBCCCD',
        900: '#C5C5C6'
    }
};
exports.baseColorTokensDark = {
    black: '#000000',
    white: '#FFFFFF',
    shadow: '#000000',
    gray: {
        100: '#333333',
        200: '#3D3D3D',
        300: '#5F5F5F',
        400: '#747474',
        500: '#878787',
        600: '#9B9B9B',
        700: '#B9B9B9',
        800: '#D7D7D7',
        900: '#EFEFEF'
    },
    red: {
        100: '#F5BDB9',
        200: '#F19D97',
        300: '#EA7168',
        400: '#E6554A',
        500: '#E02B1D',
        600: '#CC271A',
        700: '#9F1F15',
        800: '#7B1810',
        900: '#5E120C'
    },
    darkgray: {
        100: '#BEBFC0',
        200: '#9EA0A2',
        300: '#727477',
        400: '#57595D',
        500: '#2D3034',
        600: '#292C2F',
        700: '#202225',
        800: '#191A1D',
        900: '#131416'
    },
    lightgray: {
        100: '#F9F9F9',
        200: '#F2F2F2',
        300: '#ECECEC',
        400: '#E5E5E6',
        500: '#DFDFE0',
        600: '#D8D9D9',
        700: '#D2D2D3',
        800: '#CBCCCD',
        900: '#C5C5C6'
    }
};
exports.baseTokensLight = {
    size: exports.baseSizes,
    color: exports.baseColorTokensLight
};
exports.baseTokensDark = {
    size: exports.baseSizes,
    color: exports.baseColorTokensDark
};
exports.base = {
    light: exports.baseTokensLight,
    dark: exports.baseTokensDark
};
