// @amazon-devices/react-native-kepler/Libraries/Logger/kplogger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KPLogger = exports.kpLogToNative = void 0;
const kpnative_logger_1 = __importDefault(require("./kpnative-logger"));
/**
 * @brief A log implementation that logs messages through the KPNativeLogger turbomodule
 */
const kpLogToNative = (level, message) => kpnative_logger_1.default.log(level, message);
exports.kpLogToNative = kpLogToNative;
/**
 * @brief Logging methods for application wide use.
 *
 * @discussion Create an instance of this class to log messages with a
 * specified prefix tag and KPLogImplementation.
 */
class KPLogger {
    /**
     * @brief KPLogger constructor
     *
     * @param tag A tag to be inserted before formatted strings to be logged.
     * @param implementation A custom KPLogFunction to handle logging messages.
     * If null is passed KPLogToNative will be used. Custom log implementations can
     * call KPLogToNative() or any other custom log implementation to log to
     * multiple implementations if desired.
     */
    constructor(tag = '', implementation = null) {
        this._tag = tag;
        if (implementation) {
            this._logImplementation = implementation;
        }
        else {
            this._logImplementation = exports.kpLogToNative;
        }
    }
    /**
     * @brief Log a message with the debug log level
     * @param format A string with format arguments to be logged
     * @param args Arguments to supply the formatting requirements passed in format
     */
    debug(message, ...args) {
        if (this._logImplementation == null) {
            return;
        }
        try {
            const formattedMessage = this.format(message, ...args);
            const taggedMessage = this.tagMessage(formattedMessage);
            void this._logImplementation('debug', taggedMessage);
        }
        catch (e) {
            // trap and ignore logging exceptions... can't log it
        }
    }
    /**
     * @brief Log an information message
     * @param format A string with format arguments to be logged
     * @param args Arguments to supply the formatting requirements passed in format
     */
    info(message, ...args) {
        if (this._logImplementation == null) {
            return;
        }
        try {
            const formattedMessage = this.format(message, ...args);
            const taggedMessage = this.tagMessage(formattedMessage);
            void this._logImplementation('info', taggedMessage);
        }
        catch (e) {
            // trap and ignore logging exceptions... can't log it
        }
    }
    /**
     * @brief Log a warning
     * @param format A string with format arguments to be logged
     * @param args Arguments to supply the formatting requirements passed in format
     */
    warn(message, ...args) {
        if (this._logImplementation == null) {
            return;
        }
        try {
            const formattedMessage = this.format(message, ...args);
            const taggedMessage = this.tagMessage(formattedMessage);
            void this._logImplementation('warn', taggedMessage);
        }
        catch (e) {
            // trap and ignore logging exceptions... can't log it
        }
    }
    /**
     * @brief Log an error message
     * @param format A string with format arguments to be logged
     * @param args Arguments to supply the formatting requirements passed in format
     */
    error(message, ...args) {
        if (this._logImplementation == null) {
            return;
        }
        try {
            const formattedMessage = this.format(message, ...args);
            const taggedMessage = this.tagMessage(formattedMessage);
            void this._logImplementation('error', taggedMessage);
        }
        catch (e) {
            // trap and ignore logging exceptions... can't log it
        }
    }
    tagMessage(message) {
        return this._tag ? `[${this._tag}] ${message}` : `${message}`;
    }
    // FIXME: This formatting code is a placeholder to ensure that a common interface is followed
    // on the native and the JS side. It should eventually be replaced with a library that handles
    // all FMT formats.
    format(message, ...args) {
        const isUsed = [];
        let formattedString = message;
        let tempString;
        for (let k = 0; k < args.length; k++) {
            tempString = formattedString.replace('{' + k + '}', args[k]);
            if (tempString !== formattedString) {
                isUsed[k] = true;
                formattedString = tempString;
            }
        }
        for (let k = 0; k < args.length; k++) {
            if (!isUsed[k]) {
                formattedString = formattedString.replace('{}', args[k]);
            }
        }
        return formattedString;
    }
}
exports.KPLogger = KPLogger;
