// @amzn/react-native-w3cmedia/dist/AudioTrackImpl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioTrackImpl = void 0;
const W3CMediaTurboModule_1 = __importDefault(require("./turbo-modules/W3CMediaTurboModule"));
class AudioTrackImpl {
    audioTrack;
    mediaElement;
    constructor(audioTrack, mediaElement) {
        this.audioTrack = audioTrack;
        this.mediaElement = mediaElement;
    }
    get id() {
        return this.audioTrack.id;
    }
    get kind() {
        return this.audioTrack.kind;
    }
    get label() {
        return this.audioTrack.label;
    }
    get language() {
        return this.audioTrack.language;
    }
    get enabled() {
        const audioTrack = W3CMediaTurboModule_1.default.getAudioTrackById(this.mediaElement, this.id);
        return audioTrack?.enabled;
    }
    set enabled(enabled) {
        if (this.enabled !== enabled) {
            if (enabled) {
                W3CMediaTurboModule_1.default.enableAudioTrack(this.mediaElement, this.id);
            }
            else {
                W3CMediaTurboModule_1.default.disableAudioTrack(this.mediaElement, this.id);
            }
            this.audioTrack.enabled = enabled;
        }
    }
}
exports.AudioTrackImpl = AudioTrackImpl;
