// @amzn/react-native-w3cmedia/dist/LogUtils.js
"use strict";
/*
 * Copyright 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogUtil = exports.LogLevel = void 0;
/**
 * Log levels for controlling verbosity of logging
 */
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["NONE"] = 0] = "NONE";
    LogLevel[LogLevel["ERROR"] = 1] = "ERROR";
    LogLevel[LogLevel["WARN"] = 2] = "WARN";
    LogLevel[LogLevel["INFO"] = 3] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 4] = "DEBUG";
    LogLevel[LogLevel["VERBOSE"] = 5] = "VERBOSE"; // All logs
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
/**
 * Simple utility class for logging with different levels
 */
class LogUtil {
    // Current log level
    static currentLogLevel = LogLevel.INFO;
    /**
     * Set the log level
     * @param level The log level to set
     */
    static setLogLevel(level) {
        if (level >= LogLevel.NONE && level <= LogLevel.VERBOSE) {
            this.currentLogLevel = level;
        }
    }
    /**
     * Get the current log level
     * @returns The current log level
     */
    static getLogLevel() {
        return this.currentLogLevel;
    }
    /**
     * Get the current log level as a string
     * @returns The current log level name
     */
    static getCurrentLogLevelName() {
        switch (this.currentLogLevel) {
            case LogLevel.NONE: return "NONE";
            case LogLevel.ERROR: return "ERROR";
            case LogLevel.WARN: return "WARN";
            case LogLevel.INFO: return "INFO";
            case LogLevel.DEBUG: return "DEBUG";
            case LogLevel.VERBOSE: return "VERBOSE";
            default: return "UNKNOWN";
        }
    }
    /**
     * Log an error message if error logging is enabled
     * @param message The message to log
     * @param optionalParams Additional parameters to log
     */
    static error(message, ...optionalParams) {
        if (this.currentLogLevel >= LogLevel.ERROR) {
            console.error(`${message}`, ...optionalParams);
        }
    }
    /**
     * Log a warning message if warning logging is enabled
     * @param message The message to log
     * @param optionalParams Additional parameters to log
     */
    static warn(message, ...optionalParams) {
        if (this.currentLogLevel >= LogLevel.WARN) {
            console.warn(`${message}`, ...optionalParams);
        }
    }
    /**
     * Log an info message if info logging is enabled
     * @param message The message to log
     * @param optionalParams Additional parameters to log
     */
    static info(message, ...optionalParams) {
        if (this.currentLogLevel >= LogLevel.INFO) {
            console.info(`${message}`, ...optionalParams);
        }
    }
    /**
     * Log a debug message if debug logging is enabled
     * @param message The message to log
     * @param optionalParams Additional parameters to log
     */
    static debug(message, ...optionalParams) {
        if (this.currentLogLevel >= LogLevel.DEBUG) {
            console.debug(`${message}`, ...optionalParams);
        }
    }
    /**
     * Log a verbose message if verbose logging is enabled
     * @param message The message to log
     * @param optionalParams Additional parameters to log
     */
    static verbose(message, ...optionalParams) {
        if (this.currentLogLevel >= LogLevel.VERBOSE) {
            console.log(`${message}`, ...optionalParams);
        }
    }
}
exports.LogUtil = LogUtil;
