// @amzn/react-native-w3cmedia/dist/MediaErrorImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuotaExceededError = exports.NotSupportedError = exports.InvalidStateError = exports.MediaErrorImpl = void 0;
class MediaErrorImpl {
    code_;
    message_;
    constructor(code, message) {
        this.code_ = code;
        this.message_ = message;
    }
    get code() {
        return this.code_;
    }
    get message() {
        return this.message_;
    }
    get MEDIA_ERR_ABORTED() {
        return 1;
    }
    get MEDIA_ERR_DECODE() {
        return 3;
    }
    get MEDIA_ERR_NETWORK() {
        return 2;
    }
    get MEDIA_ERR_SRC_NOT_SUPPORTED() {
        return 4;
    }
}
exports.MediaErrorImpl = MediaErrorImpl;
class InvalidStateError extends Error {
    message;
    name = "InvalidStateError";
    code;
    constructor(message) {
        super(message);
        this.message = message;
        this.code = 11;
    }
}
exports.InvalidStateError = InvalidStateError;
class NotSupportedError extends Error {
    message;
    name = "NotSupportedError";
    code;
    constructor(message) {
        super(message);
        this.message = message;
        this.code = 9;
    }
}
exports.NotSupportedError = NotSupportedError;
class QuotaExceededError extends Error {
    message;
    name = "QuotaExceededError";
    code;
    constructor(message) {
        super(message);
        this.message = message;
        this.code = 22;
    }
}
exports.QuotaExceededError = QuotaExceededError;
