// @amzn/react-native-w3cmedia/dist/VideoTrackImpl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoTrackImpl = void 0;
const W3CMediaTurboModule_1 = __importDefault(require("./turbo-modules/W3CMediaTurboModule"));
class VideoTrackImpl {
    videoTrack;
    mediaElement;
    constructor(videoTrack, mediaElement) {
        this.videoTrack = videoTrack;
        this.mediaElement = mediaElement;
    }
    get id() {
        return this.videoTrack.id;
    }
    get kind() {
        return this.videoTrack.kind;
    }
    get label() {
        return this.videoTrack.label;
    }
    get language() {
        return this.videoTrack.language;
    }
    get selected() {
        const videoTrack = W3CMediaTurboModule_1.default.getVideoTrackById(this.mediaElement, this.id);
        return videoTrack?.selected;
    }
    set selected(selected) {
        if (this.selected !== selected) {
            if (selected) {
                W3CMediaTurboModule_1.default.enableVideoTrack(this.mediaElement, this.id);
            }
            else {
                W3CMediaTurboModule_1.default.disableVideoTrack(this.mediaElement, this.id);
            }
            this.videoTrack.selected = selected;
        }
    }
}
exports.VideoTrackImpl = VideoTrackImpl;
