// @amzn/react-native-w3cmedia/dist/VTTRegion.js
"use strict";
/*
 * Copyright 2024-2025 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VTTRegion = void 0;
/**
 * @hidden
 * https://w3c.github.io/webvtt/#webvtt-region
 */
class VTTRegion {
    _id;
    _width;
    _lines;
    _regionAnchorX;
    _regionAnchorY;
    _viewportAnchorX;
    _viewportAnchorY;
    _scroll;
    constructor() {
        this._id = '';
        this._width = 100;
        this._lines = 3;
        this._regionAnchorX = 0;
        this._regionAnchorY = 100;
        this._viewportAnchorX = 0;
        this._viewportAnchorY = 100;
        this._scroll = 'None';
    }
    get id() {
        return this._id;
    }
    get width() {
        return this._width;
    }
    set width(aWidth) {
        this._width = aWidth;
    }
    get lines() {
        return this._lines;
    }
    set lines(aLines) {
        this._lines = aLines;
    }
    get regionAnchorX() {
        return this._regionAnchorX;
    }
    set regionAnchorX(aRegionAnchorX) {
        this._regionAnchorX = aRegionAnchorX;
    }
    get regionAnchorY() {
        return this._regionAnchorY;
    }
    set regionAnchorY(aRegionAnchorY) {
        this._regionAnchorY = aRegionAnchorY;
    }
    get viewportAnchorX() {
        return this._viewportAnchorX;
    }
    set viewportAnchorX(aViewportAnchorX) {
        this._viewportAnchorX = aViewportAnchorX;
    }
    get viewportAnchorY() {
        return this._viewportAnchorY;
    }
    set viewportAnchorY(aViewportAnchorY) {
        this._viewportAnchorY = aViewportAnchorY;
    }
    get scroll() {
        return this._scroll;
    }
    set scroll(aScroll) {
        this._scroll = aScroll;
    }
}
exports.VTTRegion = VTTRegion;
