// @amazon-devices/react-native-kepler/Libraries/FocusManager/FocusManager.js
/*
 * Copyright (c) 2021 - 2023 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FocusManager = void 0;
const FocusManagerTurboModuleSpec_1 = __importDefault(require("./FocusManagerTurboModuleSpec"));
const tagForComponentOrHandle_1 = require("./tagForComponentOrHandle");
class FocusManagerLibrary {
    getFocused() {
        let focusedTag = FocusManagerTurboModuleSpec_1.default.getFocused();
        if (focusedTag === -1) {
            return undefined;
        }
        return focusedTag;
    }
    setNextFocus(fromTag, toTag, direction) {
        FocusManagerTurboModuleSpec_1.default.setNextFocus(tagForComponentOrHandle_1.tagForComponentOrHandle(fromTag), tagForComponentOrHandle_1.tagForComponentOrHandle(toTag), direction);
    }
    clearNextFocus(fromTag, direction = 'all') {
        FocusManagerTurboModuleSpec_1.default.clearNextFocus(tagForComponentOrHandle_1.tagForComponentOrHandle(fromTag), direction);
    }
    setFocusRoot(tag, isFocusRoot) {
        FocusManagerTurboModuleSpec_1.default.setFocusRoot(tagForComponentOrHandle_1.tagForComponentOrHandle(tag), isFocusRoot);
    }
    focus(tag) {
        FocusManagerTurboModuleSpec_1.default.focus(tagForComponentOrHandle_1.tagForComponentOrHandle(tag));
    }
    blur(tag) {
        FocusManagerTurboModuleSpec_1.default.blur(tagForComponentOrHandle_1.tagForComponentOrHandle(tag));
    }
}
exports.FocusManager = new FocusManagerLibrary();
exports.default = exports.FocusManager;
