// @amazon-devices/react-native-w3cmedia/dist/components/mediacontrols/SeekBarUI/ProgressBar/ProgressBar.js
"use strict";
/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
//Referenced {SeekBar,ProgressBar} folder from  - https://tiny.amazon.com/10ftogofw/KeplerComponentLibrary
// @ts-nocheck
const react_1 = __importStar(require("react"));
const react_native_1 = require("react-native");
const ThemeUtilities_1 = require("../Theme/ThemeUtilities");
const Indicator_1 = require("./Indicator");
const tokenResolver_progressbar_1 = require("./tokenResolver.progressbar");
/**
 * ProgressBar is a component that visualizes the status and progress of a given task.
 * If the progress is not specified or cannot be determined, an indeterminate indicator is shown.
 *
 * @returns ProgressBar component
 */
const ProgressBar = ({ progress, animationDuration, indicatorColor, trackColor, size = 'md', barStyle, indicatorStyle, ...rest }) => {
    const theme = (0, ThemeUtilities_1.useTheme)();
    const animatedProgress = (0, react_1.useRef)(new react_native_1.Animated.Value(progress || 0)).current;
    const [barWidth, setBarWidth] = (0, react_1.useState)(0);
    const isIndeterminate = progress === undefined;
    const accessibilityRole = 'progressbar';
    const accessibilityValue = isIndeterminate
        ? { text: `Progress is indeterminate.` }
        : { text: `Current progress is, ${Math.round(progress)} percent.` };
    const trackTokenStyle = (0, tokenResolver_progressbar_1.resolveTrackTokens)({
        tokens: theme.progressBar,
        size: size,
        trackColor: trackColor
    });
    const indicatorTokenStyle = (0, tokenResolver_progressbar_1.resolveIndicatorTokens)({
        tokens: theme.progressBar,
        size: size,
        indicatorColor: indicatorColor
    });
    const resolvedIndicatorAnimationDuration = (0, tokenResolver_progressbar_1.resolveIndicatorAnimationDuration)({
        isIndeterminate: isIndeterminate,
        tokens: theme.progressBar,
        animationDuration: animationDuration
    });
    const progressedWidth = animatedProgress.interpolate({
        inputRange: [0, 100],
        outputRange: ['0%', '100%'],
        extrapolate: 'clamp'
    });
    const indeterminateIndicatorSizeRatio = 0.5;
    const indeterminateIndicatorTranslateX = (0, react_1.useRef)(new react_native_1.Animated.Value(-(barWidth * indeterminateIndicatorSizeRatio))).current;
    (0, react_1.useEffect)(() => {
        if (!isIndeterminate) {
            react_native_1.Animated.timing(animatedProgress, {
                toValue: progress,
                duration: resolvedIndicatorAnimationDuration,
                useNativeDriver: false
            }).start();
        }
    }, [progress, isIndeterminate]);
    (0, react_1.useEffect)(() => {
        if (isIndeterminate && barWidth != 0) {
            indeterminateIndicatorTranslateX.setValue(-(barWidth * indeterminateIndicatorSizeRatio));
            react_native_1.Animated.loop(react_native_1.Animated.timing(indeterminateIndicatorTranslateX, {
                toValue: barWidth,
                duration: resolvedIndicatorAnimationDuration,
                easing: react_native_1.Easing.linear,
                useNativeDriver: true
            })).start();
        }
    }, [barWidth, isIndeterminate]);
    const initializeBarLayout = (event) => {
        const { width } = event.nativeEvent.layout;
        setBarWidth(width);
    };
    return (<react_native_1.View onLayout={initializeBarLayout} accessibilityRole={accessibilityRole} accessibilityValue={accessibilityValue} style={[defaultBarStyle.trackAlignment, trackTokenStyle, barStyle]} {...rest}>
      {isIndeterminate ? (<Indicator_1.IndeterminateIndicator width={barWidth} sizeRatio={indeterminateIndicatorSizeRatio} position={indeterminateIndicatorTranslateX} tokenStyle={indicatorTokenStyle} style={indicatorStyle}/>) : (<Indicator_1.DeterminateIndicator progressedWidth={progressedWidth} tokenStyle={indicatorTokenStyle} style={indicatorStyle}/>)}
    </react_native_1.View>);
};
const defaultBarStyle = react_native_1.StyleSheet.create({
    trackAlignment: {
        justifyContent: 'flex-start'
    }
});
exports.default = ProgressBar;
