// @amazon-devices/react-native-w3cmedia/dist/components/mediacontrols/SeekBarUI/Theme/ThemeUtilities.js
"use strict";
/*
 * Copyright (c) 2022 - 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTheme = exports.ThemeContext = exports.defaultContext = exports.getDefaultTheme = exports.defaultThemes = void 0;
//Referenced {SeekBar,ProgressBar} folder from  - https://tiny.amazon.com/10ftogofw/KeplerComponentLibrary
const react_1 = __importStar(require("react"));
const LogUtils_1 = require("../../../../LogUtils");
const ProgressBarUtilities_1 = require("../ProgressBar/ProgressBarUtilities");
const makeTheme = (type) => {
    return {
        progressBar: (0, ProgressBarUtilities_1.progressBarTokens)(type),
    };
};
exports.defaultThemes = {
    light: makeTheme('light'),
    dark: makeTheme('dark')
};
const getDefaultTheme = (type = 'dark') => {
    return exports.defaultThemes[type];
};
exports.getDefaultTheme = getDefaultTheme;
exports.defaultContext = {
    theme: (0, exports.getDefaultTheme)(),
    setTheme: () => {
        LogUtils_1.LogUtil.info('[Theming] setTheme() is not implemented');
    }
};
exports.ThemeContext = react_1.default.createContext(exports.defaultContext);
/**
 *
 * The `useTheme` hook returns the current theme. If the theme is not set, or
 * useTheme is called outside the context of the ThemeProvider the default theme
 * will be returned. The theme can also be set using the `setTheme` function
 * provided by the {@link useSetTheme `useSetTheme`} hook.
 * @returns theme
 */
const useTheme = () => {
    const theme = (0, react_1.useContext)(exports.ThemeContext).theme;
    if (!theme) {
        return (0, exports.getDefaultTheme)();
    }
    return theme;
};
exports.useTheme = useTheme;
