// @amzn/keplermediadescriptor/dist/MediaCapabilitiesImpl.js
"use strict";
/*
 * Copyright 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoDecoderCapabilitiesImpl = exports.AudioDecoderCapabilitiesImpl = void 0;
const KeplerMediaDescriptorTurboModule_1 = __importDefault(require("./turbo-modules/KeplerMediaDescriptorTurboModule"));
/**
 * @hidden
 *
 * This impl is internal to the TM. When the application calls @name KeplerMediaDescriptor.queryMediaCapabilities,
 * The returned json object should be converted to a valid ts object. These implementations would help in the same.
 */
class AudioDecoderCapabilitiesImpl {
    mediaFormatProfileLevelCapabilities;
    mediaCodecFeaturesCapabilities;
    decoderFeaturesCapabilities;
    constructor(jsonObject) {
        this.mediaCodecFeaturesCapabilities = jsonObject.mediaCodecFeaturesCapabilities;
        this.decoderFeaturesCapabilities = jsonObject.decodeCapabilities;
        this.mediaFormatProfileLevelCapabilities = jsonObject.mediaFormatProfileLevelCapabilities;
    }
}
exports.AudioDecoderCapabilitiesImpl = AudioDecoderCapabilitiesImpl;
class VideoDecoderCapabilitiesImpl {
    mediaFormatProfileLevelCapabilities;
    mediaCodecFeaturesCapabilities;
    videoFormatCapabilities;
    decoderFeaturesCapabilities;
    constructor(jsonObject) {
        this.mediaCodecFeaturesCapabilities = jsonObject.mediaCodecFeaturesCapabilities;
        this.mediaFormatProfileLevelCapabilities = jsonObject.mediaFormatProfileLevelCapabilities;
        this.decoderFeaturesCapabilities = jsonObject.decoderFeaturesCapabilities;
        this.videoFormatCapabilities = new VideoFormatCapabilitiesImpl(jsonObject.videoFormatCapabilities);
    }
}
exports.VideoDecoderCapabilitiesImpl = VideoDecoderCapabilitiesImpl;
class VideoFormatCapabilitiesImpl {
    mID;
    resolutions;
    colorFormats;
    constructor(jsonObject) {
        this.mID = jsonObject.mID;
        this.resolutions = jsonObject.resolutions;
        this.colorFormats = jsonObject.colorFormats;
    }
    getMaxFrameRate(resolution) {
        return KeplerMediaDescriptorTurboModule_1.default.getMaxFrameRate(this, resolution);
    }
}
